{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# psbO gene   \n",
    "\n",
    "This is an attempt to link our ptMAGs to nuclear genetic data. The psbO gene encodes the manganese-stabilising polypeptide of the photosystem II oxygen evolving complex. It is a single copy gene found in nuclear genomes. "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "3.10.5 | packaged by conda-forge | (main, Jun 14 2022, 07:04:59) [GCC 10.3.0]\n"
     ]
    }
   ],
   "source": [
    "# Check if python is 3.10.5\n",
    "import json\n",
    "import os\n",
    "import pandas as pd\n",
    "import sys\n",
    "import numpy as np\n",
    "import __init__\n",
    "\n",
    "\n",
    "print(sys.version)\n",
    "%load_ext autoreload\n",
    "%autoreload 2"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [],
   "source": [
    "# we store the important data paths in PATH_FILE\n",
    "PATH_FILE = \"../../PATHS.json\"\n",
    "\n",
    "paths_dict = json.load(open(PATH_FILE, \"r\"));"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## 1. Most abundant psbO sequences in filters where Lepto-01 is abundant"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "We start by focusing on the filters where Lepto-01 is most abundant: \n",
    "\n",
    "- 194SUR1GGZZ11 (mean coverage of NEW = 124.15, 6th most abundant plastid lineage)  \n",
    "- 194SUR0CCKK11 (mean coverage of NEW = 70.54, 3rd most abundant plastid lineage)  \n",
    "\n",
    "Let's start by extracting the top 50 most abundant psbO sequences in these two filters.\n",
    "\n",
    "I downloaded the psbO database provided by [Karlusich et al 2022](https://doi.org/10.1111/1755-0998.13592), which also includes sequence abundance per Tara filter. "
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### 1.1 Extract most abundant psbO sequences"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [],
   "source": [
    "## Path to psbO database, and sequence list\n",
    "DATABASE = paths_dict[\"DATABASES\"][\"PSBO\"][\"ROOT\"]\n",
    "\n",
    "## Path to output folder \n",
    "OUT_DIR = paths_dict[\"ANALYSIS_DATA\"][\"PSBO\"][\"ABUNDANT_PSBO\"][\"DATA\"]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [],
   "source": [
    "%%bash -s \"$DATABASE\" \"$OUT_DIR\"\n",
    "\n",
    "cat \"$1\"/PsbO_metaG.tsv | \\\n",
    "tr '\\t' '@' | \\\n",
    "grep \"194@SUR@1@GGZZ@11\" | \\\n",
    "tr '@' '\\t' | \\\n",
    "sort -nrk 7 | \\\n",
    "head -50 | \\\n",
    "cut -f 6 \\\n",
    "> \"$2\"/194SUR1GGZZ11_psbO_top50.list\n",
    "\n",
    "cat \"$1\"/PsbO_metaG.tsv | \\\n",
    "tr '\\t' '@' | \\\n",
    "grep \"194@SUR@0@CCKK@11\" | \\\n",
    "tr '@' '\\t' | \\\n",
    "sort -nrk 7 | \\\n",
    "head -50 | \\\n",
    "cut -f 6 \\\n",
    "> \"$2\"/194SUR0CCKK11_psbO_top50.list"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "[INFO]\u001b[0m 50 patterns loaded from file\n",
      "[INFO]\u001b[0m 50 patterns loaded from file\n"
     ]
    }
   ],
   "source": [
    "%%bash -s \"$DATABASE\" \"$OUT_DIR\"\n",
    "\n",
    "seqkit grep -f \"$2\"/194SUR1GGZZ11_psbO_top50.list \"$1\"/psbO_20210825.fna > \"$2\"/194SUR1GGZZ11_psbO_top50.fasta\n",
    "seqkit grep -f \"$2\"/194SUR0CCKK11_psbO_top50.list \"$1\"/psbO_20210825.fna > \"$2\"/194SUR0CCKK11_psbO_top50.fasta"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "%%bash -s \"$OUT_DIR\"\n",
    "\n",
    "seqkit stats \"$1\"/194SUR1GGZZ11_psbO_top50.fasta\n",
    "seqkit stats \"$1\"/194SUR0CCKK11_psbO_top50.fasta"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Let's add some information to our fasta files, so that they contain: (1) order of abundance, (2) number of reads mapped."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {},
   "outputs": [],
   "source": [
    "%%bash -s \"$DATABASE\" \"$OUT_DIR\"\n",
    "\n",
    "## Create a file with the abundnace rank and the number of reads\n",
    "cat \"$1\"/PsbO_metaG.tsv | \\\n",
    "tr '\\t' '@' | \\\n",
    "grep \"194@SUR@1@GGZZ@11\" | \\\n",
    "tr '@' '\\t' | \\\n",
    "sort -nrk 7 | \\\n",
    "head -50 | \\\n",
    "cat -n | \\\n",
    "cut -f1,7,8 | \\\n",
    "sed -E 's/([0-9]+)\\t(.*)\\t([0-9]+)/\\2\\trank=\\1_reads=\\3/' | \\\n",
    "awk '{$1=$1;print}' | \\\n",
    "tr ' ' '\\t' \\\n",
    "> \"$2\"/194SUR1GGZZ11_psbO_top50.metadata\n",
    "\n",
    "## Create a file with the abundance rank and the number of reads\n",
    "cat \"$1\"/PsbO_metaG.tsv | \\\n",
    "tr '\\t' '@' | \\\n",
    "grep \"194@SUR@0@CCKK@11\" | \\\n",
    "tr '@' '\\t' | \\\n",
    "sort -nrk 7 | \\\n",
    "head -50 | \\\n",
    "cat -n | \\\n",
    "cut -f1,7,8 | \\\n",
    "sed -E 's/([0-9]+)\\t(.*)\\t([0-9]+)/\\2\\trank=\\1_reads=\\3/' | \\\n",
    "awk '{$1=$1;print}' | \\\n",
    "tr ' ' '\\t' \\\n",
    "> \"$2\"/194SUR0CCKK11_psbO_top50.metadata"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {},
   "outputs": [],
   "source": [
    "%%bash -s \"$OUT_DIR\"\n",
    "\n",
    "## Create a file with the taxonomy\n",
    "grep \">\" \"$1\"/194SUR1GGZZ11_psbO_top50.fasta | \\\n",
    "tr -d '>' | \\\n",
    "cut -f 1,2 -d ' ' | \\\n",
    "sed -E 's/(.*) (.*)/\\1\\t\\1_tax=\\2/' | \\\n",
    "tr ';' '_' \\\n",
    "> \"$1\"/194SUR1GGZZ11_psbO_top50.taxonomy\n",
    "\n",
    "## Create a file with the taxonomy\n",
    "grep \">\" \"$1\"/194SUR0CCKK11_psbO_top50.fasta | \\\n",
    "tr -d '>' | \\\n",
    "cut -f 1,2 -d ' ' | \\\n",
    "sed -E 's/(.*) (.*)/\\1\\t\\1_tax=\\2/' | \\\n",
    "tr ';' '_' \\\n",
    "> \"$1\"/194SUR0CCKK11_psbO_top50.taxonomy\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "I had to manually edit the taxonomy files a bit as the column content was not consistent across the fasta headers from which this file was generated."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Now generate the file to rename the fasta headers."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "metadata": {},
   "outputs": [],
   "source": [
    "%%bash -s \"$OUT_DIR\"\n",
    "\n",
    "join <(sort -k1 \"$1\"/194SUR1GGZZ11_psbO_top50.taxonomy) <(sort -k1 \"$1\"/194SUR1GGZZ11_psbO_top50.metadata) | \\\n",
    "tr ' ' '\\t' | \\\n",
    "sed -E 's/(.*)\\t(.*)\\t(.*)/\\1\\t\\2_\\3/' \\\n",
    "> \"$1\"/194SUR1GGZZ11_psbO_top50.rename\n",
    "\n",
    "join <(sort -k1 \"$1\"/194SUR0CCKK11_psbO_top50.taxonomy) <(sort -k1 \"$1\"/194SUR0CCKK11_psbO_top50.metadata) | \\\n",
    "tr ' ' '\\t' | \\\n",
    "sed -E 's/(.*)\\t(.*)\\t(.*)/\\1\\t\\2_\\3/' \\\n",
    "> \"$1\"/194SUR0CCKK11_psbO_top50.rename"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Rename the fasta headers now!"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "%%bash -s \"$OUT_DIR\"\n",
    "\n",
    "seqkit replace -p '^(\\S+)(.+?)$' -r '{kv}' -k \"$1\"/194SUR1GGZZ11_psbO_top50.rename \"$1\"/194SUR1GGZZ11_psbO_top50.fasta \\\n",
    "> \"$1\"/194SUR1GGZZ11_psbO_top50.renamed.fasta \n",
    "\n",
    "seqkit replace -p '^(\\S+)(.+?)$' -r '{kv}' -k \"$1\"/194SUR0CCKK11_psbO_top50.rename \"$1\"/194SUR0CCKK11_psbO_top50.fasta \\\n",
    "> \"$1\"/194SUR0CCKK11_psbO_top50.renamed.fasta "
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Finally, we can add the reference psbO sequences to the extracted sequences in order to build a phylogeny. "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "[INFO]\u001b[0m 346 patterns loaded from file\n"
     ]
    }
   ],
   "source": [
    "%%bash -s \"$DATABASE\" \"$OUT_DIR\"\n",
    "\n",
    "## extract mmetsp psbO sequences\n",
    "seqkit grep -nf \"$1\"/mmetsp.list \"$1\"/psbO_20210825.fna > \"$1\"/mmetsp.fasta\n",
    "\n",
    "## add to sequences from each filter\n",
    "cat \"$1\"/mmetsp.fasta \"$2\"/194SUR1GGZZ11_psbO_top50.renamed.fasta > \"$2\"/194SUR1GGZZ11_psbO_top50.reference.fasta\n",
    "\n",
    "cat \"$1\"/mmetsp.fasta \"$2\"/194SUR0CCKK11_psbO_top50.renamed.fasta > \"$2\"/194SUR0CCKK11_psbO_top50.reference.fasta"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### 1.2 Align and trim sequences"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 22,
   "metadata": {},
   "outputs": [],
   "source": [
    "## Path to input fasta file\n",
    "DATA = paths_dict[\"ANALYSIS_DATA\"][\"PSBO\"][\"ABUNDANT_PSBO\"][\"DATA\"]\n",
    "\n",
    "## Path to output folder \n",
    "OUT_DIR = paths_dict[\"ANALYSIS_DATA\"][\"PSBO\"][\"ABUNDANT_PSBO\"][\"ALIGNMENTS\"]"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Align with mafft-linsi!"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "%%bash -s \"$DATA\" \"$OUT_DIR\"\n",
    "\n",
    "sbatch ../../uppmax_scripts/script_bin/job_mafft-linsi.sh \"$1\"/194SUR0CCKK11_psbO_top50.reference.fasta \"$2\"/194SUR0CCKK11_psbO_top50.mafft.fasta\n",
    "\n",
    "sbatch ../../uppmax_scripts/script_bin/job_mafft-linsi.sh \"$1\"/194SUR1GGZZ11_psbO_top50.reference.fasta \"$2\"/194SUR1GGZZ11_psbO_top50.mafft.fasta"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Remove gaps and semicolons from files."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "metadata": {},
   "outputs": [],
   "source": [
    "%%bash -s \"$OUT_DIR\"\n",
    "\n",
    "cat \"$1\"/194SUR0CCKK11_psbO_top50.mafft.fasta | \\\n",
    "    tr ' ' '_' | \\\n",
    "    tr ';' '_' | \\\n",
    "    tr -d '(' | tr -d ')' \\\n",
    "    | tr -d '.' \\\n",
    "    > \"$1\"/194SUR0CCKK11_psbO_top50.mafft.edit.fasta\n",
    "\n",
    "cat \"$1\"/194SUR1GGZZ11_psbO_top50.mafft.fasta | \\\n",
    "    tr ' ' '_' | \\\n",
    "    tr ';' '_' | \\\n",
    "    tr -d '(' | tr -d ')' \\\n",
    "    | tr -d '.' \\\n",
    "    > \"$1\"/194SUR1GGZZ11_psbO_top50.mafft.edit.fasta"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Trim gently with trimal."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "%%bash -s \"$OUT_DIR\"\n",
    "\n",
    "sbatch ../../uppmax_scripts/script_bin/job_trimal_ssu.sh \"$1\"/194SUR0CCKK11_psbO_top50.mafft.edit.fasta \"$1\"/194SUR0CCKK11_psbO_top50.mafft.trimal.fasta\n",
    "\n",
    "sbatch ../../uppmax_scripts/script_bin/job_trimal_ssu.sh \"$1\"/194SUR1GGZZ11_psbO_top50.mafft.edit.fasta \"$1\"/194SUR1GGZZ11_psbO_top50.mafft.trimal.fasta"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "We found a couple of duplicates in our files (not sure why...). We remove these."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 23,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "[INFO]\u001b[0m 1 duplicated records removed\n",
      "[INFO]\u001b[0m 1 duplicated records removed\n"
     ]
    }
   ],
   "source": [
    "%%bash -s \"$OUT_DIR\"\n",
    "\n",
    "seqkit rmdup -n \"$1\"/194SUR0CCKK11_psbO_top50.mafft.trimal.fasta > temp\n",
    "mv temp \"$1\"/194SUR0CCKK11_psbO_top50.mafft.trimal.fasta \n",
    "\n",
    "seqkit rmdup -n \"$1\"/194SUR1GGZZ11_psbO_top50.mafft.trimal.fasta > temp\n",
    "mv temp \"$1\"/194SUR1GGZZ11_psbO_top50.mafft.trimal.fasta"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### 1.3 Run phylogenies!"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 24,
   "metadata": {},
   "outputs": [],
   "source": [
    "## Path to input fasta file\n",
    "ALIGNMENTS = paths_dict[\"ANALYSIS_DATA\"][\"PSBO\"][\"ABUNDANT_PSBO\"][\"ALIGNMENTS\"]\n",
    "\n",
    "## Path to output folder \n",
    "OUT_DIR = paths_dict[\"ANALYSIS_DATA\"][\"PSBO\"][\"ABUNDANT_PSBO\"][\"TREES\"]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "%%bash -s \"$ALIGNMENTS\" \"$OUT_DIR\"\n",
    "\n",
    "sbatch ../../uppmax_scripts/script_bin/job_2023_10_22_raxml-ng.sh \"$1\"/194SUR0CCKK11_psbO_top50.mafft.trimal.fasta \"$2\"/194SUR0CCKK11_psbO_top50\n",
    "sbatch ../../uppmax_scripts/script_bin/job_2023_10_22_raxml-ng.sh \"$1\"/194SUR1GGZZ11_psbO_top50.mafft.trimal.fasta \"$2\"/194SUR1GGZZ11_psbO_top50\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## 2. Search for psbO gene in Arctic asembled metagenomes   \n",
    "\n",
    "We would like to:  \n",
    "\n",
    "1) Search for psbO genes in the Arctic metagenomes (where NEW is most abundant)\n",
    "2) Correlate the abundance of all hits against the abundance of NEW\n",
    "3) Build a psbO phylogeny of potential candidates (fingers crossed we get some!). The NEW psbO gene should fall in a deep position in the tree.\n",
    "5) Try and find other genes from the same contig/MAG\n",
    "6) Use those to try and determine the identity of NEW"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### 2.1. Get data\n",
    "\n",
    "We downloaded the metagenomic co-assembly from the Arctic from https://www.genoscope.cns.fr/tara/. We downloaded both the contigs >2,500nt (which were used for binning), and >1,000nt."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### 2.2 Generate contigs database\n",
    "\n",
    "We use anvi'o to generate the contigs database. This runs Prodigal in the background for gene calling. This of course may not be the best approach for eukaryotes, but it's a good start! "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "## Path to input fasta file\n",
    "DATABASE = paths_dict[\"DATABASES\"][\"PSBO\"][\"ARCTIC\"]\n",
    "\n",
    "## Path to output folder \n",
    "OUT_DIR = paths_dict[\"ANALYSIS_DATA\"][\"PSBO\"][\"ARCTIC\"][\"CANDIDATES\"]"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Submit job for contigs above 2,500 nucleotides."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "%%bash -s \"$DATABASE\" \"$OUT_DIR\"\n",
    "\n",
    "sbatch ../../uppmax_scripts/script_bin/job_anvio_contigs_db.sh \"$1\"/TARA_ARC_GGZZ_SSUU_QQSS_KKQQ_2500nt.fa \"$2\"/TARA_ARC_CONTIGS_2500.db"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "And now for the contigs above 1,000 nucleotides."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "%%bash -s \"$DATABASE\" \"$OUT_DIR\"\n",
    "\n",
    "sbatch ../../uppmax_scripts/script_bin/job_anvio_contigs_db.sh \"$1\"/TARA_ARC_GGZZ_SSUU_QQSS_KKQQ_1000nt.fa \"$2\"/TARA_ARC_CONTIGS_1000.db"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### 2.3 Run HMM search\n",
    "We use an HMM search for the psbO gene based on the PFAM accession PF01716."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "## Path to psbO HMM profile\n",
    "DATABASE = paths_dict[\"DATABASES\"][\"PSBO\"][\"ROOT\"]\n",
    "\n",
    "## Path to output folder \n",
    "OUT_DIR = paths_dict[\"ANALYSIS_DATA\"][\"PSBO\"][\"ARCTIC\"][\"CANDIDATES\"]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "%%bash -s \"$DATABASE\" \"$OUT_DIR\"\n",
    "\n",
    "sbatch ../../uppmax_scripts/script_bin/job_anvio_hmm_search.sh \"$2\"/TARA_ARC_CONTIGS_2500.db \"$1\"/psbOHmms"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "We detected 51 hits."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "%%bash -s \"$DATABASE\" \"$OUT_DIR\"\n",
    "\n",
    "sbatch ../../uppmax_scripts/script_bin/job_anvio_hmm_search.sh \"$2\"/TARA_ARC_CONTIGS_1000.db \"$1\"/psbOHmms"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "We detected 84 hits."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### 2.4 Extract HMM hits\n",
    "Now we simply extract the HMM hits!"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "## Path to psbO HMM profile\n",
    "DATABASE = paths_dict[\"DATABASES\"][\"PSBO\"][\"ROOT\"]\n",
    "\n",
    "## Path to output folder \n",
    "OUT_DIR = paths_dict[\"ANALYSIS_DATA\"][\"PSBO\"][\"ARCTIC\"][\"CANDIDATES\"]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "%%bash -s \"$OUT_DIR\"\n",
    "\n",
    "sbatch ../../uppmax_scripts/script_bin/job_anvio_hmm_extract.sh \"$1\"/TARA_ARC_CONTIGS_2500.db psbOHmms \"$1\"/psbO_TARA_ARC_CONTIGS_2500.fasta\n",
    "\n",
    "sbatch ../../uppmax_scripts/script_bin/job_anvio_hmm_extract.sh \"$1\"/TARA_ARC_CONTIGS_1000.db psbOHmms \"$1\"/psbO_TARA_ARC_CONTIGS_1000.fasta"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### 1.5 Run tree!\n",
    "\n",
    "Now we can infer a phylogeny with the extracted psbO sequences to see where they go in the psbO tree. First let us trim and align the sequences."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "## Path to input fasta file\n",
    "DATABASE = paths_dict[\"ANALYSIS_DATA\"][\"PSBO\"][\"ARCTIC\"][\"PHYLOGENY\"][\"DATA\"]\n",
    "\n",
    "## Path to output folder \n",
    "OUT_DIR = paths_dict[\"ANALYSIS_DATA\"][\"PSBO\"][\"ARCTIC\"][\"PHYLOGENY\"][\"ALIGNMENTS\"]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "%%bash -s \"$DATABASE\" \"$OUT_DIR\"\n",
    "\n",
    "sbatch ../../uppmax_scripts/script_bin/job_mafft-linsi.sh \"$1\"/mmetsp_psbO_TARA_ARC_CONTIGS_2500.fasta \"$2\"/mmetsp_psbO_TARA_ARC_CONTIGS_2500.mafft.fasta\n",
    "\n",
    "sbatch ../../uppmax_scripts/script_bin/job_mafft-linsi.sh \"$1\"/mmetsp_psbO_TARA_ARC_CONTIGS_1000.fasta \"$2\"/mmetsp_psbO_TARA_ARC_CONTIGS_1000.mafft.fasta"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "We remove gaps and semicolons from fasta headers."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "%%bash -s \"$OUT_DIR\"\n",
    "\n",
    "cat \"$1\"/mmetsp_psbO_TARA_ARC_CONTIGS_2500.mafft.fasta | \\\n",
    "    tr ' ' '_' | \\\n",
    "    tr ';' '_' | \\\n",
    "    tr -d '(' | tr -d ')' | \\\n",
    "    tr ':' '_' | \\\n",
    "    tr -d '.' \\\n",
    "    > \"$1\"/mmetsp_psbO_TARA_ARC_CONTIGS_2500.mafft.edit.fasta\n",
    "\n",
    "cat \"$1\"/mmetsp_psbO_TARA_ARC_CONTIGS_1000.mafft.fasta | \\\n",
    "    tr ' ' '_' | \\\n",
    "    tr ';' '_' | \\\n",
    "    tr -d '(' | tr -d ')' | \\\n",
    "    tr ':' '_' | \\\n",
    "    tr -d '.' \\\n",
    "    > \"$1\"/mmetsp_psbO_TARA_ARC_CONTIGS_1000.mafft.edit.fasta"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Trim gently with trimal (keep columns with 90% data)."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "## Path to output folder \n",
    "DIR = paths_dict[\"ANALYSIS_DATA\"][\"PSBO\"][\"ARCTIC\"][\"PHYLOGENY\"][\"ALIGNMENTS\"]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "%%bash -s \"$DIR\"\n",
    "\n",
    "sbatch ../../uppmax_scripts/script_bin/job_trimal_ssu.sh \"$1\"/mmetsp_psbO_TARA_ARC_CONTIGS_2500.mafft.edit.fasta \"$1\"/mmetsp_psbO_TARA_ARC_CONTIGS_2500.mafft.trimal.fasta\n",
    "\n",
    "sbatch ../../uppmax_scripts/script_bin/job_trimal_ssu.sh \"$1\"/mmetsp_psbO_TARA_ARC_CONTIGS_1000.mafft.edit.fasta \"$1\"/mmetsp_psbO_TARA_ARC_CONTIGS_1000.mafft.trimal.fasta"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Now run a tree with raxml-ng!"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "## Path to input fasta file\n",
    "ALIGNMENTS = paths_dict[\"ANALYSIS_DATA\"][\"PSBO\"][\"ARCTIC\"][\"PHYLOGENY\"][\"ALIGNMENTS\"]\n",
    "\n",
    "## Path to output folder \n",
    "OUT_DIR = paths_dict[\"ANALYSIS_DATA\"][\"PSBO\"][\"ARCTIC\"][\"PHYLOGENY\"][\"TREES\"]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "%%bash -s \"$ALIGNMENTS\" \"$OUT_DIR\"\n",
    "\n",
    "sbatch ../../uppmax_scripts/script_bin/job_raxml-ng.sh \"$1\"/mmetsp_psbO_TARA_ARC_CONTIGS_2500.mafft.trimal.fasta \"$2\"/psbO_arctic_2500\n",
    "\n",
    "sbatch ../../uppmax_scripts/script_bin/job_raxml-ng.sh \"$1\"/mmetsp_psbO_TARA_ARC_CONTIGS_1000.mafft.trimal.fasta \"$2\"/psbO_arctic_1000"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "ptMAGs",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.10.5"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
