{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Targeted coassembly  \n",
    "\n",
    "The large Arctic co-assembly did not recover the psbO gene of NEW (at least not that we can see).\n",
    "\n",
    "We now try to do a targeted co-assembly of the filters where NEW is abundant in the hope that it better assembles the leptophyte nuclear genome.  \n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Check if python is 3.10.5\n",
    "import json\n",
    "import os\n",
    "import pandas as pd\n",
    "import sys\n",
    "import numpy as np\n",
    "import __init__\n",
    "\n",
    "\n",
    "print(sys.version)\n",
    "%load_ext autoreload\n",
    "%autoreload 2"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [],
   "source": [
    "# we store the important data paths in PATH_FILE\n",
    "PATH_FILE = \"../../PATHS.json\"\n",
    "\n",
    "paths_dict = json.load(open(PATH_FILE, \"r\"));"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## 1. Megahit\n",
    "\n",
    "I tried preliminary assemblies using MetaSpades, however, I quickly ran into memory issues (both with and without normalising the reads). I then tried a co-assembly with megahit as it very memory efficient. "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "metadata": {},
   "outputs": [],
   "source": [
    "## Path to psbO database, and sequence list\n",
    "DATA = paths_dict[\"DATABASES\"][\"COASSEMBLY\"]\n",
    "\n",
    "## Path to output folder \n",
    "ASSEMBLY = paths_dict[\"ANALYSIS_DATA\"][\"COASSEMBLY\"][\"MEGAHIT\"]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "%%bash -s \"$DATA\" \"$ASSEMBLY\"\n",
    "\n",
    "R1s=$(ls \"$1\"/*F_1.fastq.gz | tr '\\n' ',' | sed 's/.$//')\n",
    "R2s=$(ls \"$1\"/*F_2.fastq.gz | tr '\\n' ',' | sed 's/.$//')\n",
    "\n",
    "sbatch ../../uppmax_scripts/script_bin/job_megahit.sh \"$R1s\" \"$R2s\" 1000 \"$2\"/out"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Yay that worked!! Before going any further, let's try to see if we recover the psbO sequence of interest in our co-assembly. We start by simplifying the contig names."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [],
   "source": [
    "## Path to output folder \n",
    "ASSEMBLY = paths_dict[\"ANALYSIS_DATA\"][\"COASSEMBLY\"][\"MEGAHIT\"]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "%%bash -s \"$ASSEMBLY\"\n",
    "\n",
    "sbatch ../../uppmax_scripts/script_bin/job_anvio_simplify.sh \"$1\"/out/final.contigs.fa \"$1\"/out/final.contigs.fixed.fa"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "I replaced the contigs file manually by going to the folder and doing:\n",
    "\n",
    "`mv final.contigs.fixed.fa final.contigs.fa`"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## 2. Create contigs database\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [],
   "source": [
    "## Path to output folder \n",
    "ASSEMBLY = paths_dict[\"ANALYSIS_DATA\"][\"COASSEMBLY\"][\"MEGAHIT\"]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "%%bash -s \"$ASSEMBLY\" \n",
    "\n",
    "sbatch ../../uppmax_scripts/script_bin/job_anvio_contigs_db.sh \"$1\"/out/final.contigs.fa \"$1\"/out/final.contigs.db"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## 3. Run HMM search for psbO gene\n",
    "We use an HMM search for the psbO gene based on the PFAM accession PF01716."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [],
   "source": [
    "## Path to psbO HMM profile\n",
    "DATABASE = paths_dict[\"DATABASES\"][\"PSBO\"][\"ROOT\"]\n",
    "\n",
    "## Path to output folder \n",
    "ASSEMBLY = paths_dict[\"ANALYSIS_DATA\"][\"COASSEMBLY\"][\"MEGAHIT\"]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "%%bash -s \"$DATABASE\" \"$ASSEMBLY\"\n",
    "\n",
    "sbatch ../../uppmax_scripts/script_bin/job_anvio_hmm_search.sh \"$2\"/out/final.contigs.db \"$1\"/psbOHmms"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "We recovered only 16 hits for the psbO gene. "
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## 4. Extract HMM hits\n",
    "Now we simply extract the HMM hits."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [],
   "source": [
    "## Path to psbO HMM profile\n",
    "DATABASE = paths_dict[\"DATABASES\"][\"PSBO\"][\"ROOT\"]\n",
    "\n",
    "## Path to output folder \n",
    "ASSEMBLY = paths_dict[\"ANALYSIS_DATA\"][\"COASSEMBLY\"][\"MEGAHIT\"]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "%%bash -s \"$ASSEMBLY\"\n",
    "\n",
    "sbatch ../../uppmax_scripts/script_bin/job_anvio_hmm_extract.sh \"$1\"/out/final.contigs.db psbOHmms \"$1\"/out/psbO_top6filters.fasta"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## 5. Run tree!\n",
    "Now we can infer a phylogeny with the extracted psbO sequences to see where they go in the psbO tree. First let us add our extracted sequences to some reference sequences. "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [],
   "source": [
    "## Path to psbO database, and sequence list\n",
    "DATABASE = paths_dict[\"DATABASES\"][\"PSBO\"][\"ROOT\"]\n",
    "\n",
    "## Path to output folder \n",
    "OUT_DIR = paths_dict[\"ANALYSIS_DATA\"][\"COASSEMBLY\"][\"MEGAHIT\"]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [],
   "source": [
    "%%bash -s \"$DATABASE\" \"$OUT_DIR\"\n",
    "\n",
    "## add to sequences from each filter\n",
    "cat \"$1\"/mmetsp.fasta \"$2\"/out/psbO_top6filters.fasta > \"$2\"/out/psbO_top6filters.reference.fasta"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Align and trim sequences!"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "%%bash -s \"$OUT_DIR\"\n",
    "\n",
    "sbatch ../../uppmax_scripts/script_bin/job_mafft-linsi.sh \"$1\"/out/psbO_top6filters.reference.fasta \"$1\"/out/psbO_top6filters.reference.mafft.fasta"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Remove gaps and semicolons from files."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [],
   "source": [
    "%%bash -s \"$OUT_DIR\"\n",
    "\n",
    "cat \"$1\"/out/psbO_top6filters.reference.mafft.fasta | \\\n",
    "    tr ' ' '_' | \\\n",
    "    tr ';' '_' | \\\n",
    "    tr -d '(' | tr -d ')' \\\n",
    "    | tr -d '.' \\\n",
    "    > \"$1\"/out/psbO_top6filters.reference.mafft.edit.fasta"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Trim gently with trimal."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Submitted batch job 47719428 on cluster rackham\n"
     ]
    }
   ],
   "source": [
    "%%bash -s \"$OUT_DIR\"\n",
    "\n",
    "sbatch ../../uppmax_scripts/script_bin/job_trimal_ssu.sh \"$1\"/out/psbO_top6filters.reference.mafft.edit.fasta \"$1\"/out/psbO_top6filters.reference.mafft.trimal.fasta"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Now run treeeee!"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "%%bash -s \"$OUT_DIR\"\n",
    "\n",
    "sbatch ../../uppmax_scripts/script_bin/jobraxml-ng.sh \"$1\"/out/psbO_top6filters.reference.mafft.trimal.fasta \"$1\"/out/psbO_top6filters"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Doesn't look like we got the psbO from NEW. My guess is that it is simply not abundant enough for the nuclear genome to be assembled."
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "ptMAGs",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.10.5"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
