{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# 18S rRNA gene from the targeted coassembly  \n",
    "\n",
    "The targeted co-assembly of NEW-abundant filters did not recover a potential corresponding psbO gene (at least not that we can see).\n",
    "\n",
    "We now try to extract the 18S genes from the targeted coassembly with the idea that the 18S gene is more likely to be recovered than the psbO gene as it tends to (but not always) have multiple copies in genomes.\n",
    "\n",
    "I'll try to first extract the 18S gene using [barrnap](https://github.com/tseemann/barrnap)."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Check if python is 3.10.5\n",
    "import json\n",
    "import os\n",
    "import pandas as pd\n",
    "import sys\n",
    "import numpy as np\n",
    "import __init__\n",
    "\n",
    "\n",
    "print(sys.version)\n",
    "%load_ext autoreload\n",
    "%autoreload 2"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {},
   "outputs": [],
   "source": [
    "# we store the important data paths in PATH_FILE\n",
    "PATH_FILE = \"../../PATHS.json\"\n",
    "\n",
    "paths_dict = json.load(open(PATH_FILE, \"r\"));"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## 1. Extract 18S sequences from the filters where NEW is abundant\n",
    "\n",
    "We use barrnap with a very relaxed length cutoff of 0.1."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [],
   "source": [
    "## Path to assembly folder \n",
    "ASSEMBLY = paths_dict[\"ANALYSIS_DATA\"][\"COASSEMBLY\"][\"MEGAHIT\"]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "%%bash -s \"$ASSEMBLY\"\n",
    "\n",
    "sbatch ../../uppmax_scripts/script_bin/job_2024_09_barrnap.sh \"$1\"/out/final.contigs.fa \"$1\"/out/18S/rrna.fasta"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "We extract the 18S sequences only. "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "%%bash -s \"$ASSEMBLY\"\n",
    "\n",
    "seqkit grep -rp \"18S\" \"$1\"/out/18S/rrna.fasta > \"$1\"/out/18S/18S.fasta\n",
    "seqkit stats \"$1\"/out/18S/18S.fasta"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "We got 95 18S sequences, from 468 to 1800 bp. What are these 18S sequences? Let us find out! "
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## 2. BLAST 18S sequences against PR2 v5\n",
    "\n",
    "We BLAST our extracted 18S sequences against [PR2 v5](https://github.com/pr2database/pr2database/releases/tag/v5.0.0) to try and identify what is present in our metagenomes. Are there things that are very unsimilar to known reference sequences?"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {},
   "outputs": [],
   "source": [
    "## Path to assembly folder \n",
    "ASSEMBLY = paths_dict[\"ANALYSIS_DATA\"][\"COASSEMBLY\"][\"MEGAHIT\"]\n",
    "\n",
    "## Path to PR2 blast database\n",
    "PR2 = paths_dict[\"DATABASES\"][\"COASSEMBLY\"][\"PR2\"]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "%%bash -s \"$ASSEMBLY\" \"$PR2\"\n",
    "\n",
    "sbatch ../../uppmax_scripts/script_bin/job_blastn.sh \"$1\"/out/18S/18S.fasta \"$2\"/pr2_version_5.0.0_SSU_taxo_long.db \"$1\"/out/18S"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "We pull the best blast hit for each query."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "metadata": {},
   "outputs": [],
   "source": [
    "%%bash -s \"$ASSEMBLY\" \n",
    "\n",
    "## Extract best blast hit for each 18S seq\n",
    "cat \"$1\"/out/18S/18S__pr2_version_5.blastout | sort -k1,1 -k12,12nr | awk '!seen[$1]++' > \"$1\"/out/18S/18S__pr2_version_5.blastout.bestHit"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "I looked at the best hits manually. Most 18S sequences were actually 16S sequences from plastids and mitochondria, leaving behind only 29 nuclear 18S sequences. Out of these most were >90% to known references (animals, haptophytes, dinoflagellates, ciliates etc). Around four were less than 90% similar, and I looked at these sequences in more detail and BLASTed them against NCBI's nt_core. Three of these sequences could be convincingly placed in existing groups. One sequence (18S_RRNA::C_000000708636:221-1073) looked very different, and it initially excited me. However, a vast proportion of the sequence simply had no hits (when I blasted in bits) and is likely a chimeric sequence from bits of 18S and non-18S sequence. \n",
    "\n",
    "To summarise, the search for the 18S of NEW failed looking at just the assembled metagenomes. I will try looking at the 18S ASVs from these filters."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## References\n",
    "\n",
    "Seemann T. barrnap 0.9 : rapid ribosomal RNA prediction. https://github.com/tseemann/barrnap\n",
    "\n",
    "Guillou, L., Bachar, D., Audic, S., Bass, D., Berney, C., Bittner, L., ... & Christen, R. (2012). The Protist Ribosomal Reference database (PR2): a catalog of unicellular eukaryote small sub-unit rRNA sequences with curated taxonomy. Nucleic acids research, 41(D1), D597-D604. \n"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.10.5"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
