{
 "cells": [
  {
   "cell_type": "markdown",
   "id": "f8535d94",
   "metadata": {},
   "source": [
    "# Plastid MAGs\n",
    "\n",
    "In this notebook, we annotate the plastid MAGs with MFannot and convert the resulting masterfiles to genbank format.  We began with a redundant set of 902 plastid MAGs. The final curated set of 660 non-redundant plastid MAGs is provided at the Figshare repository associated with the manuscript: https://doi.org/10.17044/scilifelab.28212173.\n",
    "\n",
    "### Settings"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "5e9c56e9-b915-4a94-a98f-35d6458ddd78",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "3.10.5 | packaged by conda-forge | (main, Jun 14 2022, 07:04:59) [GCC 10.3.0]\n"
     ]
    }
   ],
   "source": [
    "# Check if python is 3.10.5\n",
    "import json\n",
    "import os\n",
    "import pandas as pd\n",
    "import sys\n",
    "import numpy as np\n",
    "import __init__\n",
    "\n",
    "\n",
    "print(sys.version)\n",
    "%load_ext autoreload\n",
    "%autoreload 2"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "1e445097-c8a6-4d53-bd7c-4ec655e4da68",
   "metadata": {},
   "outputs": [],
   "source": [
    "# we store the important data paths in PATH_FILE\n",
    "PATH_FILE = \"../../PATHS.json\"\n",
    "\n",
    "paths_dict = json.load(open(PATH_FILE, \"r\"));"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "id": "635fd994-3019-4f19-a2d9-4e1f7a6c2fdc",
   "metadata": {},
   "outputs": [],
   "source": [
    "pd.set_option('display.max_columns', None)\n",
    "pd.set_option('display.max_colwidth', 150)\n",
    "pd.set_option('display.max_rows', None)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "29f0e25a",
   "metadata": {},
   "source": [
    "Preliminary analyses compared how MFannot performed by extending the proteins collection with plastomes from RefSeq (both including and excluding the highly over-represented Viridiplantae). Thomas did not find a difference in performance, so all downstream performances were carried out with the proteins collection shipped with MFannot."
   ]
  },
  {
   "cell_type": "markdown",
   "id": "25dbf53c",
   "metadata": {},
   "source": [
    "## 1. Run MFannot\n",
    "We annotate all the plastid MAGs with MFannot."
   ]
  },
  {
   "cell_type": "markdown",
   "id": "fe0fd9b4",
   "metadata": {},
   "source": [
    "We first define the input and output directories."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "68f2c05f",
   "metadata": {},
   "outputs": [],
   "source": [
    "## Define directory with samples (SMP_DIR). Paths are defined in PATHS.json in the main directory.\n",
    "SMP_DIR = paths_dict['DATABASES'][\"TARA_PLASTID_GENOMES\"]\n",
    "\n",
    "## Define output directory\n",
    "MFOUT_DIR = paths_dict[\"ANALYSIS_DATA\"][\"PLASTOMES\"][\"MFANNOTATIONS\"][\"MASTERFILE\"][\"ROOT\"]\n",
    "MFSLURM_CSV = paths_dict[\"ANALYSIS_DATA\"][\"PLASTOMES\"][\"MFANNOTATIONS\"][\"MASTERFILE\"][\"SLURMLOG\"]\n",
    "\n",
    "## Define MFannot database\n",
    "PROTEIN_COLLECTION_DB = paths_dict[\"DATABASES\"][\"MF_ANNOT_REFS\"][\"PROTEINS\"]"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "d11d2120",
   "metadata": {},
   "source": [
    "For all our genomes we start a batch of MFannot runs. Jobs are submitted through the Uppmax `SlurmSubmitter`. And as defined above, jobs are tracked in `/crex/proj/naiss2023-6-81/Mahwash/ptMAGs/00_plastid_MAGS/results/mfannot/2023_10_mfannot_slurmlog.csv`."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "id": "eafb4a2c",
   "metadata": {},
   "outputs": [],
   "source": [
    "from plastome_raw_data import PlastomeRawIterator"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "ced33c96",
   "metadata": {},
   "outputs": [],
   "source": [
    "pri = PlastomeRawIterator(SMP_DIR, suffix=\"fa\")\n",
    "\n",
    "pri.run_mfannot(MFOUT_DIR, MFSLURM_CSV, PROTEIN_COLLECTION_DB, force=False, restart_fails=False)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "51e92b62",
   "metadata": {},
   "source": [
    "### Track jobs\n",
    "Jobs can be tracked below:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "id": "345239b0",
   "metadata": {},
   "outputs": [],
   "source": [
    "from dateutil import parser\n",
    "import matplotlib.pyplot as plt\n",
    "import seaborn as sns\n",
    "\n",
    "%matplotlib inline\n",
    "sns.set_theme(style=\"whitegrid\", palette=\"pastel\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "7e205ab2",
   "metadata": {},
   "outputs": [],
   "source": [
    "job_df = pd.read_csv(MFSLURM_CSV)\n",
    "\n",
    "job_df.sort_values(by=[\"duration\"])"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "c52a38e7",
   "metadata": {},
   "source": [
    "## 2. Convert to GenBank Format\n",
    "\n",
    "We used the asn2gb tool from NCBI to convert the sqn files (files in ASN.1 syntax that contains the sequences, their features, and the metadata about submission to NCBI) to flat genbank files."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "id": "8f1ff020",
   "metadata": {},
   "outputs": [],
   "source": [
    "## Define directory with sqn files (SQN_DIR). Paths are defined in PATHS.json in the main directory.\n",
    "SQN_DIR = paths_dict[\"ANALYSIS_DATA\"][\"PLASTOMES\"][\"MFANNOTATIONS\"][\"MASTERFILE\"][\"ROOT\"]\n",
    "\n",
    "## Define output directory\n",
    "GBOUT_DIR = paths_dict[\"ANALYSIS_DATA\"][\"PLASTOMES\"][\"MFANNOTATIONS\"][\"GENBANK\"][\"ROOT\"]\n",
    "\n",
    "## Define slurm csv to track jobs \n",
    "MFSLURM_CSV = paths_dict[\"ANALYSIS_DATA\"][\"PLASTOMES\"][\"MFANNOTATIONS\"][\"GENBANK\"][\"SLURMLOG\"]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "id": "3146f0af",
   "metadata": {},
   "outputs": [],
   "source": [
    "from sqn2gb import PlastomeSQNIterator"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "de216dba",
   "metadata": {},
   "outputs": [],
   "source": [
    "psi = PlastomeSQNIterator(SQN_DIR, GBOUT_DIR, MFSLURM_CSV)\n",
    "\n",
    "psi.run_asn2gb()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "02221874",
   "metadata": {},
   "source": [
    "## 3. Extract proteome\n",
    "We extract the proteome from the GenBank files."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "5003a65e",
   "metadata": {},
   "outputs": [],
   "source": [
    "## Define directory with GenBank files\n",
    "MAG_GB = paths_dict['ANALYSIS_DATA'][\"PLASTOMES\"][\"MFANNOTATIONS\"][\"GENBANK\"][\"ROOT\"]\n",
    "\n",
    "## Define output directory to store proteomes\n",
    "MAG_PROT = paths_dict['ANALYSIS_DATA'][\"PLASTOMES\"][\"MFANNOTATIONS\"][\"PROTEOMES\"]\n",
    "\n",
    "## Define whether the file is a reference (from NCBI) or a MAG (option between \"ref\" and \"mag\").\n",
    "FILE_TYPE = \"mag\""
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 19,
   "id": "0f41ef44",
   "metadata": {},
   "outputs": [],
   "source": [
    "from gb_to_prot import mfannot_gb_prot"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 20,
   "id": "b264d00b",
   "metadata": {},
   "outputs": [],
   "source": [
    "mfannot_gb_prot(MAG_GB, MAG_PROT, FILE_TYPE)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "10bcb3de",
   "metadata": {},
   "source": [
    "Now we have all the proteomes from the plastid MAGs!"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "745efba8",
   "metadata": {},
   "source": [
    "## 4. Mitochondrial contaminants\n",
    "\n",
    "At a much later point, we identified that one of the ptMAGs had mitochondrial contamination (thanks to R2 for discovering this after the first submission)!! This is something we had not previously checked! So we will now systematically check all our ptMAGs for mitochondrial genes. I'm curious to see if the contaminant ptMAGs are assembly chimeras (one contig contains both plastid and mitochondrial genes), or if plastid and mitochondrial contigs were accidentally binned together. \n",
    "\n",
    "### 4.1 Search for mito genes\n",
    "I put together a list of canonical mitochondrial genes from [Butenko et al 2024](https://doi.org/10.1186/s12915-024-01824-1), containing 25 genes. \n",
    "\n",
    "I then searched all the genbank files for the mito genes. "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "1ca2c5e5",
   "metadata": {
    "vscode": {
     "languageId": "powershell"
    }
   },
   "outputs": [],
   "source": [
    "grep -f mfannot/mito_genes.txt mfannot/genbank/* > mfannot/mito_contaminants.tsv\n",
    "\n",
    "cat mfannot/mito_contaminants.tsv | cut -f1 -d ' ' | uniq | wc -l"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "8525c606",
   "metadata": {},
   "source": [
    "51 redundant ptMAGs had mitochondrial gene contaminants. Manual checking revealed that this corresponded to only 6 non-redundant ptMAGs in our final database (and 45 redudnant ptMAGs that are not in the final database). Interestingly, this indiactes that reducing redundancy takes care of most low-quality ptMAGs.\n",
    "\n",
    "The list of contaminants/non-contaminants is stored in the file `mfannot/mito_contaminants_ptMAGS.txt`"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "e5258629",
   "metadata": {},
   "source": [
    "### 4.2 Manually checking for chimeras/binning errors\n",
    "\n",
    "I now manually check the 6 NR ptMAGs. \n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "6f27f7c4",
   "metadata": {},
   "source": [
    "**1. CHL_AON_Bin_184_91_c**  \n",
    "Contains 4 mitochondrial genes: nad5, nad4, nad9, nad4L\n",
    "\n",
    "All for mito genes (nad5, nad4, nad9, nad4L) were on C_4 (6645 bp) along with trnA, trnE, and rpl14. The rpl14 gene indeed seems closer to a mitochondrial gene upon BLASTing.\n",
    "\n",
    "**2. CHL_AON_Bin_218_19_c**  \n",
    "Contains 2 mitochondrial genes: atp6 and cob.\n",
    "\n",
    "Both genes were found on C_0 (4639 bp) along with 16S, trnM, trnX, and 23S. Only BLASTed the 16S - definitely mitochondrial.\n",
    "\n",
    "**3. CHL_AOS_Bin_125_33_c**  \n",
    "Contains the mito gene cox11.\n",
    "\n",
    "This gene was found on C_4 (3385 bp) along with ycf60 (copy 1), psbI, ycf12, psaC, trnL, trnC, ycf60 (copy2), rpl19, and ycf46. \n",
    "\n",
    "This struck me as a bit strange as the contig clearly contains genes involved in photosynthesis. Could it be a misassembly? I BLASTed the cox11 amino acid and nucleotide sequence but found no hits. Could be a misannotation. \n",
    "\n",
    "**4. CHL_ARC_Bin_238_20_c**  \n",
    "Contains 2 mito genes: cox1 and atp8. \n",
    "\n",
    "Both genes on C_0 (12507 bp) along with orf145, orf170, orf168, orf141, 23S (copy 1), trnM (copy1), 16S (copy 1), 5S, orf116, trnY, orf159, orf102, and tatC. The 16S sequence BLASTed to some Acetobacteria (with only 80% similarity). Unclear, but my guess is that the whole contig is mitochondrial. \n",
    "\n",
    "**5. CHL_ION_Bin_67_7_c**  \n",
    "Contains 12 mitochondrial genes!! atp6 (copy 1), cox3, atp8, nad3, nad2 (copy 1), nad5, cob, nad9, nad2 (copy 2), nad6, atp6 (copy 2), cox3. \n",
    "\n",
    "atp6, cox3, atp8, nad3, nad2  on C_3 (3454 bp) along with rps14.\n",
    "nad5, cob, nad9, nad2, nad6, atp6, cox3  on C_17 (6511 bp) along with trnR, trnF, and trnS.\n",
    "\n",
    "**6. CHL_IOS_Bin_126_17_c**  \n",
    "Contains 1 mito gene: nad10.\n",
    "\n",
    "Found on C_2 (3399 bp) along with rps4 (copy 1), trnM, trnH, trnF, trnC, rps13, trnM, 23S, and orf102. Both the 23S and rps4 gene clearly BLASTed to a mitochondrial genome. \n",
    "\n",
    "\n",
    "All the contaminating contigs were removed. "
   ]
  },
  {
   "cell_type": "markdown",
   "id": "182e80be",
   "metadata": {},
   "source": [
    "## Final ptMAGs dataset\n",
    "\n",
    "The final dataset consisted of 660 ptMAGs that were:\n",
    "- non-redundant (less than 98% average nucleotide identity with other ptMAGs and our selected references)  \n",
    "- checked for mitochondrial contamination  \n",
    "- checked for chimeras (through single gene tree inference)  \n",
    "- less than 15% redundant based on 44 core genes "
   ]
  },
  {
   "cell_type": "markdown",
   "id": "2b01e782-2760-408f-8fc6-6fbfd2e56cfa",
   "metadata": {},
   "source": [
    "## References\n",
    "\n",
    "Lang, B. F., Beck, N., Prince, S., Sarrasin, M., Rioux, P., & Burger, G. (2023). Mitochondrial genome annotation with MFannot: a critical analysis of gene identification and gene model prediction. Frontiers in Plant Science, 14.\n",
    "\n",
    "Beck, N., Lang, B. F. (2010) MFannot, organelle genome annotation websever. Available at: https://github.com/BFL-lab/Mfannot.\n",
    "\n",
    "asn2gb tool from NCBI. Available at: https://ftp.ncbi.nlm.nih.gov/asn1-converters/by_program/asn2gb/ \n",
    "\n",
    "Butenko, A., Luke≈°, J., Speijer, D., & Wideman, J. G. (2024). Mitochondrial genomes revisited: why do different lineages retain different genes?. BMC biology, 22(1), 15.\n"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "ptMAGs",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.10.5"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
