{
 "cells": [
  {
   "cell_type": "markdown",
   "id": "4162c4fb",
   "metadata": {},
   "source": [
    "# References\n",
    "\n",
    "In this notebook, we obtain the reference plastomes to be used in downstream phylogenetic analyses of plastid genomes.  \n",
    "\n",
    "### Settings"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "52514193-0cab-4e6e-b2e4-a6a8e46f811b",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "3.10.5 | packaged by conda-forge | (main, Jun 14 2022, 07:04:59) [GCC 10.3.0]\n"
     ]
    }
   ],
   "source": [
    "# Check if python is 3.10.5\n",
    "import json\n",
    "import os\n",
    "import pandas as pd\n",
    "import sys\n",
    "import __init__\n",
    "\n",
    "print(sys.version)\n",
    "%load_ext autoreload\n",
    "%autoreload 2"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "a1bc7d07-7d42-4a98-8abc-1f1ebc6e2ba8",
   "metadata": {},
   "outputs": [],
   "source": [
    "# we store the important data paths in PATH_FILE\n",
    "PATH_FILE = \"../../PATHS.json\"\n",
    "\n",
    "paths_dict = json.load(open(PATH_FILE, \"r\"));"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "3be57be2",
   "metadata": {},
   "source": [
    "## 1. Retrieve reference plastomes/genomes\n",
    "The reference plastomes and genomes used in downstream phylogenetic analyses were selected manually through several rounds of phylogenies, mostly from RefSeq. The aim was to include plastomes:\n",
    "- closely related to the ptMAGs,\n",
    "- that cover the broad diversity of plastid-bearing lineages,\n",
    "- while excluding very fast-evolving lineages (such as some dinoflagellate groups) that can affect phylogenetic analyses through long-branch attraction.\n",
    "\n",
    "The final list of selected references contained 178 accession numbers (NB: this number was updated later after many more rounds of preliminary phylogenies, these can be found at the Figshare repository: https://doi.org/10.17044/scilifelab.28212173).\n",
    " "
   ]
  },
  {
   "cell_type": "markdown",
   "id": "7efc094a",
   "metadata": {},
   "source": [
    "I downloaded the references manually (both genbank and fasta formats) through BatchEntrez (https://www.ncbi.nlm.nih.gov/sites/batchentrez). \n",
    "\n",
    "This worked for 178 records. Two could not be processed:\n",
    "\n",
    "```\n",
    "Id=GCF_000315565: nuccore: Wrong UID GCF_000315565\n",
    "Id=GCF_000173555: nuccore: Wrong UID GCF_000173555\n",
    "```"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "f5d24cb9",
   "metadata": {},
   "source": [
    "I downloaded 'GCF_000315565' and 'GCF_000173555' separately."
   ]
  },
  {
   "cell_type": "markdown",
   "id": "33bbaddd",
   "metadata": {},
   "source": [
    "I edited the fasta file to keep only the accession number."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "64476abc",
   "metadata": {},
   "outputs": [],
   "source": [
    "## Fasta file directory\n",
    "REF_FASTA = paths_dict['ANALYSIS_DATA'][\"REFERENCE_ORGANIZATION\"][\"FASTA_FILES\"]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "aa8b6e1c",
   "metadata": {},
   "outputs": [],
   "source": [
    "%%bash -s \"$REF_FASTA\"\n",
    "\n",
    "cat \"$1\"/sequence.fasta | sed -E 's/(>.*)\\.[0-9].*/\\1/' > fasta\n",
    "mv fasta \"$1\"/sequence.fasta "
   ]
  },
  {
   "cell_type": "markdown",
   "id": "dbacee84",
   "metadata": {},
   "source": [
    "I then split this fasta file into multiple files, with one plastid genome per file, and the file name corresponding to the sequence header. "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "90236ef6",
   "metadata": {},
   "outputs": [],
   "source": [
    "%%bash -s \"$REF_FASTA\"\n",
    "\n",
    "awk '/^>/{if(N){close(N)} N=substr($1,2) \".fa\"; print > N; next;} {if(N) print >> N}' \"$1\"/sequence.fasta\n",
    "\n",
    "rm \"$1\"/sequence.fasta"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "4bfba65b",
   "metadata": {},
   "source": [
    "## 2. Extract proteomes\n",
    "We converted the retrieved GenBank files to protein fasta files to enable downstream phylogenetic analyses. "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "id": "bf5ff5c7",
   "metadata": {},
   "outputs": [],
   "source": [
    "## Define directory with GenBank files\n",
    "REF_GB = paths_dict['ANALYSIS_DATA'][\"REFERENCE_ORGANIZATION\"][\"GENBANK_FILES\"]\n",
    "\n",
    "## Define output directory to store proteomes\n",
    "REF_PROT = paths_dict['ANALYSIS_DATA'][\"REFERENCE_ORGANIZATION\"][\"PROTEOMES\"]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "id": "1cfc7751",
   "metadata": {},
   "outputs": [],
   "source": [
    "from gb_to_prot import refseq_gb_prot"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "id": "078d1870",
   "metadata": {},
   "outputs": [],
   "source": [
    "refseq_gb_prot(REF_GB, REF_PROT)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "c8d4a8d6",
   "metadata": {},
   "source": [
    "Now we check if there are any empty proteome files, i.e. if some genbank to fasta file conversion failed."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "fe7fbef6",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Iterate through all files in the output directory\n",
    "for filename in os.listdir(REF_PROT):\n",
    "    file_path = os.path.join(REF_PROT, filename)\n",
    "    \n",
    "    # Check if the file is a regular file and its size is 0 bytes\n",
    "    if os.path.isfile(file_path) and os.path.getsize(file_path) == 0:\n",
    "        print(f\"Empty file found: {filename}\")"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "18cb6750",
   "metadata": {},
   "source": [
    "We have empty files corresponding to accession number LC716140, NC_012575, LC716139, NC_039967, in addition to the genomes GCF_000173555, and GCF_000315565. These files are empty either becuase the CDS sequences do not have gene information, or because the CDS sequences are missing. We can deal with these sequences manually. \n",
    "First, we get the corresponding nucleotide fasta file, and then run mfannot on that.  "
   ]
  },
  {
   "cell_type": "markdown",
   "id": "efb5d110",
   "metadata": {},
   "source": [
    "Run MFannot! "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "cca26102",
   "metadata": {},
   "outputs": [],
   "source": [
    "## Define directory with fasta files for genomes that failed in the last step\n",
    "## This is also the output folder for the mfannot analysis\n",
    "REF_MFANNOT = paths_dict['ANALYSIS_DATA'][\"REFERENCE_ORGANIZATION\"][\"MF_ANNOT_REFS\"][\"ROOT\"]\n",
    "\n",
    "## Define slurm csv to track jobs \n",
    "MFSLURM_CSV = paths_dict[\"ANALYSIS_DATA\"][\"REFERENCE_ORGANIZATION\"][\"MF_ANNOT_REFS\"][\"SLURMLOG\"]\n",
    "\n",
    "## Define MFannot database\n",
    "PROTEIN_COLLECTION_DB = paths_dict[\"DATABASES\"][\"MF_ANNOT_REFS\"][\"PROTEINS\"]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "4cf4c5fd",
   "metadata": {},
   "outputs": [],
   "source": [
    "from plastome_raw_data import PlastomeRawIterator"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "614e08ac",
   "metadata": {},
   "outputs": [],
   "source": [
    "pri = PlastomeRawIterator(REF_MFANNOT, suffix=\"fa\")\n",
    "\n",
    "pri.run_mfannot(REF_MFANNOT, MFSLURM_CSV, PROTEIN_COLLECTION_DB, force=False, restart_fails=False)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "9197e37a",
   "metadata": {},
   "source": [
    "Now we can convert the MFannot output file to the genbank format. We use the asn2gb tool from NCBI to do so."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "94117af3",
   "metadata": {},
   "outputs": [],
   "source": [
    "## Define directory with sqn files (SQN_DIR). Paths are defined in PATHS.json in the main directory.\n",
    "SQN_DIR = paths_dict['ANALYSIS_DATA'][\"REFERENCE_ORGANIZATION\"][\"MF_ANNOT_REFS\"][\"ROOT\"]\n",
    "\n",
    "## Define output directory\n",
    "GBOUT_DIR = paths_dict['ANALYSIS_DATA'][\"REFERENCE_ORGANIZATION\"][\"MF_ANNOT_REFS\"][\"ROOT\"]\n",
    "\n",
    "## Define slurm csv to track jobs \n",
    "MFSLURM_CSV = paths_dict['ANALYSIS_DATA'][\"REFERENCE_ORGANIZATION\"][\"MF_ANNOT_REFS\"][\"SLURMLOG_GB\"]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "511c6c69",
   "metadata": {},
   "outputs": [],
   "source": [
    "from sqn2gb import PlastomeSQNIterator"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "a1d4f860",
   "metadata": {},
   "outputs": [],
   "source": [
    "psi = PlastomeSQNIterator(SQN_DIR, GBOUT_DIR, MFSLURM_CSV)\n",
    "\n",
    "psi.run_asn2gb()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "18496a8e",
   "metadata": {},
   "source": [
    "We track the jobs."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "id": "697054ab",
   "metadata": {},
   "outputs": [],
   "source": [
    "from dateutil import parser\n",
    "import matplotlib.pyplot as plt\n",
    "import seaborn as sns\n",
    "\n",
    "%matplotlib inline\n",
    "sns.set_theme(style=\"whitegrid\", palette=\"pastel\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "3345e6a1",
   "metadata": {},
   "outputs": [],
   "source": [
    "job_df = pd.read_csv(MFSLURM_CSV)\n",
    "\n",
    "job_df.sort_values(by=[\"duration\"])"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "b9ff993d",
   "metadata": {},
   "source": [
    "Finally we can get the proteome from the genbank file."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "id": "1c1ef5b2",
   "metadata": {},
   "outputs": [],
   "source": [
    "## Define directory with GenBank files\n",
    "GBOUT_DIR = paths_dict['ANALYSIS_DATA'][\"REFERENCE_ORGANIZATION\"][\"MF_ANNOT_REFS\"][\"ROOT\"]\n",
    "\n",
    "## Define output directory to store proteomes\n",
    "REF_PROT = paths_dict['ANALYSIS_DATA'][\"REFERENCE_ORGANIZATION\"][\"PROTEOMES\"]\n",
    "\n",
    "## Define whether the file is a reference (from NCBI) or a MAG (opt between \"ref\" and \"mag\").\n",
    "FILE_TYPE = \"ref\""
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "id": "6deed73c",
   "metadata": {},
   "outputs": [],
   "source": [
    "from gb_to_prot import mfannot_gb_prot"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "id": "060bafcf",
   "metadata": {},
   "outputs": [],
   "source": [
    "mfannot_gb_prot(GBOUT_DIR, REF_PROT, FILE_TYPE)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "9d8c194b",
   "metadata": {},
   "source": [
    "Now we have all the proteomes from the reference sequences!!"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "ee852d15",
   "metadata": {},
   "source": [
    "## 3. Get taxonomy\n",
    "\n",
    "We have the accession numbers for the all the sequences, but we don't have the taxonomy information. Let's add this now to help interpret all out phylogenies. \n",
    "\n",
    "To get the taxonomic IDs and their associated lineage information, we use the a custom python script that looks up each accession number on NCBI and extracts the corresponding taxonomy string."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "id": "cf8be86a",
   "metadata": {},
   "outputs": [],
   "source": [
    "from accession2taxonomy import acc_to_taxo"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "id": "0e9dd8d9",
   "metadata": {},
   "outputs": [],
   "source": [
    "## List of accession numbers\n",
    "LIST = paths_dict[\"ANALYSIS_DATA\"]['REFERENCE_ORGANIZATION']['LIST']\n",
    "\n",
    "## Output taxonomy file\n",
    "TAXONOMY = paths_dict[\"ANALYSIS_DATA\"]['REFERENCE_ORGANIZATION']['TAXONOMY']\n",
    "\n",
    "## Type of database to query\n",
    "DATABASE = \"nuccore\""
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "e193bcf9",
   "metadata": {},
   "outputs": [],
   "source": [
    "acc_to_taxo(LIST, TAXONOMY, DATABASE)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "c09a5da8",
   "metadata": {},
   "source": [
    "We couldn't get the taxonomy information for two accession numbers: GCF_000315565 and GCF_000173555. We add this manually. \n",
    "This is what the file looks like:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 21,
   "id": "23555e0d",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "NC_022600\tBacteria_Cyanobacteriota_Cyanophyceae_Gloeobacterales_Gloeobacteraceae_Gloeobacter_kilaueensis_JS1\n",
      "NC_005125\tBacteria_Cyanobacteriota_Cyanophyceae_Gloeobacterales_Gloeobacteraceae_Gloeobacter_violaceus_PCC_7421\n",
      "CP017675\tBacteria_Cyanobacteriota_Cyanophyceae_Gloeomargaritales_Gloeomargaritaceae_Gloeomargarita_lithophora_Alchichica-D10\n",
      "CP088016\tBacteria_Cyanobacteriota_Cyanophyceae_Gloeomargaritales_cyanobacterium_VI4D9\n",
      "GCF_000315565\tBacteria_Cyanobacteriota_Cyanophyceae_Nostocales_Symphyonemataceae_Mastigocladopsis_repens_PCC_10914\n",
      "NC_019693\tBacteria_Cyanobacteriota_Cyanophyceae_Oscillatoriophycideae_Oscillatoriales_Oscillatoriaceae_Oscillatoria_acuminata_PCC_6304\n",
      "GCF_000173555\tBacteria_Cyanobacteriota_Cyanophyceae_Oscillatoriales_Sirenicapillariaceae_Limnospira_maxima_CS-328\n",
      "NZ_AP014638\tBacteria_Cyanobacteriota_Cyanophyceae_Leptolyngbyales_Leptolyngbyaceae_Leptolyngbya_boryana_IAM_M-101\n",
      "NC_009925\tBacteria_Cyanobacteriota_Cyanophyceae_Acaryochloridales_Acaryochloridaceae_Acaryochloris_marina_MBIC11017\n",
      "CP000239\tBacteria_Cyanobacteriota_Cyanophyceae_Synechococcales_Synechococcaceae_Synechococcus_sp_JA-3-3Ab\n",
      "\n"
     ]
    }
   ],
   "source": [
    "## head TAXONOMY file\n",
    "print(\"\".join(open(TAXONOMY).readlines()[:10]))"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "b7299a55",
   "metadata": {},
   "source": [
    "Nice! We want to add the taxonomy information to all the fasta headers in the proteome files (generated in the previous step). I want to change a couple of things:\n",
    "1. The taxonomy strings are super long, so I will manually edit them to retain only informative taxonomy ranks.\n",
    "2. PhyloFisher does not like the underscore character, so we should replace it with a dash. (NB: Ended up not using PhyloFisher at the end because it discards plastid data as bacterial contaminents.)\n",
    "\n",
    "Let's get started! We start by creating a third column and replacing '_' with '-'."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 23,
   "id": "c5cc410d",
   "metadata": {},
   "outputs": [],
   "source": [
    "%%bash -s \"$TAXONOMY\"\n",
    "cat $1 | awk 'BEGIN{FS=OFS=\"\\t\"} {gsub(\"_\", \"-\", $2)} 1' | sed -E 's/(.*)\\t(.*)/\\1\\t\\2\\ttaxo=\\2/' > txt\n",
    "mv txt $1"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "f19c79a0",
   "metadata": {},
   "source": [
    "Let's examine the file now."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 24,
   "id": "25d5193f",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "NC-022600\tBacteria-Cyanobacteriota-Cyanophyceae-Gloeobacterales-Gloeobacteraceae-Gloeobacter-kilaueensis-JS1\ttaxo=Bacteria-Cyanobacteriota-Cyanophyceae-Gloeobacterales-Gloeobacteraceae-Gloeobacter-kilaueensis-JS1\n",
      "NC-005125\tBacteria-Cyanobacteriota-Cyanophyceae-Gloeobacterales-Gloeobacteraceae-Gloeobacter-violaceus-PCC-7421\ttaxo=Bacteria-Cyanobacteriota-Cyanophyceae-Gloeobacterales-Gloeobacteraceae-Gloeobacter-violaceus-PCC-7421\n",
      "CP017675\tBacteria-Cyanobacteriota-Cyanophyceae-Gloeomargaritales-Gloeomargaritaceae-Gloeomargarita-lithophora-Alchichica-D10\ttaxo=Bacteria-Cyanobacteriota-Cyanophyceae-Gloeomargaritales-Gloeomargaritaceae-Gloeomargarita-lithophora-Alchichica-D10\n",
      "CP088016\tBacteria-Cyanobacteriota-Cyanophyceae-Gloeomargaritales-cyanobacterium-VI4D9\ttaxo=Bacteria-Cyanobacteriota-Cyanophyceae-Gloeomargaritales-cyanobacterium-VI4D9\n",
      "GCF-000315565\tBacteria-Cyanobacteriota-Cyanophyceae-Nostocales-Symphyonemataceae-Mastigocladopsis-repens-PCC-10914\ttaxo=Bacteria-Cyanobacteriota-Cyanophyceae-Nostocales-Symphyonemataceae-Mastigocladopsis-repens-PCC-10914\n",
      "NC-019693\tBacteria-Cyanobacteriota-Cyanophyceae-Oscillatoriophycideae-Oscillatoriales-Oscillatoriaceae-Oscillatoria-acuminata-PCC-6304\ttaxo=Bacteria-Cyanobacteriota-Cyanophyceae-Oscillatoriophycideae-Oscillatoriales-Oscillatoriaceae-Oscillatoria-acuminata-PCC-6304\n",
      "GCF-000173555\tBacteria-Cyanobacteriota-Cyanophyceae-Oscillatoriales-Sirenicapillariaceae-Limnospira-maxima-CS-328\ttaxo=Bacteria-Cyanobacteriota-Cyanophyceae-Oscillatoriales-Sirenicapillariaceae-Limnospira-maxima-CS-328\n",
      "NZ-AP014638\tBacteria-Cyanobacteriota-Cyanophyceae-Leptolyngbyales-Leptolyngbyaceae-Leptolyngbya-boryana-IAM-M-101\ttaxo=Bacteria-Cyanobacteriota-Cyanophyceae-Leptolyngbyales-Leptolyngbyaceae-Leptolyngbya-boryana-IAM-M-101\n",
      "NC-009925\tBacteria-Cyanobacteriota-Cyanophyceae-Acaryochloridales-Acaryochloridaceae-Acaryochloris-marina-MBIC11017\ttaxo=Bacteria-Cyanobacteriota-Cyanophyceae-Acaryochloridales-Acaryochloridaceae-Acaryochloris-marina-MBIC11017\n",
      "CP000239\tBacteria-Cyanobacteriota-Cyanophyceae-Synechococcales-Synechococcaceae-Synechococcus-sp-JA-3-3Ab\ttaxo=Bacteria-Cyanobacteriota-Cyanophyceae-Synechococcales-Synechococcaceae-Synechococcus-sp-JA-3-3Ab\n",
      "\n"
     ]
    }
   ],
   "source": [
    "## head updated TAXONOMY file\n",
    "print(\"\".join(open(TAXONOMY).readlines()[:10]))"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "01efa58e",
   "metadata": {},
   "source": [
    "Looking good. I will now manually edit the third column."
   ]
  },
  {
   "cell_type": "markdown",
   "id": "eb128108",
   "metadata": {},
   "source": [
    "We can look at the edited file now. "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 26,
   "id": "536ddb80",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "NC-022600\tBacteria-Cyanobacteriota-Cyanophyceae-Gloeobacterales-Gloeobacteraceae-Gloeobacter-kilaueensis-JS1\ttaxo=Cyanobacteriota-Gloeobacterales-Gloeobacter-kilaueensis\n",
      "NC-005125\tBacteria-Cyanobacteriota-Cyanophyceae-Gloeobacterales-Gloeobacteraceae-Gloeobacter-violaceus-PCC-7421\ttaxo=Cyanobacteriota-Gloeobacterales-Gloeobacter-violaceus\n",
      "CP017675\tBacteria-Cyanobacteriota-Cyanophyceae-Gloeomargaritales-Gloeomargaritaceae-Gloeomargarita-lithophora-Alchichica-D10\ttaxo=Cyanobacteriota-Gloeomargaritales-Gloeomargarita-lithophora-Alchichica\n",
      "CP088016\tBacteria-Cyanobacteriota-Cyanophyceae-Gloeomargaritales-cyanobacterium-VI4D9\ttaxo=Cyanobacteriota-Gloeomargaritales-cyanobacterium-VI4D9\n",
      "GCF-000315565\tBacteria-Cyanobacteriota-Cyanophyceae-Nostocales-Symphyonemataceae-Mastigocladopsis-repens-PCC-10914\ttaxo=Cyanobacteriota-Nostocales-Mastigocladopsis-repens\n",
      "NC-019693\tBacteria-Cyanobacteriota-Cyanophyceae-Oscillatoriophycideae-Oscillatoriales-Oscillatoriaceae-Oscillatoria-acuminata-PCC-6304\ttaxo=Cyanobacteriota-Oscillatoriales-Oscillatoria-acuminata\n",
      "GCF-000173555\tBacteria-Cyanobacteriota-Cyanophyceae-Oscillatoriales-Sirenicapillariaceae-Limnospira-maxima-CS-328\ttaxo=Cyanobacteriota-Oscillatoriales-Limnospira-maxima\n",
      "NZ-AP014638\tBacteria-Cyanobacteriota-Cyanophyceae-Leptolyngbyales-Leptolyngbyaceae-Leptolyngbya-boryana-IAM-M-101\ttaxo=Cyanobacteriota-Leptolyngbyales-Leptolyngbya-boryana\n",
      "NC-009925\tBacteria-Cyanobacteriota-Cyanophyceae-Acaryochloridales-Acaryochloridaceae-Acaryochloris-marina-MBIC11017\ttaxo=Cyanobacteriota-Acaryochloridales-Acaryochloris-marina\n",
      "CP000239\tBacteria-Cyanobacteriota-Cyanophyceae-Synechococcales-Synechococcaceae-Synechococcus-sp-JA-3-3Ab\ttaxo=Cyanobacteriota-Synechococcales-Synechococcus-sp\n",
      "\n"
     ]
    }
   ],
   "source": [
    "## head updated TAXONOMY file\n",
    "print(\"\".join(open(TAXONOMY).readlines()[:10]))"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "ef13266d",
   "metadata": {},
   "source": [
    "Let's append the taxonomy to the fasta files now. "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "id": "87951e68",
   "metadata": {},
   "outputs": [],
   "source": [
    "## Define directory containing proteomes (fasta files). Also the output folder.\n",
    "REF_PROT = paths_dict['ANALYSIS_DATA'][\"REFERENCE_ORGANIZATION\"][\"PROTEOMES\"]\n",
    "\n",
    "## Taxonomy mapping file\n",
    "TAXONOMY = paths_dict[\"ANALYSIS_DATA\"]['REFERENCE_ORGANIZATION']['TAXONOMY']"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "id": "11e482f7",
   "metadata": {},
   "outputs": [],
   "source": [
    "from rename_fasta import FastaIterator"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "id": "d920d079",
   "metadata": {},
   "outputs": [],
   "source": [
    "fi = FastaIterator(REF_PROT, REF_PROT, TAXONOMY)\n",
    "\n",
    "fi.prepend_taxo()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "9e250d7e-3ff0-44d5-a6c7-47b60c72b31e",
   "metadata": {
    "tags": []
   },
   "source": [
    "---\n",
    "# References\n",
    "Ponce-Toledo, R.I., Deschamps, P., López-García, P., Zivanovic, Y., Benzerara, K. and Moreira, D., 2017. An early-branching freshwater cyanobacterium at the origin of plastids. Current Biology, 27(3), pp.386-391.\n",
    " "
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "ptMAGs",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.10.5"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
