{
 "cells": [
  {
   "cell_type": "markdown",
   "id": "c56a37f8",
   "metadata": {},
   "source": [
    "# Overview\n",
    "\n",
    "We generate some stats about the MAGs and the references. How complete are the MAGs? How many genes do they encode? What is their size compared to the size of the references? What is the abundance of these MAGs in the global ocean?\n",
    "\n",
    "All stats are presented in Supplementary Data 1 of the corresponding manuscript (and are also present in the Figshare repo: https://doi.org/10.17044/scilifelab.28212173)."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "5ad51f0b-dd2f-49f0-b504-a7d504b02c17",
   "metadata": {
    "tags": []
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "3.10.5 | packaged by conda-forge | (main, Jun 14 2022, 07:04:59) [GCC 10.3.0]\n"
     ]
    }
   ],
   "source": [
    "# Check if python is 3.10.5\n",
    "import sys\n",
    "import os\n",
    "import pandas as pd\n",
    "import json\n",
    "import __init__\n",
    "\n",
    "print(sys.version)\n",
    "%load_ext autoreload\n",
    "%autoreload 2"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "0078e6c1-b4b2-4f51-a208-b3d3a8509abc",
   "metadata": {},
   "outputs": [],
   "source": [
    "# we store the important data paths in PATH_FILE\n",
    "PATH_FILE = \"../../PATHS.json\"\n",
    "\n",
    "paths_dict = json.load(open(PATH_FILE, \"r\"));"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "b618bae5",
   "metadata": {},
   "source": [
    "## 1. Completeness\n",
    "\n",
    "The first question we ask is how complete are the MAGs, and we compare to the references. To answer this question, we identified a set of 44 core genes that are found in all plastids ([Puerta et al 2005](https://doi.org/10.1093/dnares/12.2.151), and [Janouškovec et al 2010](https://doi.org/10.1073/pnas.100333510)). We use presence and absence of these genes to estimate completeness of all the MAGs and the reference taxa. "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "6f47192a",
   "metadata": {},
   "outputs": [],
   "source": [
    "## List of core genes\n",
    "CORE_GENES = paths_dict[\"DATABASES\"][\"CORE_GENES\"]\n",
    "\n",
    "## Output folder (which also contains list of taxa in Figure 1)\n",
    "OUT = paths_dict[\"ANALYSIS_DATA\"][\"PLASTOMES\"][\"STATS\"][\"ROOT\"]\n",
    "\n",
    "## Proteomes path for MAGS\n",
    "PROTEOMES_MAGS = paths_dict[\"ANALYSIS_DATA\"][\"PLASTOMES\"][\"MFANNOTATIONS\"][\"PROTEOMES\"]\n",
    "\n",
    "## Proteomes PATH for references\n",
    "PROTEOMES_REFS = paths_dict[\"ANALYSIS_DATA\"][\"REFERENCE_ORGANIZATION\"][\"PROTEOMES\"]"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "52962009",
   "metadata": {},
   "source": [
    "Calculate completeness for MAGs first."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "id": "fda16375",
   "metadata": {},
   "outputs": [],
   "source": [
    "%%bash -s \"$CORE_GENES\" \"$OUT\" \"$PROTEOMES_MAGS\"\n",
    "\n",
    "printf \"Taxon\\tType\\tNumber of genes detected\\tCompleteness percentage\\n\" > $2/completeness_taxa.txt\n",
    "\n",
    "cat $2/mags.list | while read line; do\n",
    "    genes=$(grep -f $1 $3/\"$line\".fasta -w | cut -f1 -d ';' | sort -u | wc -l)\n",
    "    completeness=$(awk 'BEGIN {print ('$genes'/44*100)}')\n",
    "    printf \"$line\\tmag\\t$genes\\t$completeness\\n\" >> $2/completeness_taxa.txt\n",
    "done"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "af53c07a",
   "metadata": {},
   "source": [
    "Now estimate completeness for all references and add to the same file."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "id": "23995c31",
   "metadata": {},
   "outputs": [],
   "source": [
    "%%bash -s \"$CORE_GENES\" \"$OUT\" \"$PROTEOMES_REFS\"\n",
    "\n",
    "cat $2/refs.list | while read line; do\n",
    "    genes=$(grep -f $1 $3/\"$line\".fasta -w | cut -f1 -d ';' | sort -u | wc -l)\n",
    "    completeness=$(awk 'BEGIN {print ('$genes'/44*100)}')\n",
    "    printf \"$line\\tref\\t$genes\\t$completeness\\n\" >> $2/completeness_taxa.txt\n",
    "done"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "969b88cb",
   "metadata": {},
   "source": [
    "We can now plot the completeness of the MAGs vs. the references in R using the script `src/stats.R`. "
   ]
  },
  {
   "cell_type": "markdown",
   "id": "d8ca3753",
   "metadata": {},
   "source": [
    "## 2. Redundancy\n",
    "\n",
    "We can also calculate the redundancy of the MAGs and references using the same set of 44 genes."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "id": "a4c3ef7b",
   "metadata": {},
   "outputs": [],
   "source": [
    "## List of core genes\n",
    "CORE_GENES = paths_dict[\"DATABASES\"][\"CORE_GENES\"]\n",
    "\n",
    "## Output folder (which also contains list of taxa in Figure 1)\n",
    "OUT = paths_dict[\"ANALYSIS_DATA\"][\"PLASTOMES\"][\"STATS\"][\"ROOT\"]\n",
    "\n",
    "## Proteomes path for MAGS\n",
    "PROTEOMES_MAGS = paths_dict[\"ANALYSIS_DATA\"][\"PLASTOMES\"][\"MFANNOTATIONS\"][\"PROTEOMES\"]\n",
    "\n",
    "## Proteomes PATH for references\n",
    "PROTEOMES_REFS = paths_dict[\"ANALYSIS_DATA\"][\"REFERENCE_ORGANIZATION\"][\"PROTEOMES\"]"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "b9042db6",
   "metadata": {},
   "source": [
    "First MAGs!"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "id": "74313bd3",
   "metadata": {},
   "outputs": [],
   "source": [
    "%%bash -s \"$CORE_GENES\" \"$OUT\" \"$PROTEOMES_MAGS\"\n",
    "\n",
    "## read list of genes\n",
    "gene_list=($(cat $1))\n",
    "\n",
    "# Generate the header line with gene names\n",
    "header=\"Taxa\\t${gene_list[@]}\"\n",
    "\n",
    "# Initialize the content with taxa and zeros\n",
    "content=\"\"\n",
    "\n",
    "# Read list of taxa\n",
    "taxa=($(cat $2/mags.list))\n",
    "\n",
    "# Iterate through taxa and genes, check number of genes, and append the number\n",
    "for taxon in ${taxa[@]}; do\n",
    "    line=\"${taxon}\"\n",
    "    for gene in \"${gene_list[@]}\"; do\n",
    "        count=$(grep -w \"${gene}\" \"$3/${taxon}.fasta\" | wc -l)\n",
    "        line+=\"\\t${count}\"\n",
    "    done\n",
    "    content+=\"${line}\\n\"\n",
    "done\n",
    "\n",
    "# Combine header and content and write to gene_presence_absence.tsv\n",
    "echo -e \"$header\\n\" > \"$2\"/redundancy_taxa.txt\n",
    "echo -e \"$content\" >> \"$2\"/redundancy_taxa.txt"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "a4afc187",
   "metadata": {},
   "source": [
    "Then references!"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "id": "b6a176b5",
   "metadata": {},
   "outputs": [],
   "source": [
    "%%bash -s \"$CORE_GENES\" \"$OUT\" \"$PROTEOMES_REFS\"\n",
    "\n",
    "## read list of genes\n",
    "gene_list=($(cat $1))\n",
    "\n",
    "# Initialize the content with taxa and zeros\n",
    "content=\"\"\n",
    "\n",
    "# Read list of taxa\n",
    "taxa=($(cat $2/refs.list))\n",
    "\n",
    "# Iterate through taxa and genes, check number of genes, and append the number\n",
    "for taxon in ${taxa[@]}; do\n",
    "    line=\"${taxon}\"\n",
    "    for gene in \"${gene_list[@]}\"; do\n",
    "        count=$(grep -w \"${gene}\" \"$3/${taxon}.fasta\" | wc -l)\n",
    "        line+=\"\\t${count}\"\n",
    "    done\n",
    "    content+=\"${line}\\n\"\n",
    "done\n",
    "\n",
    "# Write to redundancy_taxa.txt\n",
    "echo -e \"$content\" >> \"$2\"/redundancy_taxa.txt"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "ddb5e310",
   "metadata": {},
   "source": [
    "## 3. Size\n",
    "\n",
    "What is the total size of the MAGs and the references in number of bp? We calculate the size of each genome using [seqkit](https://bioinf.shenwei.me/seqkit/). "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "id": "71187064",
   "metadata": {},
   "outputs": [],
   "source": [
    "## Output folder (which also contains list of taxa in Figure 1)\n",
    "OUT = paths_dict[\"ANALYSIS_DATA\"][\"PLASTOMES\"][\"STATS\"][\"ROOT\"]\n",
    "\n",
    "## Fasta path for MAGS\n",
    "MAGS = paths_dict[\"DATABASES\"][\"TARA_PLASTID_GENOMES\"][\"REDUNDANCY_FILTERED\"]\n",
    "\n",
    "## Fasta PATH for references\n",
    "REFS = paths_dict[\"ANALYSIS_DATA\"][\"REFERENCE_ORGANIZATION\"][\"FASTA_FILES\"]"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "307e0965",
   "metadata": {},
   "source": [
    "Calculate size of MAGs first."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "id": "16975ecd",
   "metadata": {},
   "outputs": [],
   "source": [
    "%%bash -s \"$OUT\" \"$MAGS\"\n",
    "\n",
    "printf \"Taxon\\tType\\tSize\\n\" > $1/size_taxa.txt\n",
    "\n",
    "cat $1/mags.list | while read line; do\n",
    "    length=$(seqkit stats $2/\"$line\".fa -T | cut -f 5 | grep -v \"sum\")\n",
    "    printf \"$line\\tmag\\t$length\\n\" >> $1/size_taxa.txt\n",
    "done"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "126a223d",
   "metadata": {},
   "source": [
    "Now calculate the size of the references."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "id": "a7bf49ab",
   "metadata": {},
   "outputs": [],
   "source": [
    "%%bash -s \"$OUT\" \"$REFS\"\n",
    "\n",
    "cat $1/refs.list | while read line; do\n",
    "    length=$(seqkit stats $2/\"$line\".fa -T | cut -f 5 | grep -v \"sum\")\n",
    "    printf \"$line\\tref\\t$length\\n\" >> $1/size_taxa.txt\n",
    "done"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "757eca1c",
   "metadata": {},
   "source": [
    "Plot in R using the script `src/stats.R`. "
   ]
  },
  {
   "cell_type": "markdown",
   "id": "6dc88eca",
   "metadata": {},
   "source": [
    "## 4. GC content\n",
    "\n",
    "What is the GC content of each MAG and reference? We calculate it using seqkit again."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "19e3a2ad",
   "metadata": {},
   "outputs": [],
   "source": [
    "%%bash -s \"$OUT\" \"$MAGS\"\n",
    "\n",
    "printf \"Taxon\\tType\\tGCcontent\\n\" > $1/gc_taxa.txt\n",
    "\n",
    "cat $1/mags.list | while read line; do\n",
    "    gc=$(seqkit stats $2/\"$line\".fa -T -a | cut -f 16 | grep -v \"GC\")\n",
    "    printf \"$line\\tmag\\t$gc\\n\" >> $1/gc_taxa.txt\n",
    "done"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "3a52131e",
   "metadata": {},
   "source": [
    "And now the references!"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "7d0abc83",
   "metadata": {},
   "outputs": [],
   "source": [
    "%%bash -s \"$OUT\" \"$REFS\"\n",
    "\n",
    "cat $1/refs.list | while read line; do\n",
    "    gc=$(seqkit stats $2/\"$line\".fa -T -a | cut -f 16 | grep -v \"GC\")\n",
    "    printf \"$line\\tref\\t$gc\\n\" >> $1/gc_taxa.txt\n",
    "done"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "111eae02",
   "metadata": {},
   "source": [
    "Plot in R using the script `src/stats.R`. "
   ]
  },
  {
   "cell_type": "markdown",
   "id": "4c81424a",
   "metadata": {},
   "source": [
    "## 5. Taxonomy\n",
    "\n",
    "Getting the taxonomy of each MAG and reference is important for mapping around the tree. The taxonomy is derived from the position of the taxon in the full concatenated tree in Figure 1. \n",
    "\n",
    "We collect the taxonomy in the file `taxonomy_taxa.txt`."
   ]
  },
  {
   "cell_type": "markdown",
   "id": "7d8a04b6",
   "metadata": {},
   "source": [
    "At the same time, we can generate a file containing the original MAG/taxa names, the new MAG/taxa names, and the taxon names used in the phylogenies. We do this manually, and the file is called `correspondence_taxa.txt`"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "b48d98c8",
   "metadata": {},
   "source": [
    "## 6. Number of genes\n",
    "\n",
    "How many genes are encoded in the MAGs vs. the references? For the MAGs we have data from MFAnnot, while ther references have been annotated in different ways. For the sake of consistency, I decided to annotate the references with MFAnnot too.  "
   ]
  },
  {
   "cell_type": "markdown",
   "id": "91499fe4",
   "metadata": {},
   "source": [
    "### 6.1. Annotate references with MFAnnot for consistency"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "id": "07a28fb2",
   "metadata": {},
   "outputs": [],
   "source": [
    "## Define directory with samples (SMP_DIR). Paths are defined in PATHS.json in the main directory.\n",
    "REFS = paths_dict['ANALYSIS_DATA'][\"REFERENCE_ORGANIZATION\"][\"FASTA_FILES\"]\n",
    "\n",
    "## Define output directory\n",
    "MFOUT_DIR = paths_dict[\"ANALYSIS_DATA\"][\"REFERENCE_ORGANIZATION\"][\"MF_ANNOT_REFS\"][\"ROOT\"]\n",
    "MFSLURM_CSV = paths_dict[\"ANALYSIS_DATA\"][\"REFERENCE_ORGANIZATION\"][\"MF_ANNOT_REFS\"][\"SLURMLOG\"]\n",
    "\n",
    "## Define MFannot database\n",
    "PROTEIN_COLLECTION_DB = paths_dict[\"DATABASES\"][\"MF_ANNOT_REFS\"][\"PROTEINS\"]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "id": "c00a3951",
   "metadata": {},
   "outputs": [],
   "source": [
    "from plastome_raw_data import PlastomeRawIterator"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "980aaf97",
   "metadata": {},
   "outputs": [],
   "source": [
    "pri = PlastomeRawIterator(REFS, suffix=\"fa\")\n",
    "\n",
    "pri.run_mfannot(MFOUT_DIR, MFSLURM_CSV, PROTEIN_COLLECTION_DB, force=False, restart_fails=False)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "d0042849",
   "metadata": {},
   "source": [
    "### 6.2. Count number of genes\n",
    "\n",
    "Now we can get the number of genes for each MAG and reference sequence."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 19,
   "id": "41be2112",
   "metadata": {},
   "outputs": [],
   "source": [
    "## Output folder (which also contains list of taxa in Figure 1)\n",
    "OUT = paths_dict[\"ANALYSIS_DATA\"][\"PLASTOMES\"][\"STATS\"][\"ROOT\"]\n",
    "\n",
    "## MFAnnot path for MAGS\n",
    "MAGS = paths_dict[\"ANALYSIS_DATA\"][\"PLASTOMES\"][\"MFANNOTATIONS\"][\"MASTERFILE\"][\"ROOT\"]\n",
    "\n",
    "## Fasta PATH for references\n",
    "REFS = paths_dict[\"ANALYSIS_DATA\"][\"REFERENCE_ORGANIZATION\"][\"MF_ANNOT_REFS\"][\"ROOT\"]"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "d37a9e1c",
   "metadata": {},
   "source": [
    "Calculate number of genes of MAGs first."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 20,
   "id": "c67b59d0",
   "metadata": {},
   "outputs": [],
   "source": [
    "%%bash -s \"$OUT\" \"$MAGS\"\n",
    "\n",
    "printf \"Taxon\\tType\\tNumber of genes\\n\" > $1/no-of-genes_taxa.txt\n",
    "\n",
    "cat $1/mags.list | while read line; do\n",
    "    genes=$(grep \"Gene Totals\" $2/\"$line\".mf | cut -f2 -d ':' | tr -d ' ')\n",
    "    printf \"$line\\tmag\\t$genes\\n\" >> $1/no-of-genes_taxa.txt\n",
    "done"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "93af5bee",
   "metadata": {},
   "source": [
    "And now the references!"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 21,
   "id": "4d6b6202",
   "metadata": {},
   "outputs": [],
   "source": [
    "%%bash -s \"$OUT\" \"$REFS\"\n",
    "\n",
    "cat $1/refs.list | while read line; do\n",
    "    genes=$(grep \"Gene Totals\" $2/\"$line\".mf | cut -f2 -d ':' | tr -d ' ')\n",
    "    printf \"$line\\tref\\t$genes\\n\" >> $1/no-of-genes_taxa.txt\n",
    "done"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "258c89f4",
   "metadata": {},
   "source": [
    "Plot in R using the script `src/stats.R`. "
   ]
  },
  {
   "cell_type": "markdown",
   "id": "f2673a5b",
   "metadata": {},
   "source": [
    "---\n",
    "## References\n",
    "\n",
    "Puerta, M. V. S., Bachvaroff, T. R., & Delwiche, C. F. (2005). The complete plastid genome sequence of the haptophyte Emiliania huxleyi: a comparison to other plastid genomes. DNA Research, 12(2), 151-156. \n",
    "\n",
    "Janouškovec, J., Horák, A., Oborník, M., Lukeš, J., & Keeling, P. J. (2010). A common red algal origin of the apicomplexan, dinoflagellate, and heterokont plastids. Proceedings of the National Academy of Sciences, 107(24), 10949-10954.\n",
    "\n",
    "Shen, W., Le, S., Li, Y., & Hu, F. (2016). SeqKit: a cross-platform and ultrafast toolkit for FASTA/Q file manipulation. PloS one, 11(10), e0163962."
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "ptMAGs",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.10.5"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
