{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Synteny within leptophytes\n",
    "\n",
    "Here we look at how rearranged (if it all) the four plastid genomes of leptophytes are. \n",
    "\n",
    "One genome, REFM_CHLORO_00001, is composed of a single contig. However, the rest of the plastid genomes are composed of multiple contigs, and may not be in the correct \"order\". So we first fix that manually."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Check if python is 3.10.5\n",
    "import json\n",
    "import os\n",
    "import pandas as pd\n",
    "import sys\n",
    "import numpy as np\n",
    "import __init__\n",
    "\n",
    "\n",
    "print(sys.version)\n",
    "%load_ext autoreload\n",
    "%autoreload 2"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [],
   "source": [
    "# we store the important data paths in PATH_FILE\n",
    "PATH_FILE = \"../../PATHS.json\"\n",
    "\n",
    "paths_dict = json.load(open(PATH_FILE, \"r\"));"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "\n",
    "## 1. TARA_CHLORO_00478  \n",
    "This ptMAG (original name: CHL_MED_Bin_250_53_c) is the second most complete genome after REFM_CHLORO_00001. It is composed of six contigs. \n",
    "\n",
    "By comparing plastid maps of TARA_CHLORO_00478 with that of REFM_CHLORO_00001 manually, I determined that the contigs should be joined together in this order:\n",
    "- C_0 (should be reverse complemented)\n",
    "- C_3 (shares the genes psbE and trnG-TCC with C_0; should be reverse complemented)  \n",
    "- C_1 (a small gap between C_3 and C_1; should be reverse complemented)  \n",
    "- C_4 (no gap between C_1 and C_4)  \n",
    "- C_5 (shares the gene psaA with C_4; should be reverse complemented)  \n",
    "- C_2 (a gap between C_5 and C_2)  \n",
    "\n",
    "I manually placed the contigs in the correct order. \n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## 2. TARA_CHLORO_00158\n",
    "This ptMAG (original name: CHL_AOS_Bin_125_5_c) is the third most complete genome after REFM_CHLORO_00001. It is composed of three contigs. \n",
    "\n",
    "By comparing plastid maps of TARA_CHLORO_00158 with that of REFM_CHLORO_00001 manually, I determined that the contigs should be joined together in this order:  \n",
    "- C_1 (should be reverse complemented)  \n",
    "- C_0 (big gap between C_0 and C_1)    \n",
    "- C_2 (overlaps with C_0 for psbC and psbD)    \n",
    "\n",
    "As before, I manually placed the contigs in the correct order."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## 3. TARA_CHLORO_00332  \n",
    "This ptMAG (original name, CHL_ARC_Bin_77_4_c) is the least complete genome. It is composed of 15 contigs. \n",
    "\n",
    "By comparing plastid maps of TARA_CHLORO_00332 with that of REFM_CHLORO_00001 manually, I determined that the contigs should be joined together in this order:  \n",
    "\n",
    "- C_5   \n",
    "- C_12 (reverse complement)    \n",
    "- C_14 (reverse complement)   \n",
    "- C_2 (reverse complement)  \n",
    "- C_6  \n",
    "- C_13 (reverse complement)  \n",
    "- C_9 (reverse complement)  \n",
    "- C_3 (reverse complement)  \n",
    "- C_0  \n",
    "- C_11  \n",
    "- C_10  \n",
    "- C_1 (reverse complement)  \n",
    "- C_7   \n",
    "- C_4  \n",
    "- C_8     \n",
    "\n",
    "As before, I manually placed the contigs in the correct order."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## 4. Synteny analyses\n",
    "\n",
    "I used the genbank files to visualise synteny using the tool [PyGenomeViz](https://pygenomeviz.streamlit.app/). Sequence similarity was calculated using MMseqs for reciprocal best-hit CDS search.\n",
    "\n",
    "Overall, the genomes are very similar in terms of gene order and gene content, but there are some very minor differences. \n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### 4.1. Gene order\n",
    "\n",
    "Here, I focus more on protein coding genes, and not on tRNA sequences. Two main differences could be observed. \n",
    "\n",
    "1. **psaD inversion**  \n",
    "The psaD gene was inverted in the lineage comprising TARA_CHLORO_00478, TARA_CHLORO_00158, and TARA_CHLORO_00332.\n",
    "\n",
    "2. **psaB - groEL region inversion**  \n",
    "A nearly 10 kbp region spanning from the psaB gene to the groEL gene is inverted in the lineage comprising TARA_CHLORO_00478 and TARA_CHLORO_00158."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## References\n",
    "\n",
    "Parada, A. E., Needham, D. M., & Fuhrman, J. A. (2016). Every base matters: assessing small subunit rRNA primers for marine microbiomes with mock communities, time series and global field samples. Environmental microbiology, 18(5), 1403-1414. https://doi.org/10.1111/1462-2920.13023\n",
    "\n",
    "Vernette, C., Henry, N., Lecubin, J., de Vargas, C., Hingamp, P., & Lescot, M. (2021). The Ocean barcode atlas: A web service to explore the biodiversity and biogeography of marine organisms. Molecular Ecology Resources, 21(4), 1347-1358. https://doi.org/10.1111/1755-0998.13322\n",
    "\n",
    "Sichert, A., Corzett, C. H., Schechter, M. S., Unfried, F., Markert, S., Becher, D., ... & Hehemann, J. \n",
    "H. (2020). Verrucomicrobia use hundreds of enzymes to digest the algal polysaccharide fucoidan. Nature m\n",
    "icrobiology, 5(8), 1026-1039. https://doi.org/10.1038/s41564-020-0720-2"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "ptMAGs",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.10.5"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
