{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Gene content\n",
    "\n",
    "We know confidently that leptophytes are a red plastid lineage. But what is the affiliation of leptophytes? Are they a new red algal lineage? Are they a deep branching cryptophyte, or haptophyte? \n",
    "\n",
    "The plastid gene content can give clues about the affiliation of leptophytes. "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "3.10.5 | packaged by conda-forge | (main, Jun 14 2022, 07:04:59) [GCC 10.3.0]\n"
     ]
    }
   ],
   "source": [
    "# Check if python is 3.10.5\n",
    "import json\n",
    "import os\n",
    "import pandas as pd\n",
    "import sys\n",
    "import numpy as np\n",
    "import __init__\n",
    "\n",
    "\n",
    "print(sys.version)\n",
    "%load_ext autoreload\n",
    "%autoreload 2"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [],
   "source": [
    "# we store the important data paths in PATH_FILE\n",
    "PATH_FILE = \"../../PATHS.json\"\n",
    "\n",
    "paths_dict = json.load(open(PATH_FILE, \"r\"));"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## 1. Run OrthoFinder\n",
    "\n",
    "To confidently identify orthologs across plastid genomes (including the orf genes), and infer presence/absence patterns, we run OrthoFinder. We start by considering all the reference genomes used in the subset concatenated phylogeny + the Lepto-01 plastid genome."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [],
   "source": [
    "## Path to proteomes of selected taxa\n",
    "PROTEOMES = paths_dict[\"ANALYSIS_DATA\"][\"COMP_GEN\"][\"GENE_CONTENT\"]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "%%bash -s \"$PROTEOMES\"\n",
    "\n",
    "sbatch /home/mahja/ptMAGs/uppmax_scripts/script_bin/job_orthofinder.sh \"$1\"/."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Let's have a look at the results. \n",
    "\n",
    "First, we look at the file: `Comparative_Genomics_Statistics/Statistics_Overall.tsv`. OrthoFinder assigned 7648 genes (98.5% of genes) to 275 orthogroups. This looks good. \n",
    "\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "I manually checked the `N0.tsv` file produced by Orthofinder in the `Phylogenetic_Hierarchical_Orthogroups` folder. Heatmap of gene presence-absence (Figure 2) was generated with the R script `heatmap.R` in the `src` folder.  "
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## References\n",
    "\n",
    "Parada, A. E., Needham, D. M., & Fuhrman, J. A. (2016). Every base matters: assessing small subunit rRNA primers for marine microbiomes with mock communities, time series and global field samples. Environmental microbiology, 18(5), 1403-1414. https://doi.org/10.1111/1462-2920.13023\n",
    "\n",
    "Vernette, C., Henry, N., Lecubin, J., de Vargas, C., Hingamp, P., & Lescot, M. (2021). The Ocean barcode atlas: A web service to explore the biodiversity and biogeography of marine organisms. Molecular Ecology Resources, 21(4), 1347-1358. https://doi.org/10.1111/1755-0998.13322\n",
    "\n",
    "Sichert, A., Corzett, C. H., Schechter, M. S., Unfried, F., Markert, S., Becher, D., ... & Hehemann, J. \n",
    "H. (2020). Verrucomicrobia use hundreds of enzymes to digest the algal polysaccharide fucoidan. Nature m\n",
    "icrobiology, 5(8), 1026-1039. https://doi.org/10.1038/s41564-020-0720-2"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.10.5"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
