{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Reference SGTs\n",
    "\n",
    "In this notebook, we assemble a phylogenetic dataset from reference plastomes only. "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "3.10.5 | packaged by conda-forge | (main, Jun 14 2022, 07:04:59) [GCC 10.3.0]\n"
     ]
    }
   ],
   "source": [
    "# Check if python is 3.10.5\n",
    "import json\n",
    "import os\n",
    "import pandas as pd\n",
    "import sys\n",
    "import numpy as np\n",
    "import __init__\n",
    "\n",
    "\n",
    "print(sys.version)\n",
    "%load_ext autoreload\n",
    "%autoreload 2"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [],
   "source": [
    "# we store the important data paths in PATH_FILE\n",
    "PATH_FILE = \"../../PATHS.json\"\n",
    "\n",
    "paths_dict = json.load(open(PATH_FILE, \"r\"));"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## 1. Gene lists\n",
    "\n",
    "Which genes should we use to infer plastid phylogenies? The gene list was finalised in an iterative manner following many rounds of preliminary phylogenies. Here, we go through the final gene list and how it came about. \n",
    "\n",
    "We start with the gene dataset of [Janouškovec et al 2010](https://doi.org/10.1073/pnas.1003335107). This gene dataset includes 68 orthologous plastid proteins (the full dataset), and a subset of 34 slow evolving genes (subset).\n",
    "\n",
    "These gene_lists are defined here:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [],
   "source": [
    "## Full gene dataset\n",
    "JANO_FULL = paths_dict['DATABASES'][\"GENE_LISTS\"][\"JANO_FULL\"]\n",
    "\n",
    "## Subset of slow evolving genes\n",
    "JANO_CONS = paths_dict['DATABASES'][\"GENE_LISTS\"][\"JANO_CONS\"]"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Let's look at the JANO_FULL file."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "gene\tgene_name\n",
      "acsF\tacsF\n",
      "acsF\tycf59\n",
      "atpA\tatpA\n",
      "atpB\tatpB\n",
      "atpH\tatpH\n",
      "atpI\tatpI\n",
      "ccs1\tccs1\n",
      "ccs1\tycf44\n",
      "ccsA\tccsA\n",
      "\n"
     ]
    }
   ],
   "source": [
    "## head JANO_FILE\n",
    "print(\"\".join(open(JANO_FULL).readlines()[:10]))"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "The first column is the gene, while the second column lists all the commonly used gene names (there might be multiple). For example the gene acsF is also known as ycf59. "
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "We later came across [Ponce-Toledo et al 2017](https://doi.org/10.1016/j.cub.2016.11.056) and [Pietluch et al 2024](https://doi.org/10.1093/gbe/evae192) which both had slightly non-overlapping 97 gene datasets. We compared those datasets to our 68 gene set. 50 genes were shared between our dataset and the Ponce-Todeldo dataset, with 18 genes present *only* in our dataset. We found 47 genes were present only in the Ponce-Toledo dataset, out of which 31 were not present in leptophytes (meaning that it was not necessary to add those). This left an additional 16 genes which we could add to our dataset from Ponce-Toledo et al 2017. Similarly, there were 12 additional genes we could add from Pietluch et al 2024. "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [],
   "source": [
    "## New gene dataset\n",
    "GENES_ADD = paths_dict['DATABASES'][\"GENE_LISTS\"][\"GENES_ADD\"]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "atpF\n",
      "atpG\n",
      "atpD\n",
      "atpE\n",
      "groEL\n",
      "dnaK\n",
      "chlI\n",
      "psaF\n",
      "rpl1\n",
      "rpl12\n",
      "rpl21\n",
      "rpl33\n",
      "rps10\n",
      "rps20\n",
      "rps9\n",
      "rbcL\n",
      "cbbX\n",
      "petL\n",
      "petM\n",
      "psaI\n",
      "psaJ\n",
      "psaL\n",
      "psaM\n",
      "psbL\n",
      "psbX\n",
      "psbY\n",
      "rbcS\n",
      "rpl34\n"
     ]
    }
   ],
   "source": [
    "%%bash -s \"$GENES_ADD\"\n",
    "cat $1"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## 2. Single gene trees of ref taxa\n",
    "\n",
    "We first needed to extract the genes of interest from the reference taxa. (It is worth mentioning that we had done some preliminary phylogenetic analyses with sequences retrieved based on gene names - however, this approach led to errors in some cases due to inconsistency in gene names.)\n",
    "\n",
    "To confirm that the genes from the reference taxa are: \n",
    "- free of potential contaminants \n",
    "- free of paralogs (rare in plastids, but we should still check), \n",
    "\n",
    "we start by inferring single gene trees of the reference sequences only."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### 2.1 Extract gene fasta files"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Step 1 is extracting the 68 (later 93 in total) gene sequences from each taxon. In a preliminary analysis, we tried using the gene names themselves to do so - however this did not work super well due to differences in labelling and poor annotation in some cases (particularly for the cyanobacteria - for example the gene atpE was labelled as atpH in some cyanos).\n",
    "\n",
    "So instead, we decided to use BLAST to pull the relevant genes from each taxon. As query, we opted to use sequences from the brown alga _Fucus vesiculosus_, which has a well-annotated plastid genome ([Corguillé et al 2009](https://doi.org/10.1186/1471-2148-9-253)), and contains all 68 genes we are interested in. "
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### 2.1.1. Extract genes from _Fucus vesiculosus_\n",
    "\n",
    "We extract the genes from _Fucus vesiculosus_ which has the accession number NC_016735."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "## Folder containing proteomes of reference taxa\n",
    "## Provided here (in the ptMAGs folder): https://doi.org/10.17044/scilifelab.28212173\n",
    "REF_PROTEOMES = paths_dict['ANALYSIS_DATA']['SINGLE_GENE_ANALYSIS']['PROTEOMES']['REF_PROTEOMES']\n",
    "\n",
    "## File of gene set. We start by extracting the full dataset of Janouškovec et al 2010.\n",
    "JANO_FULL = paths_dict['DATABASES']['GENE_LISTS']['JANO_FULL']\n",
    "\n",
    "## Output folder where we place the gene extracts for Fucus\n",
    "## (Intermediate files - not provided)\n",
    "WORKING_DATASET = paths_dict['ANALYSIS_DATA']['SINGLE_GENE_ANALYSIS']['DATASET']['REF']['WORKING']['REF']"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "%%bash -s \"$REF_PROTEOMES\" \"$JANO_FULL\" \"$WORKING_DATASET\"\n",
    "\n",
    "cut -f 1 \"$2\" | uniq | grep -v \"gene\" | while read line; do\n",
    "    echo $line\n",
    "    seqkit grep -rp \"gene=$line;\" \"$1\"/NC_016735.taxo.fasta > \"$3\"/\"$line\".fasta\n",
    "done"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "And do the same for the additional 28 genes."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {},
   "outputs": [],
   "source": [
    "## Folder containing proteomes of reference taxa\n",
    "REF_PROTEOMES = paths_dict['ANALYSIS_DATA']['SINGLE_GENE_ANALYSIS']['PROTEOMES']['REF_PROTEOMES']\n",
    "\n",
    "## File of gene set. \n",
    "GENES_ADD = paths_dict['DATABASES'][\"GENE_LISTS\"][\"GENES_ADD\"]\n",
    "\n",
    "## Output folder where we place the gene extracts for Fucus\n",
    "WORKING_DATASET = paths_dict['ANALYSIS_DATA']['SINGLE_GENE_ANALYSIS']['DATASET']['REF']['WORKING']['REF']"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "%%bash -s \"$REF_PROTEOMES\" \"$GENES_ADD\" \"$WORKING_DATASET\"\n",
    "\n",
    "cat \"$2\" | while read line; do\n",
    "    echo $line\n",
    "    seqkit grep -rp \"gene=$line;\" \"$1\"/NC_016735.taxo.fasta > \"$3\"/\"$line\".fasta\n",
    "done"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### 2.1.2. Create a blast database for all the other reference proteomes\n",
    "\n",
    "Create blast databases for the other 178 reference taxa. "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {},
   "outputs": [],
   "source": [
    "## Folder containing proteomes of reference taxa\n",
    "REF_PROTEOMES = paths_dict['ANALYSIS_DATA']['SINGLE_GENE_ANALYSIS']['PROTEOMES']['REF_PROTEOMES']\n",
    "\n",
    "## Output folder that will contain the blast dbs of the remaining 179 taxa\n",
    "REFS_TO_ADD = paths_dict['ANALYSIS_DATA']['SINGLE_GENE_ANALYSIS']['DATASET']['REF']['WORKING']['TOADD']"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "%%bash -s \"$REF_PROTEOMES\" \"$REFS_TO_ADD\"\n",
    "\n",
    "## load the blast module on the Uppmax cluster\n",
    "module load bioinfo-tools\n",
    "module load blast\n",
    "\n",
    "for i in \"$1\"/*taxo.fasta; do\n",
    "    if [ \"$i\" != \"NC_016735.taxo.fasta\" ]; then\n",
    "        taxon=$(basename $i | cut -f1 -d '.')\n",
    "        makeblastdb -in $i -dbtype prot -out \"$2\"/\"$taxon\".db\n",
    "    fi\n",
    "done"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### 2.1.3. Blast extracted _Fucus_ genes against the blast databases\n",
    "\n",
    "We first create a text file called `blastdb_name.txt` with the names of the blast databases."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "metadata": {},
   "outputs": [],
   "source": [
    "%%bash -s \"$REFS_TO_ADD\"\n",
    "\n",
    "for i in \"$1\"/*; do\n",
    "    echo $i | \\\n",
    "    cut -f 1 -d '.' | \\\n",
    "    sed -E 's/(.*)/\\1\\.db/'\n",
    "done | uniq > \"$1\"/blastdb_name.txt"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "metadata": {},
   "outputs": [],
   "source": [
    "## Folder containing the Fucus fasta files that we will use as blast query\n",
    "WORKING_DATASET = paths_dict['ANALYSIS_DATA']['SINGLE_GENE_ANALYSIS']['DATASET']['REF']['WORKING']['REF']\n",
    "\n",
    "## Folder containing the blast dbs\n",
    "REFS_TO_ADD = paths_dict['ANALYSIS_DATA']['SINGLE_GENE_ANALYSIS']['DATASET']['REF']['WORKING']['TOADD']"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "We submit the blast jobs now!"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "%%bash -s \"$WORKING_DATASET\" \"$REFS_TO_ADD\" \n",
    "\n",
    "for i in  \"$1\"/*.fasta; do\n",
    "    sbatch ../../uppmax_scripts/script_bin/job_running_blast.sh \\\n",
    "        $i \"$2\"/blastdb_name.txt\n",
    "    sleep 1\n",
    "done"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### 2.1.4. Extract best blast hits\n",
    "\n",
    "Blast output files are generated in the folder containing the single genes from _Fucus_, our query. For each gene, we want to:\n",
    "- parse the blast output files, \n",
    "- extract the best hit,\n",
    "- pull out the corresponding sequence"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "metadata": {},
   "outputs": [],
   "source": [
    "## Full gene dataset\n",
    "JANO_FULL = paths_dict['DATABASES'][\"GENE_LISTS\"][\"JANO_FULL\"]\n",
    "\n",
    "## 28 gene dataset\n",
    "GENES_ADD = paths_dict['DATABASES'][\"GENE_LISTS\"][\"GENES_ADD\"]\n",
    "\n",
    "## Folder containing the Fucus fasta files and the blast outputs\n",
    "WORKING_DATASET = paths_dict['ANALYSIS_DATA']['SINGLE_GENE_ANALYSIS']['DATASET']['REF']['WORKING']['REF']\n",
    "\n",
    "## Folder containing proteomes of reference taxa\n",
    "REF_PROTEOMES = paths_dict['ANALYSIS_DATA']['SINGLE_GENE_ANALYSIS']['PROTEOMES']['REF_PROTEOMES']\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "We first pull the best blast hit. We can do that by taking the first line of the blast output (since the blast output is already sorted). The second column of the blast tabular output contains the name of the best sequence."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "%%bash -s \"$JANO_FULL\" \"$WORKING_DATASET\"\n",
    "\n",
    "## Extract best blast hit for each gene and taxon\n",
    "cut -f 1 $1 | grep -v \"gene\" | sort -u | while read line; do\n",
    "    for i in \"$2\"/\"$line\".fasta__*; do\n",
    "        cat $i | head -n 1 | cut -f2 >> \"$2\"/\"$line\"_toAdd.list\n",
    "    done\n",
    "done"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "%%bash -s \"$GENES_ADD\" \"$WORKING_DATASET\"\n",
    "\n",
    "## Extract best blast hit for each gene and taxon\n",
    "cat $1 | while read line; do\n",
    "    for i in \"$2\"/\"$line\".fasta__*; do\n",
    "        cat $i | head -n 1 | cut -f2 >> \"$2\"/\"$line\"_toAdd.list\n",
    "    done\n",
    "done"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "We pull the corresponding sequences now. \n",
    "\n",
    "First we concatenate all the proteomes together to make it easier to search the sequences."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "metadata": {},
   "outputs": [],
   "source": [
    "%%bash -s \"$REF_PROTEOMES\"\n",
    "cat \"$1\"/*.taxo.fasta > \"$1\"/all.fasta"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Next we use [seqkit](https://github.com/shenwei356/seqkit) to pull out the sequences for each gene."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "%%bash -s \"$WORKING_DATASET\" \"$REF_PROTEOMES\"\n",
    "\n",
    "for i in \"$1\"/*_toAdd.list; do\n",
    "    gene=$(basename $i | cut -f 1 -d '.')\n",
    "    seqkit grep -f $i \"$2\"/all.fasta > \"$1\"/\"$gene\".fasta\n",
    "done"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Put together the query and the extracted sequences. "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {},
   "outputs": [],
   "source": [
    "%%bash -s \"$JANO_FULL\" \"$WORKING_DATASET\"\n",
    "\n",
    "cut -f 1 $1 | grep -v \"gene\" | sort -u | while read line; do\n",
    "    cat \"$2\"/\"$line\".fasta \"$2\"/\"$line\"_toAdd.fasta > \"$2\"/\"$line\".all.fasta\n",
    "done"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "And the additional 28 genes too!"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "%%bash -s \"$GENES_ADD\" \"$WORKING_DATASET\"\n",
    "\n",
    "cat $1 | while read line; do\n",
    "    cat \"$2\"/\"$line\".fasta \"$2\"/\"$line\"_toAdd.fasta > \"$2\"/\"$line\".all.fasta\n",
    "done"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "At this stage, we may have duplicates, particularly in the cyanobacteria. To avoid problems downstream, we can rename duplicated sequences. We can use seqkit for this, which appends '_N' to duplicated sequence IDs to make them unique."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 19,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Folder containing output fasta files\n",
    "DATASET = paths_dict['ANALYSIS_DATA']['SINGLE_GENE_ANALYSIS']['DATASET']['REF']['WORKING']['ROOT']"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 20,
   "metadata": {},
   "outputs": [],
   "source": [
    "%%bash -s \"$WORKING_DATASET\" \"$DATASET\"\n",
    "\n",
    "for i in \"$1\"/*all.fasta; do\n",
    "    gene=$(basename $i | cut -f 1 -d '.')\n",
    "    cat $i | \\\n",
    "    seqkit rename > \"$2\"/\"$gene\".fasta\n",
    "done"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Before running the trees, we probably need to format the fasta headers so that RAxML (or any tree inference programme) is happy. That means replacing ';' and '.' with '-'. \n",
    "\n",
    "We also edited our fasta headers so they would work with PhyloFisher (we ended up not using the Phylofisher workflow, as it removes all hits to bacterial orthogroups when collecting homologs via the fisher algorithm, which might not be appropriate for plastid genes). Since PhylloFisher expects the fasta headers to be identical across all genes for each taxon, we'll need to drop the gene name from the fasta header. I do somehow want to keep the protein ID or contig, so we can have that at the end of the sequence after an underscore (as PhyloFisher seems to ignore everything after the underscore)."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 22,
   "metadata": {},
   "outputs": [],
   "source": [
    "%%bash -s \"$DATASET\"\n",
    "\n",
    "for i in \"$1\"/*fasta; do\n",
    "    cat $i | sed -E 's/gene=.*;accession/accession/' | \\\n",
    "    tr '_' '-' | \\\n",
    "    sed -E 's/;protein/_protein/' | \\\n",
    "    sed -E 's/;contig/_contig/' | \\\n",
    "    tr ';' '-' | tr '.' '-' \\\n",
    "    > txt\n",
    "    mv txt $i\n",
    "done"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### 2.2 Align, trim, and infer single gene trees"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "For our preliminary reference SGTs, we align the genes with mafft --auto. "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 23,
   "metadata": {},
   "outputs": [],
   "source": [
    "from gene_iterator import GeneIterator"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 24,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Folder with extracted gene dataset\n",
    "DATASET = paths_dict['ANALYSIS_DATA']['SINGLE_GENE_ANALYSIS']['DATASET']['REF']['WORKING']['ROOT']\n",
    "\n",
    "# Read_genes\n",
    "JANO_FULL = paths_dict['DATABASES']['GENE_LISTS']['JANO_FULL']\n",
    "genes = set(line.split()[0].strip() for line in open(JANO_FULL, \"r\"))\n",
    "\n",
    "# Read genes\n",
    "GENES_ADD = paths_dict['DATABASES'][\"GENE_LISTS\"][\"GENES_ADD\"]\n",
    "genes_ADD = set(line.split()[0].strip() for line in open(GENES_ADD, \"r\"))\n",
    "\n",
    "# Directory for mafft output\n",
    "MAFFT_DIR = paths_dict['ANALYSIS_DATA']['SINGLE_GENE_ANALYSIS']['ALIGNMENTS']['PRELIM_REF']['MAFFT']\n",
    "\n",
    "# Slurmlog csv\n",
    "SLURMLOG = paths_dict['ANALYSIS_DATA']['SINGLE_GENE_ANALYSIS']['ALIGNMENTS']['PRELIM_REF']['MAFFTLOG']\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "gi = GeneIterator(DATASET, gene_list=genes, suffix=\"fasta\")\n",
    "gi.unlock_pipeline()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "gi = GeneIterator(DATASET, gene_list=genes_ADD, suffix=\"fasta\")\n",
    "gi.unlock_pipeline()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "gi.run_mafft(MAFFT_DIR, SLURMLOG)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "I quickly scanned a few of the alignments and then trimmed with trimal, using a gap thrashold of 0.8 (aggressive trimming)."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 28,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Directory for mafft files\n",
    "MAFFT_DIR = paths_dict['ANALYSIS_DATA']['SINGLE_GENE_ANALYSIS']['ALIGNMENTS']['PRELIM_REF']['MAFFT']\n",
    "\n",
    "# Read_genes\n",
    "JANO_FULL = paths_dict['DATABASES']['GENE_LISTS']['JANO_FULL']\n",
    "genes = set(line.split()[0].strip() for line in open(JANO_FULL, \"r\"))\n",
    "\n",
    "# Read genes\n",
    "GENES_ADD = paths_dict['DATABASES'][\"GENE_LISTS\"][\"GENES_ADD\"]\n",
    "genes_add = set(line.split()[0].strip() for line in open(GENES_ADD, \"r\"))\n",
    "\n",
    "# Directory for trimal output\n",
    "TRIMAL_DIR = paths_dict['ANALYSIS_DATA']['SINGLE_GENE_ANALYSIS']['ALIGNMENTS']['PRELIM_REF']['TRIMAL']\n",
    "\n",
    "# Slurmlog csv\n",
    "SLURMLOG = paths_dict['ANALYSIS_DATA']['SINGLE_GENE_ANALYSIS']['ALIGNMENTS']['PRELIM_REF']['TRIMALLOG']\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "gi = GeneIterator(MAFFT_DIR, gene_list=genes, suffix=\"fasta\")\n",
    "gi.unlock_pipeline()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "gi = GeneIterator(MAFFT_DIR, gene_list=genes_add, suffix=\"fasta\")\n",
    "gi.unlock_pipeline()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "gi.run_trimal(TRIMAL_DIR, SLURMLOG, MAFFT_DIR)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Now we are ready to run trees! For these preliminary trees, we'll use raxml-ng v1.2 with LG4X model and 100 rapid bootstrap searches."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 31,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Directory for raxml trees\n",
    "TREES = paths_dict['ANALYSIS_DATA']['SINGLE_GENE_ANALYSIS']['TREES']['PRELIM_REF']"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "We manually created soft links for the processed fasta files (from TRIMAL_DIR) in the TREES directory.\n",
    "\n",
    "We can now submit jobs to run the trees."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "%%bash -s \"$TREES\" \n",
    "\n",
    "for i in \"$1\"/*fasta; do\n",
    "    gene=$(basename $i | cut -f 1 -d '.')\n",
    "    echo $gene\n",
    "    sbatch /crex/proj/naiss2023-6-81/Mahwash/beta-Cyclocitral/uppmax_scripts/script_bin/job_raxml.sh \"$i\" \"$gene\"\n",
    "    sleep 1\n",
    "done "
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### 2.3 Manually go through the single gene trees\n",
    "\n",
    "We manually checked the single gene trees. Once checked, each tree was saved as `{file name}.checked`.\n",
    "\n",
    " Tips were marked in red (#ff0000) if they were duplicates (so that each taxon should have only one ortholog), or if they clearly belonged to another gene (super long branches). \n",
    "\n",
    "We marked paralogs in blue (#0000ff). As these are plastid genes, there werent really any paralogs in the genes we looked at, but for the clpC gene, one distinct clade was made of clpB sequences which was marked in blue. This will help distinguish between the clpC and clpB gene when we extract sequences from the MAGs. "
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## 3. Get clean reference sequences for each gene"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### 3.1 Remove contaminants\n",
    "\n",
    "To get clean sequences for each gene, we remove sequences marked in red or blue in the previous step. "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Gene list\n",
    "JANO_FULL = paths_dict['DATABASES']['GENE_LISTS']['JANO_FULL']\n",
    "\n",
    "# Read genes\n",
    "GENES_ADD = paths_dict['DATABASES'][\"GENE_LISTS\"][\"GENES_ADD\"]\n",
    "\n",
    "# Directory for raxml trees\n",
    "TREES = paths_dict['ANALYSIS_DATA']['SINGLE_GENE_ANALYSIS']['TREES']['PRELIM_REF']\n",
    "\n",
    "# Directory for input fasta files (unaligned and untrimmed)\n",
    "DATASET = paths_dict['ANALYSIS_DATA']['SINGLE_GENE_ANALYSIS']['DATASET']['REF']['WORKING']['ROOT']\n",
    "\n",
    "# Out directory\n",
    "FINAL = paths_dict['ANALYSIS_DATA']['SINGLE_GENE_ANALYSIS']['DATASET']['REF']['FINAL']"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "%%bash -s \"$JANO_FULL\" \"$TREES\" \"$DATASET\" \"$FINAL\"\n",
    "\n",
    "## Get gene list\n",
    "genes=$(cut -f 1 $1 | grepp -v \"gene\" | sort -u | tr '\\n' ' ')\n",
    "\n",
    "## Loop over gene list, get list of marked sequences from trees, and remove the marked sequences \n",
    "## from the corresponding fasta files to get a clean fasta file for each gene\n",
    "for gene in $genes\n",
    "do\n",
    "    echo $gene \n",
    "    grep \"color=\" \"$2\"/RAxML_bipartitions.\"$gene\".selected | \\\n",
    "    sed -E 's/\\s+(.*)\\[.*\\]/\\1/' | \\\n",
    "    tr -d \"'\" \\\n",
    "    > \"$2\"/marked.list\n",
    "    seqkit grep -f \"$2\"/marked.list \"$3\"/\"$gene\".fasta -v | \\\n",
    "    sed -E 's/(>.*)_protein.*/\\1/' | \\\n",
    "    sed -E 's/(>.*)_contig.*/\\1/' \\\n",
    "    > \"$4\"/\"$gene\".fas\n",
    "done"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "%%bash -s \"$GENES_ADD\" \"$TREES\" \"$DATASET\" \"$FINAL\"\n",
    "\n",
    "## Get gene list\n",
    "genes=$(cat $1 | tr '\\n' ' ')\n",
    "\n",
    "## Loop over gene list, get list of marked sequences from trees, and remove the marked sequences \n",
    "## from the corresponding fasta files to get a clean fasta file for each gene\n",
    "for gene in $genes\n",
    "do\n",
    "    echo $gene \n",
    "    grep \"color=\" \"$2\"/\"$gene\".raxml.support.checked | \\\n",
    "    sed -E 's/\\s+(.*)\\[.*\\]/\\1/' | \\\n",
    "    tr -d \"'\" \\\n",
    "    > \"$2\"/marked.list\n",
    "    seqkit grep -f \"$2\"/marked.list \"$3\"/\"$gene\".fasta -v | \\\n",
    "    sed -E 's/(>.*)_protein.*/\\1/' | \\\n",
    "    sed -E 's/(>.*)_contig.*/\\1/' \\\n",
    "    > \"$4\"/\"$gene\".fas\n",
    "done"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "At this I noticed one of the taxa had a `/` in the header. To avoid confusion. I changed the header as follows:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [],
   "source": [
    "%%bash -s \"$FINAL\"\n",
    "\n",
    "for i in \"$1\"/*fas\n",
    "do \n",
    "    cat $i | \\\n",
    "    sed -E 's/taxo=Eukaryota-Sar-Stramenopiles-Ochrophyta-Eustigmatophyceae-sp-Chic-10\\/23-P-6w-accession=NC-040296/taxo=Eukaryota-Sar-Stramenopiles-Ochrophyta-Eustigmatophyceae-sp-accession=NC-040296/' \\\n",
    "    > fasta\n",
    "    mv fasta \"$i\"ta\n",
    "done"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "How many sequences do have for each gene?"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "%%bash -s \"$FINAL\"\n",
    "\n",
    "seqkit stats \"$1\"/*fasta"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## References\n",
    "\n",
    "Janouškovec, J., Horák, A., Oborník, M., Lukeš, J., & Keeling, P. J. (2010). A common red algal origin of the apicomplexan, dinoflagellate, and heterokont plastids. Proceedings of the National Academy of Sciences, 107(24), 10949-10954. https://doi.org/10.1073/pnas.1003335107\n",
    "\n",
    "Le Corguillé, G., Pearson, G., Valente, M., Viegas, C., Gschloessl, B., Corre, E., ... & Leblanc, C. (2009). Plastid genomes of two brown algae, Ectocarpus siliculosus and Fucus vesiculosus: further insights on the evolution of red-algal derived plastids. BMC Evolutionary Biology, 9(1), 1-14. https://doi.org/10.1186/1471-2148-9-253\n",
    "\n",
    "Shen, W., Le, S., Li, Y., & Hu, F. (2016). SeqKit: a cross-platform and ultrafast toolkit for FASTA/Q file manipulation. PloS one, 11(10), e0163962. https://doi.org/10.1371/journal.pone.0163962\n",
    "\n",
    "Rambaut, A. (2009). FigTree. Tree figure drawing tool. http://tree.bio.ed.ac.uk/software/figtree/.\n",
    "\n",
    "Tice, A. K., Žihala, D., Pánek, T., Jones, R. E., Salomaki, E. D., Nenarokov, S., ... & Brown, M. W. (2021). PhyloFisher: a phylogenomic package for resolving eukaryotic relationships. PLoS Biology, 19(8), e3001365. https://doi.org/10.1371/journal.pbio.3001365\n",
    "\n",
    "Pietluch, F., Mackiewicz, P., Ludwig, K., & Gagat, P. (2024). A New Model and Dating for the Evolution of Complex Plastids of Red Alga Origin. Genome Biology and Evolution, evae192. https://doi.org/10.1093/gbe/evae192\n",
    "\n",
    "Ponce-Toledo, R. I., Deschamps, P., López-García, P., Zivanovic, Y., Benzerara, K., & Moreira, D. (2017). An early-branching freshwater cyanobacterium at the origin of plastids. Current Biology, 27(3), 386-391. https://doi.org/10.1016/j.cub.2016.11.056 \n",
    "\n"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "ptMAGs",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.10.5"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
