{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Concatenated phylogenies\n",
    "\n",
    "In this notebook, we infer the large concatenated phylogeny (Figure 1 in teh manuscript). Preliminary phylogenies were based on a reduced gene set (not shown/documented here for the sake of brevity)."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Check if python is 3.10.5\n",
    "import json\n",
    "import os\n",
    "import pandas as pd\n",
    "import sys\n",
    "import numpy as np\n",
    "import __init__\n",
    "\n",
    "print(sys.version)\n",
    "%load_ext autoreload\n",
    "%autoreload 2"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [],
   "source": [
    "# we store the important data paths in PATH_FILE\n",
    "PATH_FILE = \"../../PATHS.json\"\n",
    "\n",
    "paths_dict = json.load(open(PATH_FILE, \"r\"));"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## 1. Full tree - 93 genes\n",
    "\n",
    "### 1.1. Assembled dataset\n",
    "**Genes**  \n",
    "We have the clean dataset for 96 genes, however, we will exclude ascF (too few taxa), psbH (slightly weird gene tree), and rbcL (different origins of the gene in red and green algae) leaving us with 93 genes.  \n",
    "\n",
    "**Taxa**  \n",
    "We exclude 7 ptMAGs that had high redundancy values. We also exclude two redundant refs and add one ref from prelimnary versions (v1 of the preprint on bioRxiv).   \n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### 1.2 Some stats\n",
    "\n",
    "We calculate some stats regarding gene and taxa occupancy. "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [],
   "source": [
    "## Dataset\n",
    "DATASET = paths_dict['ANALYSIS_DATA'][\"CONCAT_GENE_ANALYSIS\"][\"FULL\"][\"DATASET\"][\"V3\"]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "%%bash -s \"$DATASET\"\n",
    "\n",
    "/home/mahja/ptMAGs/src/get_stats.sh \"$1\""
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### 3.3 Filter with prequal\n",
    "\n",
    "In one case, we filter sequence stretches with no clear homology. We use a posterior probability threshold of 0.95. "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Folder with extracted gene dataset\n",
    "DATASET = paths_dict['ANALYSIS_DATA'][\"CONCAT_GENE_ANALYSIS\"][\"FULL\"][\"DATASET\"][\"V3\"]\n",
    "\n",
    "# Output folder for prequal files\n",
    "PREQUAL = paths_dict['ANALYSIS_DATA'][\"CONCAT_GENE_ANALYSIS\"][\"FULL\"][\"ALIGNMENTS\"][\"V3\"][\"PREQUAL\"]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "%%bash -s \"$DATASET\" \"$PREQUAL\"\n",
    "\n",
    "for i in \"$1\"/*fasta\n",
    "do \n",
    "    sbatch ../../uppmax_scripts/script_bin/job_prequal.sh $i $2\n",
    "    sleep 1\n",
    "done"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### 3.3 Align\n",
    "We align with mafft-ginsi. "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [],
   "source": [
    "from gene_iterator import GeneIterator"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Folder with prequal-filtered gene dataset\n",
    "PREQUAL = paths_dict['ANALYSIS_DATA'][\"CONCAT_GENE_ANALYSIS\"][\"FULL\"][\"ALIGNMENTS\"][\"V3\"][\"PREQUAL\"]\n",
    "\n",
    "# Read_genes\n",
    "GENE_LIST = paths_dict['DATABASES']['GENE_LISTS']['GENES_93']\n",
    "genes = list(map(lambda x : x.strip(), open(GENE_LIST, \"r\").readlines()))\n",
    "\n",
    "# Directory for mafft output\n",
    "MAFFT_DIR = paths_dict['ANALYSIS_DATA'][\"CONCAT_GENE_ANALYSIS\"][\"FULL\"][\"ALIGNMENTS\"][\"V3\"][\"MAFFT\"]\n",
    "\n",
    "# Slurmlog csv\n",
    "SLURMLOG = paths_dict['ANALYSIS_DATA'][\"CONCAT_GENE_ANALYSIS\"][\"FULL\"][\"ALIGNMENTS\"][\"V3\"][\"MAFFTLOG\"]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "gi = GeneIterator(PREQUAL, gene_list=genes, suffix=\"fasta\")\n",
    "gi.unlock_pipeline()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "gi.run_mafft(MAFFT_DIR, SLURMLOG)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### 3.4 Trim\n",
    "We trim the alignments with BMGE (BLOSUM35 matrix, filter columns > 80% gaps)."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Folder containing aligned fasta files\n",
    "MAFFT_DIR = paths_dict['ANALYSIS_DATA'][\"CONCAT_GENE_ANALYSIS\"][\"FULL\"][\"ALIGNMENTS\"][\"V3\"][\"MAFFT\"]\n",
    "\n",
    "# Read_genes\n",
    "GENE_LIST = paths_dict['DATABASES']['GENE_LISTS']['GENES_93']\n",
    "genes = list(map(lambda x : x.strip(), open(GENE_LIST, \"r\").readlines()))\n",
    "\n",
    "# Directory for BMGE output\n",
    "BMGE_DIR = paths_dict['ANALYSIS_DATA'][\"CONCAT_GENE_ANALYSIS\"][\"FULL\"][\"ALIGNMENTS\"][\"V3\"][\"BMGE\"]\n",
    "\n",
    "# Slurmlog csv\n",
    "SLURMLOG = paths_dict['ANALYSIS_DATA'][\"CONCAT_GENE_ANALYSIS\"][\"FULL\"][\"ALIGNMENTS\"][\"V3\"][\"BMGELOG\"]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "gi = GeneIterator(MAFFT_DIR, gene_list=genes, suffix=\"fasta\")\n",
    "gi.unlock_pipeline()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "gi.run_bmge(BMGE_DIR, SLURMLOG, MAFFT_DIR)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### 3.5 Concatenate"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Directory for aligned and trimmed fasta files\n",
    "BMGE_DIR = paths_dict[\"ANALYSIS_DATA\"][\"CONCAT_GENE_ANALYSIS\"][\"FULL\"][\"ALIGNMENTS\"][\"V3\"][\"BMGE\"]\n",
    "\n",
    "## Output directory for fasta files\n",
    "CONCAT_DIR = paths_dict[\"ANALYSIS_DATA\"][\"CONCAT_GENE_ANALYSIS\"][\"FULL\"][\"ALIGNMENTS\"][\"V3\"][\"CONCAT\"]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "%%bash -s \"$BMGE_DIR\" \"$CONCAT_DIR\"\n",
    "\n",
    "files=(\"$1\"/*fasta)\n",
    "\n",
    "perl /home/mahja/ptMAGs/src/cat_fasta.pl -f \"${files[@]}\" > \"$2\"/concat_839t_93g_prequal_ginsi_bmge.fasta\n",
    "mv partitions.txt \"$2\"/partitions_prequal_ginsi_bmge.txt"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "%%bash -s \"$CONCAT_DIR\"\n",
    "\n",
    "seqkit stats \"$1\"/concat_839t_93g_prequal_ginsi_bmge.fasta"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "The concatenated file has 839 sequences, and 19,242 aligned sites. "
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### 3.6 Infer tree\n",
    "\n",
    "We run the tree in IQTree, using the best fitting site-homogenous model (matrix LG or cpREV). We treat the concatenated alignment as one big partition. We infer support using 1000 ultrafast bootstraps."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "metadata": {},
   "outputs": [],
   "source": [
    "## Directory for fasta file\n",
    "CONCAT_DIR = paths_dict[\"ANALYSIS_DATA\"][\"CONCAT_GENE_ANALYSIS\"][\"FULL\"][\"ALIGNMENTS\"][\"V3\"][\"CONCAT\"]\n",
    "\n",
    "## Output directory for tree\n",
    "TREE_DIR = paths_dict[\"ANALYSIS_DATA\"][\"CONCAT_GENE_ANALYSIS\"][\"FULL\"][\"TREES\"][\"V3\"]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Submitted batch job 55859943 on cluster rackham\n"
     ]
    }
   ],
   "source": [
    "%%bash -s \"$CONCAT_DIR\" \"$TREE_DIR\"\n",
    "\n",
    "sbatch ../../uppmax_scripts/script_bin/job_iqtree.sh \"$1\"/concat_839t_93g_prequal_ginsi_bmge.fasta \"$2\"/concat_839t_93g_prequal_ginsi_bmge_ML"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## References\n",
    "\n",
    "Katoh, K., & Standley, D. M. (2013). MAFFT multiple sequence alignment software version 7: improvements in performance and usability. Molecular biology and evolution, 30(4), 772-780. https://doi.org/10.1093/molbev/mst010\n",
    "\n",
    "Capella-Gutiérrez, S., Silla-Martínez, J. M., & Gabaldón, T. (2009). trimAl: a tool for automated alignment trimming in large-scale phylogenetic analyses. Bioinformatics, 25(15), 1972-1973. https://doi.org/10.1093/bioinformatics/btp348\n",
    "\n",
    "cat_fasta.pl script. Written by Martin Ryberg. Available at https://github.com/mr-y/my_bioinfPerls/blob/master/cat_fasta.pl. "
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.10.5"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
