{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Subset phylogeny with more complex models\n",
    "\n",
    "To illucidate the position of leptophytes (also referred to as NEW, back when we did not have a name for the lineage), we selected a subset of MAGs and references that were more complete (107 taxa in total) so that we could use more complex mixture models on our dataset of 93 genes. "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "3.10.5 | packaged by conda-forge | (main, Jun 14 2022, 07:04:59) [GCC 10.3.0]\n"
     ]
    }
   ],
   "source": [
    "# Check if python is 3.10.5\n",
    "import json\n",
    "import os\n",
    "import pandas as pd\n",
    "import sys\n",
    "import numpy as np\n",
    "import __init__\n",
    "\n",
    "\n",
    "print(sys.version)\n",
    "%load_ext autoreload\n",
    "%autoreload 2"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [],
   "source": [
    "# we store the important data paths in PATH_FILE\n",
    "PATH_FILE = \"../../PATHS.json\"\n",
    "\n",
    "paths_dict = json.load(open(PATH_FILE, \"r\"));"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## 1. Get dataset\n",
    "\n",
    "Our taxon selection included mostly red plastids, using glaucophytes and viridiplantae as outgroups. We chose to not include cyanobacteria in an effort to reduce compositional heterogeneity (plastid genomes tend to be greatly AT rich compared to cyanobacteria). \n",
    "\n",
    "Here, we generate our dataset as a subset of the V2 concatenated phylogeny. "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {},
   "outputs": [],
   "source": [
    "## V2 full dataset\n",
    "DATASET_FULL = paths_dict['ANALYSIS_DATA'][\"CONCAT_GENE_ANALYSIS\"][\"FULL\"][\"DATASET\"][\"V2\"]\n",
    "\n",
    "# Read_genes\n",
    "GENE_LIST = paths_dict['DATABASES']['GENE_LISTS']['GENES_93']\n",
    "\n",
    "## Output folder for V14 subset dataset\n",
    "DATASET = paths_dict['ANALYSIS_DATA'][\"CONCAT_GENE_ANALYSIS\"][\"SUBSET\"][\"DATASET\"][\"V14\"]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "%%bash -s \"$GENE_LIST\" \"$DATASET_FULL\" \"$DATASET\"\n",
    "\n",
    "cat $1 | while read gene\n",
    "do\n",
    "    seqkit grep -f \"$3/extract.txt\" \"$2\"/\"$gene\".fasta > \"$3\"/\"$gene\".fasta\n",
    "done"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Get stats!"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "%%bash -s \"$DATASET\"\n",
    "\n",
    "/home/mahja/ptMAGs/src/get_stats.sh \"$1\""
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## 2. Run prequal \n",
    "\n",
    "We remove sequence stretches with no clear homology using Prequal. We used a posterior probability threshold of 0.95."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Folder with extracted gene dataset\n",
    "DATASET = paths_dict['ANALYSIS_DATA'][\"CONCAT_GENE_ANALYSIS\"][\"SUBSET\"][\"DATASET\"][\"V14\"]\n",
    "\n",
    "# Output folder for prequal files\n",
    "PREQUAL = paths_dict['ANALYSIS_DATA'][\"CONCAT_GENE_ANALYSIS\"][\"SUBSET\"][\"ALIGNMENTS\"][\"V14\"][\"PREQUAL\"]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "%%bash -s \"$DATASET\" \"$PREQUAL\"\n",
    "\n",
    "for i in \"$1\"/*fasta\n",
    "do \n",
    "    sbatch ../../uppmax_scripts/script_bin/job_prequal.sh $i $2\n",
    "    sleep 1\n",
    "done"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## 3. Align\n",
    "\n",
    "We align the genes with mafft-ginsi using the --unalign 0.6 option to avoid over-alignment."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [],
   "source": [
    "from gene_iterator import GeneIterator"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Folder with prequal-filtered gene dataset\n",
    "PREQUAL = paths_dict['ANALYSIS_DATA'][\"CONCAT_GENE_ANALYSIS\"][\"SUBSET\"][\"ALIGNMENTS\"][\"V14\"][\"PREQUAL\"]\n",
    "\n",
    "# Read_genes\n",
    "GENE_LIST = paths_dict['DATABASES']['GENE_LISTS']['GENES_93']\n",
    "genes = list(map(lambda x : x.strip(), open(GENE_LIST, \"r\").readlines()))\n",
    "\n",
    "# Directory for mafft output\n",
    "MAFFT_DIR = paths_dict['ANALYSIS_DATA'][\"CONCAT_GENE_ANALYSIS\"][\"SUBSET\"][\"ALIGNMENTS\"][\"V14\"][\"MAFFT\"]\n",
    "\n",
    "# Slurmlog csv\n",
    "SLURMLOG = paths_dict['ANALYSIS_DATA'][\"CONCAT_GENE_ANALYSIS\"][\"SUBSET\"][\"ALIGNMENTS\"][\"V14\"][\"MAFFTLOG\"]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "gi = GeneIterator(PREQUAL, gene_list=genes, suffix=\"fasta\")\n",
    "gi.unlock_pipeline()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "gi.run_mafft(MAFFT_DIR, SLURMLOG)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## 4. Trim\n",
    "\n",
    "We trim the alignments with BMGE (BLOSUM35 matrix, filter columns > 80% gaps)."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Folder containing aligned fasta files\n",
    "MAFFT_DIR = paths_dict['ANALYSIS_DATA'][\"CONCAT_GENE_ANALYSIS\"][\"SUBSET\"][\"ALIGNMENTS\"][\"V14\"][\"MAFFT\"]\n",
    "\n",
    "# Read_genes\n",
    "GENE_LIST = paths_dict['DATABASES']['GENE_LISTS']['GENES_93']\n",
    "genes = list(map(lambda x : x.strip(), open(GENE_LIST, \"r\").readlines()))\n",
    "\n",
    "# Directory for trimal output\n",
    "BMGE_DIR = paths_dict['ANALYSIS_DATA'][\"CONCAT_GENE_ANALYSIS\"][\"SUBSET\"][\"ALIGNMENTS\"][\"V14\"][\"BMGE\"]\n",
    "\n",
    "# Slurmlog csv\n",
    "SLURMLOG = paths_dict['ANALYSIS_DATA'][\"CONCAT_GENE_ANALYSIS\"][\"SUBSET\"][\"ALIGNMENTS\"][\"V14\"][\"BMGELOG\"]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "gi = GeneIterator(MAFFT_DIR, gene_list=genes, suffix=\"fasta\")\n",
    "gi.unlock_pipeline()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "gi.run_bmge(BMGE_DIR, SLURMLOG, MAFFT_DIR)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## 5. Concatenate"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Directory for aligned and trimmed fasta files\n",
    "BMGE_DIR = paths_dict['ANALYSIS_DATA'][\"CONCAT_GENE_ANALYSIS\"][\"SUBSET\"][\"ALIGNMENTS\"][\"V14\"][\"BMGE\"]\n",
    "\n",
    "## Output directory for fasta files\n",
    "CONCAT_DIR = paths_dict[\"ANALYSIS_DATA\"][\"CONCAT_GENE_ANALYSIS\"][\"SUBSET\"][\"ALIGNMENTS\"][\"V14\"][\"CONCAT\"]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "%%bash -s \"$BMGE_DIR\" \"$CONCAT_DIR\"\n",
    "\n",
    "files=(\"$1\"/*fasta)\n",
    "\n",
    "perl /home/mahja/ptMAGs/src/cat_fasta.pl -f \"${files[@]}\" > \"$2\"/concat_107t_93g_prequal_ginsi_bmge.fasta\n",
    "mv partitions.txt \"$2\"/partitions_prequal_ginsi_bmge.txt"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "I manually edited the header the names slightly. \n",
    "- Removed the * after New\n",
    "- Changed ZHAN22 to Haptophyte from New\n",
    "\n",
    "The alignment has 107 taxa, 93 genes, and 20,292 positions. "
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## 6. Run cpREV+C60+G tree\n",
    "\n",
    "We started by running a tree with the cpREV+C60+G model (the best fit model as assessed by model fit tests on preliminary phylogenies). This phylogeny will be used as the guide tree for calculating the parameters of the MEOW models. "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {},
   "outputs": [],
   "source": [
    "## Directory for fasta file\n",
    "CONCAT_DIR = paths_dict[\"ANALYSIS_DATA\"][\"CONCAT_GENE_ANALYSIS\"][\"SUBSET\"][\"ALIGNMENTS\"][\"V14\"][\"CONCAT\"]\n",
    "\n",
    "## Output directory for tree\n",
    "TREE_DIR = paths_dict[\"ANALYSIS_DATA\"][\"CONCAT_GENE_ANALYSIS\"][\"SUBSET\"][\"TREES\"][\"V14\"]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "%%bash -s \"$CONCAT_DIR\" \"$TREE_DIR\"\n",
    "\n",
    "sbatch ../../uppmax_scripts/script_bin/job_iqtree.sh \"$1\"/concat_107t_93g_prequal_ginsi_bmge.fasta \"$2\"/concat_107t_93g_prequal_ginsi_bmge_cpREV-C60-G"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "The resulting tree recovered leptophytes as sister to cryptophytes and haptophytes (62% ufb support for the monophyly of cryptophytes and haptophytes). It also recovered the complex plastids as monophyletic (88% ufb support)."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## 7. Run CAT-GTR tree\n",
    "\n",
    "Three chains of PhyloBayes were started on the lab cluster. We convert the fasta file to a phylip file first."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "metadata": {},
   "outputs": [],
   "source": [
    "## Directory for fasta file\n",
    "CONCAT_DIR = paths_dict[\"ANALYSIS_DATA\"][\"CONCAT_GENE_ANALYSIS\"][\"SUBSET\"][\"ALIGNMENTS\"][\"V14\"][\"CONCAT\"]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "%%bash -s \"$CONCAT_DIR\"\n",
    "\n",
    "perl ../../src/fasta2phylip.pl -f \"$1\"/concat_107t_93g_prequal_ginsi_bmge.fasta -o \"$1\"/concat_107t_93g_prequal_ginsi_bmge.phy\n",
    "\n",
    "cat \"$1\"/concat_107t_93g_prequal_ginsi_bmge.phy | tr '=' '_' > phylip\n",
    "mv phylip \"$1\"/concat_107t_93g_prequal_ginsi_bmge.phy"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "The following command was used to run PhyloBayes (with three chains)."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "mpirun -np 10 pb_mpi -d concat_107t_93g_prequal_ginsi_bmge.phy -cat -gtr -dgam 4 -dc concat_107t_93g_prequal_ginsi_bmge_catgtr_chain1"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "We then ran posterior predictive tests in PhyloBayes. \n",
    "\n",
    "*Compositional heterogeneity.*  \n",
    "The test simulates alignments based on model parameter configurations sampled from the chains (chain 2 in this case) and then performs a posterior predictive test of compositional homogeneity. The test statistic is the maximum squared deviation between global and taxon-specific empirical frequencies. The observed value is computed on the true data, and compared with its null posterior predictive distribution. We saw that the observed max heterogeneity is an order of magnitude higher, and is significantly different (p value = 0). Same for the mean heterogeneity.\n",
    "\n",
    "*Diversity test.*  \n",
    "The diversity (div) test is doing a similar thing: simulating an alignment based on the model parameters from chain 2. The test statistic is the mean diversity per site (mean number of distinct amino-acid per sites). This is also not being sufficiently well captured."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "mpirun -np 20 /opt/pbmpi-master_1.9/data/readpb_mpi -ppred -comp -x 2000 1 9000 concat_107t_93g_prequal_ginsi_bmge_chain2 > concat_107t_93g_prequal_ginsi_bmge_chain2_comp.out\n",
    "mpirun -np 20 /opt/pbmpi-master_1.9/data/readpb_mpi -ppred -div -x 2000 1 9000 concat_107t_93g_prequal_ginsi_bmge_chain2 > concat_107t_93g_prequal_ginsi_bmge_chain2_div.out"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## 8. Run tree with CATPMSF model\n",
    "\n",
    "We use the appraoch of [Szánthó et al 2023](https://doi.org/10.1093/sysbio/syad013) to run the CATPMSF model. In the paper, the authors infer a ML tree using a site homogenous model and use it as a fixed tree for the PhyloBayes analysis (using CATGTR). Here, we will generate a tree under the LG+G model to use as fixed tree for the CATGTR analysis. The results of the phylobayes analysis will be used to extract site-specific stationary distributions and exchangeabilities."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### 8.1. Run tree with site homogenous model"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [],
   "source": [
    "## Directory for fasta file\n",
    "CONCAT_DIR = paths_dict[\"ANALYSIS_DATA\"][\"CONCAT_GENE_ANALYSIS\"][\"SUBSET\"][\"ALIGNMENTS\"][\"V14\"][\"CONCAT\"]\n",
    "\n",
    "## Directory for tree\n",
    "TREE_DIR = paths_dict[\"ANALYSIS_DATA\"][\"CONCAT_GENE_ANALYSIS\"][\"SUBSET\"][\"TREES\"][\"V14\"]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "%%bash -s \"$CONCAT_DIR\" \"$TREE_DIR\"\n",
    "\n",
    "sbatch ../../uppmax_scripts/script_bin/job_iqtree.sh \"$1\"/concat_107t_93g_prequal_ginsi_bmge.fasta \"$2\"/catpmsf/concat_107t_93g_prequal_ginsi_bmge_LG-G"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### 8.2. Use as guide tree to run PhyloBayes"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [],
   "source": [
    "## Directory for fasta file\n",
    "CONCAT_DIR = paths_dict[\"ANALYSIS_DATA\"][\"CONCAT_GENE_ANALYSIS\"][\"SUBSET\"][\"ALIGNMENTS\"][\"V14\"][\"CONCAT\"]\n",
    "\n",
    "## Directory for tree\n",
    "TREE_DIR = paths_dict[\"ANALYSIS_DATA\"][\"CONCAT_GENE_ANALYSIS\"][\"SUBSET\"][\"TREES\"][\"V14\"]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "%%bash -s \"$CONCAT_DIR\" \"$TREE_DIR\"\n",
    "\n",
    "sbatch ../../uppmax_scripts/script_bin/job_pb.sh \"$1\"/concat_107t_93g_prequal_ginsi_bmge.phy \"$2\"/catpmsf/concat_107t_93g_prequal_ginsi_bmge_LG-G.treefile \"$2\"/catpmsf/concat_107t_93g_prequal_ginsi_bmge_catgtr_chain1\n",
    "sleep 1\n",
    "sbatch ../../uppmax_scripts/script_bin/job_pb.sh \"$1\"/concat_107t_93g_prequal_ginsi_bmge.phy \"$2\"/catpmsf/concat_107t_93g_prequal_ginsi_bmge_LG-G.treefile \"$2\"/catpmsf/concat_107t_93g_prequal_ginsi_bmge_catgtr_chain2"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### 8.3 Check for convergence\n",
    "\n",
    "The chains were run for more than 5,000 cycles after which we checked for convergence. "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [],
   "source": [
    "## Directory for tree\n",
    "TREE_DIR = paths_dict[\"ANALYSIS_DATA\"][\"CONCAT_GENE_ANALYSIS\"][\"SUBSET\"][\"TREES\"][\"V14\"]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "%%bash -s \"$TREE_DIR\"\n",
    "\n",
    "sbatch ../../uppmax_scripts/script_bin/job_pb-convergence.sh \"$1\"/catpmsf/concat_107t_93g_prequal_ginsi_bmge_catgtr"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "name                effsize\trel_diff\n",
      "\n",
      "loglik              35\t\t1.6552\n",
      "length              113\t\t0.0624932\n",
      "alpha               330\t\t0.525332\n",
      "Nmode               146\t\t0.227368\n",
      "statent             118\t\t3.53681\n",
      "statalpha           110\t\t0.210322\n",
      "rrent               104\t\t0.979222\n",
      "rrmean              3525\t\t0.0110885\n"
     ]
    }
   ],
   "source": [
    "%%bash -s \"$TREE_DIR\"\n",
    "\n",
    "cat \"$1\"/catpmsf/tracecomp.contdiff"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "The effective sample sizes are above 100 for 7 out of 8 variables when setting burn-in to 1800 cycles. While this is not ideal, running the chains for longer is unlikely to improve things. I therefore decided to proceed as is (plus [Szánthó et al 2023](https://doi.org/10.1093/sysbio/syad013) failed to get convergence in all the empirical datasets they tested)."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### 8.4 Export and convert to site distributions and exchangeabilities"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "%%bash -s \"$TREE_DIR\"\n",
    "\n",
    "sbatch ../../uppmax_scripts/script_bin/job_sitedists_exchangeabilities.sh \"$1\"/catpmsf/concat_107t_93g_prequal_ginsi_bmge_catgtr_chain1"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "The generated exchangeabilities file is in paml format, so I manually turned it into a nexus file which is required by iqtree v2. "
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### 8.5. Run tree"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [],
   "source": [
    "## Directory for fasta file\n",
    "CONCAT_DIR = paths_dict[\"ANALYSIS_DATA\"][\"CONCAT_GENE_ANALYSIS\"][\"SUBSET\"][\"ALIGNMENTS\"][\"V14\"][\"CONCAT\"]\n",
    "\n",
    "## Directory for tree\n",
    "TREE_DIR = paths_dict[\"ANALYSIS_DATA\"][\"CONCAT_GENE_ANALYSIS\"][\"SUBSET\"][\"TREES\"][\"V14\"]"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Run tree with CAT-PMSF model and 100 non-parametric bootstraps."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Submitted batch job 9730163 on cluster snowy\n"
     ]
    }
   ],
   "source": [
    "%%bash -s \"$CONCAT_DIR\" \"$TREE_DIR\"\n",
    "\n",
    "sbatch ../../uppmax_scripts/script_bin/job_iqtree.sh \"$1\"/concat_107t_93g_prequal_ginsi_bmge.fasta \"$2\"/catpmsf/concat_107t_93g_prequal_ginsi_bmge_CATPMSF_fbp \"$2\"/catpmsf/concat_107t_93g_prequal_ginsi_bmge_catgtr_chain1.exchangeabilities \"$2\"/catpmsf/concat_107t_93g_prequal_ginsi_bmge_catgtr_chain1.sitefreq"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "I was also curious about the LG-PMSF model so I ran that with 1000 ufb."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Submitted batch job 9730164 on cluster snowy\n"
     ]
    }
   ],
   "source": [
    "%%bash -s \"$CONCAT_DIR\" \"$TREE_DIR\"\n",
    "\n",
    "sbatch ../../uppmax_scripts/script_bin/job_2023_11_16_iqtree.sh \"$1\"/concat_107t_93g_prequal_ginsi_bmge.fasta \"$2\"/catpmsf/concat_107t_93g_prequal_ginsi_bmge_LGPMSF \"$2\"/catpmsf/concat_107t_93g_prequal_ginsi_bmge_catgtr_chain1.sitefreq"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## 9. Estimate alternative topologies with CAT-PMSF model"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "We want to assess several alternative hypotheses relating to (1) the position of NEW, and (2) the monophyly of complex plastids. These branches in question are the ones with the lowest statistical support in all analyses run so far. We set up the following constraints: \n",
    "\n",
    "1. NEW sister to (haptophytes + cryptophytes), complex plastids monophyletic   \n",
    "2. NEW sister to (haptophytes + cryptophytes), complex plastids non-monophyletic    \n",
    "3. NEW sister to cryptophytes, complex plastids monophyletic   \n",
    "4. NEW sister to cryptophytes, complex plastids not monophyletic   \n",
    "5. NEW sister to haptophytes, complex plastids monophyletic  \n",
    "6. NEW sister to haptophytes, complex plastids non-monophyletic  \n",
    "\n",
    "For each constraint, we set up 5 independent searches to check that we were estimating best ML tree."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {},
   "outputs": [],
   "source": [
    "## Directory for fasta file\n",
    "CONCAT_DIR = paths_dict[\"ANALYSIS_DATA\"][\"CONCAT_GENE_ANALYSIS\"][\"SUBSET\"][\"ALIGNMENTS\"][\"V14\"][\"CONCAT\"]\n",
    "\n",
    "## Directory for tree\n",
    "TREE_DIR = paths_dict[\"ANALYSIS_DATA\"][\"CONCAT_GENE_ANALYSIS\"][\"SUBSET\"][\"TREES\"][\"V14\"]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "%%bash -s \"$CONCAT_DIR\" \"$TREE_DIR\"\n",
    "\n",
    "sbatch ../../uppmax_scripts/script_bin/job_iqtree.sh \"$1\"/concat_107t_93g_prequal_ginsi_bmge.fasta \"$2\"/catpmsf/concat_107t_93g_catpmsf_new-sister-c_complex-mono \"$2\"/catpmsf/concat_107t_93g_prequal_ginsi_bmge_catgtr_chain1.exchangeabilities \"$2\"/catpmsf/concat_107t_93g_prequal_ginsi_bmge_catgtr_chain1.sitefreq \"$2\"/constraints/new-sister-c_complex-mono.tre\n",
    "sleep 1\n",
    "\n",
    "sbatch ../../uppmax_scripts/script_bin/job_iqtree.sh \"$1\"/concat_107t_93g_prequal_ginsi_bmge.fasta \"$2\"/catpmsf/concat_107t_93g_catpmsf_new-sister-c_complex-non-mono \"$2\"/catpmsf/concat_107t_93g_prequal_ginsi_bmge_catgtr_chain1.exchangeabilities \"$2\"/catpmsf/concat_107t_93g_prequal_ginsi_bmge_catgtr_chain1.sitefreq \"$2\"/constraints/new-sister-c_complex-non-mono.tre\n",
    "sleep 1\n",
    "\n",
    "sbatch ../../uppmax_scripts/script_bin/job_iqtree.sh \"$1\"/concat_107t_93g_prequal_ginsi_bmge.fasta \"$2\"/catpmsf/concat_107t_93g_catpmsf_new-sister-h_complex-mono \"$2\"/catpmsf/concat_107t_93g_prequal_ginsi_bmge_catgtr_chain1.exchangeabilities \"$2\"/catpmsf/concat_107t_93g_prequal_ginsi_bmge_catgtr_chain1.sitefreq \"$2\"/constraints/new-sister-h_complex-mono.tre\n",
    "sleep 1\n",
    "\n",
    "sbatch ../../uppmax_scripts/script_bin/job_iqtree.sh \"$1\"/concat_107t_93g_prequal_ginsi_bmge.fasta \"$2\"/catpmsf/concat_107t_93g_catpmsf_new-sister-h_complex-non-mono \"$2\"/catpmsf/concat_107t_93g_prequal_ginsi_bmge_catgtr_chain1.exchangeabilities \"$2\"/catpmsf/concat_107t_93g_prequal_ginsi_bmge_catgtr_chain1.sitefreq \"$2\"/constraints/new-sister-h_complex-non-mono.tre\n",
    "sleep 1\n",
    "\n",
    "sbatch ../../uppmax_scripts/script_bin/job_iqtree.sh \"$1\"/concat_107t_93g_prequal_ginsi_bmge.fasta \"$2\"/catpmsf/concat_107t_93g_catpmsf_new-sister-h-c_complex-mono \"$2\"/catpmsf/concat_107t_93g_prequal_ginsi_bmge_catgtr_chain1.exchangeabilities \"$2\"/catpmsf/concat_107t_93g_prequal_ginsi_bmge_catgtr_chain1.sitefreq \"$2\"/constraints/new-sister-h-c_complex-mono.tre\n",
    "sleep 1\n",
    "\n",
    "sbatch ../../uppmax_scripts/script_bin/job_iqtree.sh \"$1\"/concat_107t_93g_prequal_ginsi_bmge.fasta \"$2\"/catpmsf/concat_107t_93g_catpmsf_new-sister-h-c_complex-non-mono \"$2\"/catpmsf/concat_107t_93g_prequal_ginsi_bmge_catgtr_chain1.exchangeabilities \"$2\"/catpmsf/concat_107t_93g_prequal_ginsi_bmge_catgtr_chain1.sitefreq \"$2\"/constraints/new-sister-h-c_complex-non-mono.tre"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "We can now check if the AU test rejects any of the topologies under the cat-pmsf model. Let's concatenate the trees we want to test first."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {},
   "outputs": [],
   "source": [
    "%%bash -s \"$TREE_DIR\"\n",
    "\n",
    "cat \"$1\"/catpmsf/concat_107t_93g_catpmsf_new-sister-h-c_complex-mono.treefile \\\n",
    "    \"$1\"/catpmsf/concat_107t_93g_catpmsf_new-sister-h-c_complex-non-mono.treefile \\\n",
    "    \"$1\"/catpmsf/concat_107t_93g_catpmsf_new-sister-c_complex-mono.treefile \\\n",
    "    \"$1\"/catpmsf/concat_107t_93g_catpmsf_new-sister-h_complex-mono.treefile \\\n",
    "    \"$1\"/catpmsf/concat_107t_93g_catpmsf_new-sister-h_complex-non-mono.treefile \\\n",
    "    \"$1\"/catpmsf/concat_107t_93g_catpmsf_new-sister-c_complex-non-mono.treefile \\\n",
    "    > \"$1\"/catpmsf/au_test.trees"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Submitted batch job 9735351 on cluster snowy\n"
     ]
    }
   ],
   "source": [
    "%%bash -s \"$CONCAT_DIR\" \"$TREE_DIR\"\n",
    "\n",
    "sbatch ../../uppmax_scripts/script_bin/job_topology_test.sh \\\n",
    "    \"$1\"/concat_107t_93g_prequal_ginsi_bmge.fasta \\\n",
    "    \"$2\"/catpmsf/concat_107t_93g_prequal_ginsi_bmge_au-test \\\n",
    "    \"$2\"/catpmsf/concat_107t_93g_prequal_ginsi_bmge_catgtr_chain1.exchangeabilities \\\n",
    "    \"$2\"/catpmsf/concat_107t_93g_prequal_ginsi_bmge_catgtr_chain1.sitefreq \\\n",
    "    \"$2\"/catpmsf/concat_107t_93g_prequal_ginsi_bmge_CATPMSF_fbp.treefile \\\n",
    "    \"$2\"/catpmsf/au_test.trees"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## 10. Estimate site profiles with MEOW \n",
    "I will now estimate site profiles using [MEOW](https://github.com/jdaneau/pm), which is an extension of MAMMaL. The difference between MAMMaL and MEOW is that while MAMMaL only uses high rate sites for estimating custom site profiles, MEOW uses all sites in the input alignment for estimating custom site profiles. \n",
    "\n",
    "Here, we will follow [Williamson et al 2024](https://doi.org/10.1101/2024.09.04.611237) for using MEOW."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### 10.1. Get phylip file for MEOW input\n",
    "\n",
    "MEOW requires a phylip file with 10 character headers as input. So we replace the fasta headers of the V14 concat file, and convert to phylip format."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [],
   "source": [
    "## Directory for V12 concat fasta files\n",
    "CONCAT_DIR = paths_dict[\"ANALYSIS_DATA\"][\"CONCAT_GENE_ANALYSIS\"][\"SUBSET\"][\"ALIGNMENTS\"][\"V14\"][\"CONCAT\"]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "%%bash -s \"$CONCAT_DIR\"\n",
    "\n",
    "## replace fasta headers\n",
    "seqkit replace -p '^(\\S+)$' -r '{kv}$2' \\\n",
    "-k \"$1\"/replace_headers.txt \"$1\"/concat_107t_93g_prequal_ginsi_bmge.fasta \\\n",
    "> \"$1\"/concat_107t_93g_prequal_ginsi_bmge_replaced.fasta"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "%%bash -s \"$CONCAT_DIR\"\n",
    "\n",
    "## convert to phylip\n",
    "perl ../../src/fasta2phylip.pl -f \"$1\"/concat_107t_93g_prequal_ginsi_bmge_replaced.fasta \\\n",
    "-o \"$1\"/concat_107t_93g_prequal_ginsi_bmge_replaced.phylip\n",
    "\n",
    "## replace '=' with '_' so it will be consistent with the tree in the next step\n",
    "cat \"$1\"/concat_107t_93g_prequal_ginsi_bmge_replaced.phylip | \\\n",
    "tr '=' '_' \\\n",
    "> phylip\n",
    "\n",
    "mv phylip \"$1\"/concat_107t_93g_prequal_ginsi_bmge_replaced.phylip"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### 10.2. Get tree for MEOW input\n",
    "We will use the tree inferred with the V14 concat phylogeny and the cpREV+C60+G model as a guide tree for inferring the custom 80 site frequency profiles with MEOW. We now need to replace the tip labels of this tree so that it is consistent with the input phylip file."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [],
   "source": [
    "## Directory for fasta file\n",
    "CONCAT_DIR = paths_dict[\"ANALYSIS_DATA\"][\"CONCAT_GENE_ANALYSIS\"][\"SUBSET\"][\"ALIGNMENTS\"][\"V14\"][\"CONCAT\"]\n",
    "\n",
    "## Directory for tree\n",
    "TREE_DIR = paths_dict[\"ANALYSIS_DATA\"][\"CONCAT_GENE_ANALYSIS\"][\"SUBSET\"][\"TREES\"][\"V14\"]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "metadata": {},
   "outputs": [],
   "source": [
    "%%bash -s \"$CONCAT_DIR\" \"$TREE_DIR\"\n",
    "\n",
    "## copy the phylip file to the new mammal directory \n",
    "cp \"$1\"/concat_107t_93g_prequal_ginsi_bmge_replaced.phylip \"$2\"/meow/.\n",
    "\n",
    "perl ../../src/replace_tip_labels.pl \"$2\"/concat_107t_93g_prequal_ginsi_bmge_cpREV-C60-G.treefile \\\n",
    "<(cat \"$1\"/replace_headers.txt | tr '=' '_') \\\n",
    "\"$2\"/meow/concat_107t_93g_prequal_ginsi_bmge_cpREV-C60-G_replaced.treefile"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### 10.3. Run MEOW\n",
    "\n",
    "We run MEOW. Following Williamson et al 2024, we remove invariant sites from the alignment from consideration and use a guide tree (inferred with a C60 mixture model) to help estimate site rates of the variable sites using the Discrete\n",
    "Gamma Probability Estimate (DGPE) method of site rate estimation ([Susko et al 2003](https://doi.org/10.1080/10635150390235395))."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "metadata": {},
   "outputs": [],
   "source": [
    "## Directory for fasta file\n",
    "CONCAT_DIR = paths_dict[\"ANALYSIS_DATA\"][\"CONCAT_GENE_ANALYSIS\"][\"SUBSET\"][\"ALIGNMENTS\"][\"V14\"][\"CONCAT\"]\n",
    "\n",
    "## Directory for tree\n",
    "TREE_DIR = paths_dict[\"ANALYSIS_DATA\"][\"CONCAT_GENE_ANALYSIS\"][\"SUBSET\"][\"TREES\"][\"V14\"]"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Starting first with MEOW(80,0)."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "%%bash -s \"$TREE_DIR\"\n",
    "\n",
    "meow -s \"$1\"/meow/concat_106t_93g_prequal_ginsi_bmge_replaced.phylip \\\n",
    "    -ch 80 \\\n",
    "    -cl 0 \\\n",
    "    -t \"$1\"/meow/concat_106t_93g_prequal_ginsi_bmge_cpREV-C60-G_replaced.treefile \\\n",
    "    -ri \\\n",
    "    -p R \\\n",
    "    -C 5 \\\n",
    "    -f H \\ \n",
    "    -o \"$1\"/meow/meow_80_0"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Then MEOW(60,20). After initially getting an error (\"Error in solve.QP(Dmat = Sigma * 2, dvec = rep(0, ntaxa), Amat = Amat,  : matrix D in quadratic function is not positive definite!\"), I used the `-l` flag which does not use likelihood weights for the low rate partition (a custom version of the script kindly provided by Hector Baños). (Custom script provided in the `src` directory)."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "%%bash -s \"$TREE_DIR\"\n",
    "\n",
    "meow_custom.R -s \"$1\"/meow/concat_106t_93g_prequal_ginsi_bmge_replaced.phylip \\\n",
    "    -ch 60 \\\n",
    "    -cl 20 \\\n",
    "    -t \"$1\"/meow/concat_106t_93g_prequal_ginsi_bmge_cpREV-C60-G_replaced.treefile \\\n",
    "    -ri \\\n",
    "    -p R \\\n",
    "    -C 5 \\\n",
    "    -f H \\ \n",
    "    -l \\\n",
    "    -o \"$1\"/meow/meow_60_20"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Finally, MEOW(40,40)."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "%%bash -s \"$TREE_DIR\"\n",
    "\n",
    "meow_custom.R -s \"$1\"/meow/concat_106t_93g_prequal_ginsi_bmge_replaced.phylip \\\n",
    "    -ch 40 \\\n",
    "    -cl 40 \\\n",
    "    -t \"$1\"/meow/concat_106t_93g_prequal_ginsi_bmge_cpREV-C60-G_replaced.treefile \\\n",
    "    -ri \\\n",
    "    -p R \\\n",
    "    -C 5 \\\n",
    "    -f H \\ \n",
    "    -l \\\n",
    "    -o \"$1\"/meow/meow_40_40"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### 10.4. IQTree Model Test\n",
    "\n",
    "I was interested in trying a regular model test in IQTree using the BIC criteria. I'll be testing the models:  \n",
    "  \n",
    "- cpREV+MEOW(40,40)+G   \n",
    "- cpREV+MEOW(60,20)+G   \n",
    "- cpREV+MEOW(80,0)+G     \n",
    "- LG+MEOW(40,40)+G    \n",
    "- LG+MEOW(60,20)+G    \n",
    "- LG+MEOW(80,0)+G   \n",
    "\n",
    "There is a small caveat of using BIC. According to Hector Baños: \"Since the classes are estimated from the data, these are also parameters, nonetheless, these are not accounted for in model finder and even if you do account for them they were not estimated under the likelihood estimation.  This is why you cannot compare  LG+C60+G vs LG+MAM60+G using a model finder (and we do cross-validation). Without giving it much thought, I think if you only compare  MEOW(60,0), MEOW(40,20), and MEOW(30,30) then you can get a proxy since all these models estimate the classes similarly so AIC and BIC can give you a good estimation.\"\n",
    "\n",
    "So, I use BIC to get a reasonable estimation of the best model since I'm testing models with 80 classes. "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [],
   "source": [
    "## Directory for fasta file\n",
    "CONCAT_DIR = paths_dict[\"ANALYSIS_DATA\"][\"CONCAT_GENE_ANALYSIS\"][\"SUBSET\"][\"ALIGNMENTS\"][\"V14\"][\"CONCAT\"]\n",
    "\n",
    "## Directory for tree\n",
    "TREE_DIR = paths_dict[\"ANALYSIS_DATA\"][\"CONCAT_GENE_ANALYSIS\"][\"SUBSET\"][\"TREES\"][\"V14\"]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "%%bash -s \"$CONCAT_DIR\" \"$TREE_DIR\"\n",
    "\n",
    "sbatch ../../uppmax_scripts/script_bin/job_iqtree.sh \"$1\"/concat_107t_93g_prequal_ginsi_bmge.fasta \"$2\"/meow/concat_107t_93g_prequal_ginsi_bmge_modeltest \"$2\"/meow/meow.nex"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "The best fit model was LG+MEOW(60,20)+G. We now run a tree with that model. "
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## 11. Run tree with MEOW(60,20) model\n",
    "\n",
    "The best fit model! We run an unconstrained tree and estimate support with 1000 ufb."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [],
   "source": [
    "## Directory for fasta file\n",
    "CONCAT_DIR = paths_dict[\"ANALYSIS_DATA\"][\"CONCAT_GENE_ANALYSIS\"][\"SUBSET\"][\"ALIGNMENTS\"][\"V14\"][\"CONCAT\"]\n",
    "\n",
    "## Directory for tree\n",
    "TREE_DIR = paths_dict[\"ANALYSIS_DATA\"][\"CONCAT_GENE_ANALYSIS\"][\"SUBSET\"][\"TREES\"][\"V14\"]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "%%bash -s \"$CONCAT_DIR\" \"$TREE_DIR\"\n",
    "\n",
    "sbatch ../../uppmax_scripts/script_bin/job_iqtree.sh \"$1\"/concat_107t_93g_prequal_ginsi_bmge.fasta \"$2\"/meow/concat_107t_93g_prequal_ginsi_bmge_LG-MEOW6020-G \"$2\"/meow/meow.nex"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "The topology recovered from this tree was exactly the same as the LG+C60+G tree: (1) NEW sister to cryptophytes and haptophytes (61% ufb suport for the monophyly of cryptophytes and haptophytes), and (2) complex plastids monophyletic (but with lower support this time of 68%)."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## 12. Estimate alternative topologies with LG+MEOW(60,20)+G model\n",
    "\n",
    "We want to assess several alternative hypotheses relating to (1) the position of NEW, and (2) the monophyly of complex plastids. These branches in question are the ones with the lowest statistical support in all analyses run so far. We set up the following constraints: \n",
    "\n",
    "1. NEW sister to (haptophytes + cryptophytes), complex plastids monophyletic   \n",
    "2. NEW sister to (haptophytes + cryptophytes), complex plastids non-monophyletic    \n",
    "3. NEW sister to cryptophytes, complex plastids monophyletic   \n",
    "4. NEW sister to cryptophytes, complex plastids not monophyletic   \n",
    "5. NEW sister to haptophytes, complex plastids monophyletic  \n",
    "6. NEW sister to haptophytes, complex plastids non-monophyletic  \n",
    "\n",
    "For each constraint, we set up 10 independent searches, as [Liu et al 2024](https://doi.org/10.1093/sysbio/syae031) showed that the number of tree searches considerably influences the identification of phylogenies with the highest log-likelihood scores."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [],
   "source": [
    "## Directory for fasta file\n",
    "CONCAT_DIR = paths_dict[\"ANALYSIS_DATA\"][\"CONCAT_GENE_ANALYSIS\"][\"SUBSET\"][\"ALIGNMENTS\"][\"V14\"][\"CONCAT\"]\n",
    "\n",
    "## Directory for tree\n",
    "TREE_DIR = paths_dict[\"ANALYSIS_DATA\"][\"CONCAT_GENE_ANALYSIS\"][\"SUBSET\"][\"TREES\"][\"V14\"]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "%%bash -s \"$CONCAT_DIR\" \"$TREE_DIR\"\n",
    "\n",
    "sbatch ../../uppmax_scripts/script_bin/job_iqtree.sh \"$1\"/concat_107t_93g_prequal_ginsi_bmge.fasta \"$2\"/meow/concat_107t_93g_LG-MEOW6020-G_new-sister-c_complex-mono_2 \"$2\"/meow/meow.nex \"$2\"/constraints/new-sister-c_complex-mono.tre\n",
    "sleep 1\n",
    "\n",
    "sbatch ../../uppmax_scripts/script_bin/job_iqtree.sh \"$1\"/concat_107t_93g_prequal_ginsi_bmge.fasta \"$2\"/meow/concat_107t_93g_LG-MEOW6020-G_new-sister-c_complex-non-mono_2 \"$2\"/meow/meow.nex \"$2\"/constraints/new-sister-c_complex-non-mono.tre\n",
    "sleep 1\n",
    "\n",
    "sbatch ../../uppmax_scripts/script_bin/job_iqtree.sh \"$1\"/concat_107t_93g_prequal_ginsi_bmge.fasta \"$2\"/meow/concat_107t_93g_LG-MEOW6020-G_new-sister-h_complex-mono_2 \"$2\"/meow/meow.nex \"$2\"/constraints/new-sister-h_complex-mono.tre\n",
    "sleep 1\n",
    "\n",
    "sbatch ../../uppmax_scripts/script_bin/job_iqtree.sh \"$1\"/concat_107t_93g_prequal_ginsi_bmge.fasta \"$2\"/meow/concat_107t_93g_LG-MEOW6020-G_new-sister-h_complex-non-mono_2 \"$2\"/meow/meow.nex \"$2\"/constraints/new-sister-h_complex-non-mono.tre\n",
    "sleep 1\n",
    "\n",
    "sbatch ../../uppmax_scripts/script_bin/job_iqtree.sh \"$1\"/concat_107t_93g_prequal_ginsi_bmge.fasta \"$2\"/meow/concat_107t_93g_LG-MEOW6020-G_new-sister-h-c_complex-mono_2 \"$2\"/meow/meow.nex \"$2\"/constraints/new-sister-h-c_complex-mono.tre\n",
    "sleep 1\n",
    "\n",
    "sbatch ../../uppmax_scripts/script_bin/job_iqtree.sh \"$1\"/concat_107t_93g_prequal_ginsi_bmge.fasta \"$2\"/meow/concat_107t_93g_LG-MEOW6020-G_new-sister-h-c_complex-non-mono_2 \"$2\"/meow/meow.nex \"$2\"/constraints/new-sister-h-c_complex-non-mono.tre"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## 13. AU test of constrained trees inferred with LG+MEOW(60,20) model\n",
    "\n",
    "We perform the Approximately Unbiased (AU) test to see if any of the following topologies is rejected by the AU test:\n",
    "\n",
    "1. NEW sister to (haptophytes + cryptophytes), complex plastids monophyletic   \n",
    "2. NEW sister to (haptophytes + cryptophytes), complex plastids non-monophyletic    \n",
    "3. NEW sister to cryptophytes, complex plastids monophyletic   \n",
    "4. NEW sister to cryptophytes, complex plastids not monophyletic   \n",
    "5. NEW sister to haptophytes, complex plastids monophyletic  \n",
    "6. NEW sister to haptophytes, complex plastids non-monophyletic "
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "We first concatenate the trees that we want to test."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [],
   "source": [
    "## Directory for fasta file\n",
    "CONCAT_DIR = paths_dict[\"ANALYSIS_DATA\"][\"CONCAT_GENE_ANALYSIS\"][\"SUBSET\"][\"ALIGNMENTS\"][\"V14\"][\"CONCAT\"]\n",
    "\n",
    "## Directory for tree\n",
    "TREE_DIR = paths_dict[\"ANALYSIS_DATA\"][\"CONCAT_GENE_ANALYSIS\"][\"SUBSET\"][\"TREES\"][\"V14\"]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [],
   "source": [
    "%%bash -s \"$TREE_DIR\"\n",
    "\n",
    "cat \"$1\"/meow/concat_107t_93g_LG-MEOW6020-G_new-sister-h-c_complex-mono.treefile \\\n",
    "    \"$1\"/meow/concat_107t_93g_LG-MEOW6020-G_new-sister-h-c_complex-non-mono.treefile \\\n",
    "    \"$1\"/meow/concat_107t_93g_LG-MEOW6020-G_new-sister-c_complex-mono.treefile \\\n",
    "    \"$1\"/meow/concat_107t_93g_LG-MEOW6020-G_new-sister-h_complex-mono_2.treefile \\\n",
    "    \"$1\"/meow/concat_107t_93g_LG-MEOW6020-G_new-sister-h_complex-non-mono.treefile \\\n",
    "    \"$1\"/meow/concat_107t_93g_LG-MEOW6020-G_new-sister-c_complex-non-mono_2.treefile \\\n",
    "    > \"$1\"/meow/au_test.trees"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "We set up the AU test! Here we estimate model parameters based on the \"best\" ML tree reconstructed previously."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Submitted batch job 9730443 on cluster snowy\n"
     ]
    }
   ],
   "source": [
    "%%bash -s \"$CONCAT_DIR\" \"$TREE_DIR\"\n",
    "\n",
    "sbatch ../../uppmax_scripts/script_bin/job_topology_test.sh \\\n",
    "    \"$1\"/concat_107t_93g_prequal_ginsi_bmge.fasta \\\n",
    "    \"$2\"/meow/concat_107t_93g_prequal_ginsi_bmge_au-test \\\n",
    "    \"$2\"/meow/meow.nex \\\n",
    "    \"$2\"/meow/concat_107t_93g_prequal_ginsi_bmge_LG-MEOW6020-G.treefile \\\n",
    "    \"$2\"/meow/au_test.trees"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "None of the topologies was rejected (unsurprisingly). Perhaps the guide tree used to estimate parameters is super important? I use the best scoring constrained tree (which had a higher log-likelihood than the unconstrained tree used before) found so far to run the AU test. "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Submitted batch job 9735347 on cluster snowy\n"
     ]
    }
   ],
   "source": [
    "%%bash -s \"$CONCAT_DIR\" \"$TREE_DIR\"\n",
    "\n",
    "sbatch ../../uppmax_scripts/script_bin/job_topology_test.sh \\\n",
    "    \"$1\"/concat_107t_93g_prequal_ginsi_bmge.fasta \\\n",
    "    \"$2\"/meow/concat_107t_93g_prequal_ginsi_bmge_au-test2 \\\n",
    "    \"$2\"/meow/meow.nex \\\n",
    "    \"$2\"/meow/concat_107t_93g_LG-MEOW6020-G_new-sister-h-c_complex-mono.treefile \\\n",
    "    \"$2\"/meow/au_test.trees"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## 14. Removing compositionally heterogenous sites\n",
    "\n",
    "The previous models (CAT-GTR, MOEW models) assume that the amino acid frequencies for every branch in the tree will be similar. However, this assumption might not hold true, and serious violations can mislead phylogenetic inference. This may be the case as our dataset includes cyanidiales which are known extremophiles. "
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### 14.1. Investigating compositional heterogeneity\n",
    "We first need to calculate the amino acid composition for each taxon. This is easily done using the tool [nRCFV_Reader](https://github.com/JFFleming/RCFV_Reader) ([Fleming and Struck 2023](https://doi.org/10.1186/s12859-023-05270-8))."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [],
   "source": [
    "## Directory for fasta file\n",
    "CONCAT_DIR = paths_dict[\"ANALYSIS_DATA\"][\"CONCAT_GENE_ANALYSIS\"][\"SUBSET\"][\"ALIGNMENTS\"][\"V14\"][\"CONCAT\"]\n",
    "\n",
    "## Directory for tree\n",
    "TREE_DIR = paths_dict[\"ANALYSIS_DATA\"][\"CONCAT_GENE_ANALYSIS\"][\"SUBSET\"][\"TREES\"][\"V14\"]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "%%bash -s \"$CONCAT_DIR\" \"$TREE_DIR\"\n",
    "\n",
    "perl ../../src/RCFVReader_v1.pl protein \"$1\"/concat_107t_93g_prequal_ginsi_bmge.fasta \"$2\"/comp_het/concat_107t_93g_aa_comp"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "%%bash -s \"$TREE_DIR\"\n",
    "\n",
    "cat \"$1\"/comp_het/concat_107t_93g_aa_comp.ncsRCFV.txt"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "We see that amino acids I, V, A, and N have the highest normalised relative compositional frequency variability. \n",
    "\n",
    "The tool also generates a frequency file to which we can add taxonomic affiliation."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "%%bash -s \"$TREE_DIR\"\n",
    "\n",
    "cat \"$1\"/comp_het/concat_107t_93g_aa_comp.Frequencies.txt | \\\n",
    "    sed -E 's/(.*Ochrophyta.*)/\\1\\tOchrophyta/' | \\\n",
    "    sed -E 's/(.*Haptophyta.*)/\\1\\tHaptophyta/' | \\\n",
    "    sed -E 's/(.*Cryptop.*)/\\1\\tCryptophyta/' | \\\n",
    "    sed -E 's/(.*Glauco.*)/\\1\\tGlaucophyta/' | \\\n",
    "    sed -E 's/(.*Cyanidiales.*)/\\1\\tCyanidiophytina/' | \\\n",
    "    sed -E 's/(.*Galdieriales.*)/\\1\\tCyanidiophytina/' | \\\n",
    "    sed -E 's/(.*Florideo.*)/\\1\\tRhodophytina/' | \\\n",
    "    sed -E 's/(.*Compsopogono.*)/\\1\\tRhodophytina/' | \\\n",
    "    sed -E 's/(.*Porphyridiales.*)/\\1\\tRhodophytina/' | \\\n",
    "    sed -E 's/(.*Rhodello.*)/\\1\\tRhodophytina/' | \\\n",
    "    sed -E 's/(.*Bangiales.*)/\\1\\tRhodophytina/' | \\\n",
    "    sed -E 's/(.*Stylonema.*)/\\1\\tRhodophytina/' | \\\n",
    "    sed -E 's/(.*Viridiplantae.*)/\\1\\tViridiplantae/' | \\\n",
    "    sed -E 's/(.*New.*)/\\1\\tNEW/' | \\\n",
    "    grep -v \"Mean_Freq_Across_Taxa\" | \\\n",
    "    sed -E 's/(NAME.*)/\\1\\tGroup/' > \"$1\"/comp_het/concat_107_93g_aa_freq_group.txt"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "The table was analysed in the R script `PCA_aa_usage.R`. "
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### 14.2. Removing compositionally heterogenous sites and inferring a phylogeny\n",
    "\n",
    "Here, we trim the concatenated alignment with BMGE's stationary-based trimming. This method trims the alignment until the remaining characters are compositionally homogeneous, as assessed by Stuart’s test of marginal homogeneity between each pair of sequences."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [],
   "source": [
    "## Directory for fasta file\n",
    "CONCAT_DIR = paths_dict[\"ANALYSIS_DATA\"][\"CONCAT_GENE_ANALYSIS\"][\"SUBSET\"][\"ALIGNMENTS\"][\"V14\"][\"CONCAT\"]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "%%bash -s \"$CONCAT_DIR\"\n",
    "\n",
    "sbatch ../../uppmax_scripts/script_bin/job_bmge.sh \"$1\"/concat_107t_93g_prequal_ginsi_bmge.fasta \"$1\"/concat_107t_93g_prequal_ginsi_bmge.bmge.fasta"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "The final alignment has 107 taxa and 15,754 sites (i.e. roughly about 5,000 sites were removed). What is the new nRCFV score?"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [],
   "source": [
    "## Directory for fasta file\n",
    "CONCAT_DIR = paths_dict[\"ANALYSIS_DATA\"][\"CONCAT_GENE_ANALYSIS\"][\"SUBSET\"][\"ALIGNMENTS\"][\"V14\"][\"CONCAT\"]\n",
    "\n",
    "## Directory for tree\n",
    "TREE_DIR = paths_dict[\"ANALYSIS_DATA\"][\"CONCAT_GENE_ANALYSIS\"][\"SUBSET\"][\"TREES\"][\"V14\"]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "%%bash -s \"$CONCAT_DIR\" \"$TREE_DIR\"\n",
    "\n",
    "perl ../../src/RCFVReader_v1.pl protein \"$1\"/concat_107t_93g_prequal_ginsi_bmge.bmge.fasta \"$2\"/comp_het/concat_107t_93g_comp-rem_aa_comp"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "The nRCFV is much lower now, 0.00112250162175195 compared to 0.00311920771907216 for the untreated dataset."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "I decided to infer a tree with the LG+C60+G model. "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [],
   "source": [
    "## Directory for fasta file\n",
    "CONCAT_DIR = paths_dict[\"ANALYSIS_DATA\"][\"CONCAT_GENE_ANALYSIS\"][\"SUBSET\"][\"ALIGNMENTS\"][\"V14\"][\"CONCAT\"]\n",
    "\n",
    "## Directory for tree\n",
    "TREE_DIR = paths_dict[\"ANALYSIS_DATA\"][\"CONCAT_GENE_ANALYSIS\"][\"SUBSET\"][\"TREES\"][\"V14\"]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "%%bash -s \"$CONCAT_DIR\" \"$TREE_DIR\"\n",
    "\n",
    "sbatch ../../uppmax_scripts/script_bin/job_iqtree.sh \"$1\"/concat_107t_93g_prequal_ginsi_bmge.bmge.fasta \"$2\"/comp_het/concat_107t_93g_prequal_ginsi_bmge_bmge_LG-C60-G"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "And with the MEOW80 model."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "%%bash -s \"$CONCAT_DIR\" \"$TREE_DIR\"\n",
    "\n",
    "sbatch ../../uppmax_scripts/script_bin/job_iqtree.sh \"$1\"/concat_107t_93g_prequal_ginsi_bmge.bmge.fasta \"$2\"/comp_het/stationary_trimming/concat_107t_93g_prequal_ginsi_bmge_bmge_LG-MEOW6020-G \"$2\"/meow/meow.nex"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### 14.3 Inferring constrained trees\n",
    "\n",
    "As before, we can infer constrained trees under the LG+MEOW80+G model and compare their likelihoods."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [],
   "source": [
    "## Directory for fasta file\n",
    "CONCAT_DIR = paths_dict[\"ANALYSIS_DATA\"][\"CONCAT_GENE_ANALYSIS\"][\"SUBSET\"][\"ALIGNMENTS\"][\"V14\"][\"CONCAT\"]\n",
    "\n",
    "## Directory for tree\n",
    "TREE_DIR = paths_dict[\"ANALYSIS_DATA\"][\"CONCAT_GENE_ANALYSIS\"][\"SUBSET\"][\"TREES\"][\"V14\"]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "%%bash -s \"$CONCAT_DIR\" \"$TREE_DIR\"\n",
    "\n",
    "sbatch ../../uppmax_scripts/script_bin/job_iqtree.sh \"$1\"/concat_107t_93g_prequal_ginsi_bmge.bmge.fasta \"$2\"/comp_het/stationary_trimming/concat_107t_93g_LG-MEOW6020-G_new-sister-c_complex-mono \"$2\"/meow/meow.nex \"$2\"/constraints/new-sister-c_complex-mono.tre\n",
    "sleep 1\n",
    "\n",
    "sbatch ../../uppmax_scripts/script_bin/job_iqtree.sh \"$1\"/concat_107t_93g_prequal_ginsi_bmge.bmge.fasta \"$2\"/comp_het/stationary_trimming/concat_107t_93g_LG-MEOW6020-G_new-sister-c_complex-non-mono \"$2\"/meow/meow.nex \"$2\"/constraints/new-sister-c_complex-non-mono.tre\n",
    "sleep 1\n",
    "\n",
    "sbatch ../../uppmax_scripts/script_bin/job_iqtree.sh \"$1\"/concat_107t_93g_prequal_ginsi_bmge.bmge.fasta \"$2\"/comp_het/stationary_trimming/concat_107t_93g_LG-MEOW6020-G_new-sister-h_complex-mono \"$2\"/meow/meow.nex \"$2\"/constraints/new-sister-h_complex-mono.tre\n",
    "sleep 1\n",
    "\n",
    "sbatch ../../uppmax_scripts/script_bin/job_iqtree.sh \"$1\"/concat_107t_93g_prequal_ginsi_bmge.bmge.fasta \"$2\"/comp_het/stationary_trimming/concat_107t_93g_LG-MEOW6020-G_new-sister-h_complex-non-mono \"$2\"/meow/meow.nex \"$2\"/constraints/new-sister-h_complex-non-mono.tre\n",
    "sleep 1\n",
    "\n",
    "sbatch ../../uppmax_scripts/script_bin/job_iqtree.sh \"$1\"/concat_107t_93g_prequal_ginsi_bmge.bmge.fasta \"$2\"/comp_het/stationary_trimming/concat_107t_93g_LG-MEOW6020-G_new-sister-h-c_complex-mono \"$2\"/meow/meow.nex \"$2\"/constraints/new-sister-h-c_complex-mono.tre\n",
    "sleep 1\n",
    "\n",
    "sbatch ../../uppmax_scripts/script_bin/job_iqtree.sh \"$1\"/concat_107t_93g_prequal_ginsi_bmge.bmge.fasta \"$2\"/comp_het/stationary_trimming/concat_107t_93g_LG-MEOW6020-G_new-sister-h-c_complex-non-mono \"$2\"/meow/meow.nex \"$2\"/constraints/new-sister-h-c_complex-non-mono.tre"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### 14.4 Topology test\n",
    "\n",
    "We do the AU test to see which topologies can be statistically rejected. First, concatenate the trees we are testing."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {},
   "outputs": [],
   "source": [
    "%%bash -s \"$TREE_DIR\"\n",
    "\n",
    "cat \"$1\"/comp_het/stationary_trimming/concat_107t_93g_LG-MEOW6020-G_new-sister-c_complex-mono.treefile \\\n",
    "    \"$1\"/comp_het/stationary_trimming/concat_107t_93g_LG-MEOW6020-G_new-sister-c_complex-non-mono.treefile \\\n",
    "    \"$1\"/comp_het/stationary_trimming/concat_107t_93g_LG-MEOW6020-G_new-sister-h-c_complex-mono.treefile \\\n",
    "    \"$1\"/comp_het/stationary_trimming/concat_107t_93g_LG-MEOW6020-G_new-sister-h-c_complex-non-mono.treefile \\\n",
    "    \"$1\"/comp_het/stationary_trimming/concat_107t_93g_LG-MEOW6020-G_new-sister-h_complex-mono.treefile \\\n",
    "    \"$1\"/comp_het/stationary_trimming/concat_107t_93g_LG-MEOW6020-G_new-sister-h_complex-non-mono.treefile \\\n",
    "    > \"$1\"/comp_het/stationary_trimming/au_test.trees\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "We estimate parameters from the unconstrained ML tree."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "%%bash -s \"$CONCAT_DIR\" \"$TREE_DIR\"\n",
    "\n",
    "sbatch ../../uppmax_scripts/script_bin/job_topology_test.sh \"$1\"/concat_107t_93g_prequal_ginsi_bmge.bmge.fasta \"$2\"/comp_het/stationary_trimming/AU_test \"$2\"/meow/meow.nex \"$2\"/comp_het/stationary_trimming/concat_107t_93g_prequal_ginsi_bmge_bmge_LG-MEOW6020-G.treefile \"$2\"/comp_het/stationary_trimming/au_test.trees"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### 14.5 CAT-GTR tree\n",
    "\n",
    "We run a Bayesian phylogenetic inference in PhyloBayes. \n",
    "\n",
    "Convert alignment to phylip format first."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "%%bash -s \"$CONCAT_DIR\"\n",
    "\n",
    "perl ../../src/fasta2phylip.pl -f \"$1\"/concat_107t_93g_prequal_ginsi_bmge.bmge.fasta -o \"$1\"/concat_107t_93g_prequal_ginsi_bmge.bmge.phy\n",
    "\n",
    "cat \"$1\"/concat_107t_93g_prequal_ginsi_bmge.bmge.phy | tr '=' '_' > phylip\n",
    "mv phylip \"$1\"/concat_107t_93g_prequal_ginsi_bmge.bmge.phy"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Run three chains of PhyloBayes."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "%%bash -s \"$CONCAT_DIR\" \"$TREE_DIR\"\n",
    "\n",
    "sbatch ../../uppmax_scripts/script_bin/job_pb.sh \"$1\"/concat_107t_93g_prequal_ginsi_bmge.bmge.phy \"$2\"/comp_het/stationary_trimming/concat_107t_93g_prequal_ginsi_bmge.bmge_catgtr_chain1\n",
    "sleep 1\n",
    "\n",
    "sbatch ../../uppmax_scripts/script_bin/job_pb.sh \"$1\"/concat_107t_93g_prequal_ginsi_bmge.bmge.phy \"$2\"/comp_het/stationary_trimming/concat_107t_93g_prequal_ginsi_bmge.bmge_catgtr_chain2\n",
    "sleep 1\n",
    "\n",
    "sbatch ../../uppmax_scripts/script_bin/job_pb.sh \"$1\"/concat_107t_93g_prequal_ginsi_bmge.bmge.phy \"$2\"/comp_het/stationary_trimming/concat_107t_93g_prequal_ginsi_bmge.bmge_catgtr_chain3\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## 15. Modelling compositional heterogeneity with the GFmix model\n",
    "\n",
    "The GFmix model modifies the vector of amino acid frequencies in a branch-specific manner to account for shifts in the relative frequencies of amino acids in different branches of the tree. While Williamson et al 2024 used multiple partitions (5 partitions), we will use a single partition. This is because all of our genes are plastid encoded genes, and we do expect there to be lesser discrepancy in compositional heterogeneity compared to Williamson et al 2024, which had both mitochondrial and nuclear encoded genes. Here, we estimate the likelihood of the six ML trees, corresponding to different topologies, under the GFmix model. "
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "GFmix expects the sequence file to be in phylip format with names being 10 characters long, which should correspond to the names in the treefile. Let's fulfill these requirements. We already have the sequence file: `concat_107t_93g_prequal_ginsi_bmge_replaced.phylip`. Let's convert the names of the six treefiles we inferred earlier. "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [],
   "source": [
    "%%bash -s \"$CONCAT_DIR\" \"$TREE_DIR\"\n",
    "\n",
    "## copy the phylip file to the tree directory \n",
    "cp \"$1\"/concat_107t_93g_prequal_ginsi_bmge_replaced.phylip \"$2\"/comp_het/gfmix/.\n",
    "\n",
    "## replace tip labels of trees inferred with LG+MEOW80+G model\n",
    "perl ../../src/replace_tip_labels.pl \"$2\"/meow/concat_107t_93g_LG-MEOW6020-G_new-sister-h-c_complex-mono.treefile \\\n",
    "<(cat \"$1\"/replace_headers.txt | tr '=' '_') \\\n",
    "\"$2\"/comp_het/gfmix/concat_107t_93g_LG-MEOW6020-G_new-sister-h-c_complex-mono_replaced.treefile\n",
    "\n",
    "perl ../../src/replace_tip_labels.pl \"$2\"/meow/concat_107t_93g_LG-MEOW6020-G_new-sister-h-c_complex-non-mono.treefile \\\n",
    "<(cat \"$1\"/replace_headers.txt | tr '=' '_') \\\n",
    "\"$2\"/comp_het/gfmix/concat_107t_93g_LG-MEOW6020-G_new-sister-h-c_complex-non-mono_replaced.treefile\n",
    "\n",
    "perl ../../src/replace_tip_labels.pl \"$2\"/meow/concat_107t_93g_LG-MEOW6020-G_new-sister-c_complex-mono.treefile \\\n",
    "<(cat \"$1\"/replace_headers.txt | tr '=' '_') \\\n",
    "\"$2\"/comp_het/gfmix/concat_107t_93g_LG-MEOW6020-G_new-sister-c_complex-mono_replaced.treefile\n",
    "\n",
    "perl ../../src/replace_tip_labels.pl \"$2\"/meow/concat_107t_93g_LG-MEOW6020-G_new-sister-h_complex-mono_2.treefile \\\n",
    "<(cat \"$1\"/replace_headers.txt | tr '=' '_') \\\n",
    "\"$2\"/comp_het/gfmix/concat_107t_93g_LG-MEOW6020-G_new-sister-h_complex-mono_replaced.treefile\n",
    "\n",
    "perl ../../src/replace_tip_labels.pl \"$2\"/meow/concat_107t_93g_LG-MEOW6020-G_new-sister-h_complex-non-mono.treefile \\\n",
    "<(cat \"$1\"/replace_headers.txt | tr '=' '_') \\\n",
    "\"$2\"/comp_het/gfmix/concat_107t_93g_LG-MEOW6020-G_new-sister-h_complex-non-mono_replaced.treefile\n",
    "\n",
    "perl ../../src/replace_tip_labels.pl \"$2\"/meow/concat_107t_93g_LG-MEOW6020-G_new-sister-c_complex-non-mono_2.treefile \\\n",
    "<(cat \"$1\"/replace_headers.txt | tr '=' '_') \\\n",
    "\"$2\"/comp_het/gfmix/concat_107t_93g_LG-MEOW6020-G_new-sister-c_complex-non-mono_replaced.treefile"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Also copy the IQTree files."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {},
   "outputs": [],
   "source": [
    "%%bash -s \"$TREE_DIR\"\n",
    "\n",
    "## replace tip labels of trees inferred with LG+MEOW80+G model\n",
    "cp \"$1\"/meow/concat_107t_93g_LG-MEOW6020-G_new-sister-h-c_complex-mono.iqtree \\\n",
    "\"$1\"/comp_het/gfmix/.\n",
    "\n",
    "cp \"$1\"/meow/concat_107t_93g_LG-MEOW6020-G_new-sister-h-c_complex-non-mono.iqtree \\\n",
    "\"$1\"/comp_het/gfmix/.\n",
    "\n",
    "cp \"$1\"/meow/concat_107t_93g_LG-MEOW6020-G_new-sister-c_complex-mono.iqtree \\\n",
    "\"$1\"/comp_het/gfmix/.\n",
    "\n",
    "cp \"$1\"/meow/concat_107t_93g_LG-MEOW6020-G_new-sister-h_complex-mono_2.iqtree \\\n",
    "\"$1\"/comp_het/gfmix/.\n",
    "\n",
    "cp \"$1\"/meow/concat_107t_93g_LG-MEOW6020-G_new-sister-h_complex-non-mono.iqtree \\\n",
    "\"$1\"/comp_het/gfmix/.\n",
    "\n",
    "cp \"$1\"/meow/concat_107t_93g_LG-MEOW6020-G_new-sister-c_complex-non-mono_2.iqtree \\\n",
    "\"$1\"/comp_het/gfmix/."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### 15.1 Binomial test to determine groups of enriched/depleted amino acids"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "We now estimate the groups of amino acids that are depleted and enriched in different lineages from our data. We don't have any a priori knowledge of how groups might differ (e.g. as is known in the case of halophiles vs non-halophiles), nor do we know which amino acids might be involved in composition bias. \n",
    "\n",
    "We now use an R script (from Williamson et al 2024, kindly provided by Charley McCarthey) to:\n",
    "\n",
    "1) reads in the alignment file and tabulate the counts of all 20 amino acids,\n",
    "2) perform a chi-squared test on that table, and use the residuals from that test to build a hierarchically-clustered tree of taxa based on UPGMA,\n",
    "3) cut the tree into two groups, and use them to run the binomial test for our alignment (the groups are also printed to screen),\n",
    "4) print the results for each amino acid are to screen or to an output file,\n",
    "5) present the results are presented as a three column table of amino acid, Z-score and GFmix class assignment based on Z-score."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [],
   "source": [
    "## Directory for fasta file\n",
    "CONCAT_DIR = paths_dict[\"ANALYSIS_DATA\"][\"CONCAT_GENE_ANALYSIS\"][\"SUBSET\"][\"ALIGNMENTS\"][\"V14\"][\"CONCAT\"]\n",
    "\n",
    "## Directory for tree\n",
    "TREE_DIR = paths_dict[\"ANALYSIS_DATA\"][\"CONCAT_GENE_ANALYSIS\"][\"SUBSET\"][\"TREES\"][\"V14\"]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "%%bash -s \"$CONCAT_DIR\" \"$TREE_DIR\"\n",
    "\n",
    "Rscript ../../src/quickBinomial.R -aln \"$1\"/concat_107t_93g_prequal_ginsi_bmge_replaced.phylip -fmt phylip -out \"$2\"/comp_het/gfmix/bionimal_results.txt"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "The two groups of taxa were found to be Cyanidiococcus-yangmingshanensis-accession_NC-051883 and Cyanidioschyzon-merolae-accession_NC-004799, vs. everything else. \n",
    "\n",
    "The groups of amino acids that are enriched/depleted in the two groups of lineages are as follows:\n",
    "\n",
    "G-class: I, N, F, K, T, D, S, E   \n",
    "F-class: R, C, V, W, H, M, A, Q"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### 15.2 Run gfmix"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Run the gfmix_custombins model! (An updated version should be available via the official release on Ed Susko's website. The version used for our analyses is provided in the `src` folder)."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "%%bash -s \"$TREE_DIR\"\n",
    "\n",
    "sbatch ../../uppmax_scripts/script_bin/job_gfmix.sh \"$1\"/comp_het/gfmix/concat_107t_93g_prequal_ginsi_bmge_replaced.phylip \\\n",
    "    \"$1\"/comp_het/gfmix/concat_107t_93g_LG-MEOW6020-G_new-sister-c_complex-mono_replaced.treefile \\\n",
    "    \"$1\"/comp_het/gfmix/concat_107t_93g_LG-MEOW6020-G_new-sister-c_complex-mono.iqtree \\\n",
    "    \"$1\"/comp_het/gfmix/meow_60_20.aafreq.dat \\\n",
    "    \"$1\"/comp_het/gfmix/glauco.rootfile\n",
    "sleep 1\n",
    "\n",
    "sbatch ../../uppmax_scripts/script_bin/job_gfmix.sh \"$1\"/comp_het/gfmix/concat_107t_93g_prequal_ginsi_bmge_replaced.phylip \\\n",
    "    \"$1\"/comp_het/gfmix/concat_107t_93g_LG-MEOW6020-G_new-sister-c_complex-non-mono_replaced.treefile \\\n",
    "    \"$1\"/comp_het/gfmix/concat_107t_93g_LG-MEOW6020-G_new-sister-c_complex-non-mono_2.iqtree \\\n",
    "    \"$1\"/comp_het/gfmix/meow_60_20.aafreq.dat \\\n",
    "    \"$1\"/comp_het/gfmix/glauco.rootfile\n",
    "sleep 1\n",
    "\n",
    "sbatch ../../uppmax_scripts/script_bin/job_gfmix.sh \"$1\"/comp_het/gfmix/concat_107t_93g_prequal_ginsi_bmge_replaced.phylip \\\n",
    "    \"$1\"/comp_het/gfmix/concat_107t_93g_LG-MEOW6020-G_new-sister-h-c_complex-mono_replaced.treefile \\\n",
    "    \"$1\"/comp_het/gfmix/concat_107t_93g_LG-MEOW6020-G_new-sister-h-c_complex-mono.iqtree \\\n",
    "    \"$1\"/comp_het/gfmix/meow_60_20.aafreq.dat \\\n",
    "    \"$1\"/comp_het/gfmix/glauco.rootfile\n",
    "sleep 1\n",
    "\n",
    "sbatch ../../uppmax_scripts/script_bin/job_gfmix.sh \"$1\"/comp_het/gfmix/concat_107t_93g_prequal_ginsi_bmge_replaced.phylip \\\n",
    "    \"$1\"/comp_het/gfmix/concat_107t_93g_LG-MEOW6020-G_new-sister-h-c_complex-non-mono_replaced.treefile \\\n",
    "    \"$1\"/comp_het/gfmix/concat_107t_93g_LG-MEOW6020-G_new-sister-h-c_complex-non-mono.iqtree \\\n",
    "    \"$1\"/comp_het/gfmix/meow_60_20.aafreq.dat \\\n",
    "    \"$1\"/comp_het/gfmix/glauco.rootfile\n",
    "sleep 1\n",
    "\n",
    "sbatch ../../uppmax_scripts/script_bin/job_gfmix.sh \"$1\"/comp_het/gfmix/concat_107t_93g_prequal_ginsi_bmge_replaced.phylip \\\n",
    "    \"$1\"/comp_het/gfmix/concat_107t_93g_LG-MEOW6020-G_new-sister-h_complex-mono_replaced.treefile \\\n",
    "    \"$1\"/comp_het/gfmix/concat_107t_93g_LG-MEOW6020-G_new-sister-h_complex-mono_2.iqtree \\\n",
    "    \"$1\"/comp_het/gfmix/meow_60_20.aafreq.dat \\\n",
    "    \"$1\"/comp_het/gfmix/glauco.rootfile\n",
    "sleep 1\n",
    "\n",
    "sbatch ../../uppmax_scripts/script_bin/job_gfmix.sh \"$1\"/comp_het/gfmix/concat_107t_93g_prequal_ginsi_bmge_replaced.phylip \\\n",
    "    \"$1\"/comp_het/gfmix/concat_107t_93g_LG-MEOW6020-G_new-sister-h_complex-non-mono_replaced.treefile \\\n",
    "    \"$1\"/comp_het/gfmix/concat_107t_93g_LG-MEOW6020-G_new-sister-h_complex-non-mono.iqtree \\\n",
    "    \"$1\"/comp_het/gfmix/meow_60_20.aafreq.dat \\\n",
    "    \"$1\"/comp_het/gfmix/glauco.rootfile\n",
    "sleep 1\n",
    "\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## 16. Modelling heterotachy with the GHOST model\n",
    "\n",
    "All models used so far assume that the substitution rate for each site is constant across all lineages. However, we know that substitution rates can vary across lineages. To accomodate rate variation across lineages, we use the GHOST model (General Heterogeneous evolution On a Single Topology; [Crotty et al 2019](https://doi.org/10.1093/sysbio/syz051)). GHOST is a mixture model comprised of several site classes, each having a separate set of model parameters and edge lengths on the same tree topology.\n",
    "\n",
    "We follow [Williamson et al 2024](https://doi.org/10.1101/2024.09.04.611237) and calculate the likelihoods of the six topologies under the GHOST model. We will evaluate each topology with 4, 6, and 8 linked heterotachy classes (LG+MEOW80+H4/6/8)."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [],
   "source": [
    "## Directory for fasta file\n",
    "CONCAT_DIR = paths_dict[\"ANALYSIS_DATA\"][\"CONCAT_GENE_ANALYSIS\"][\"SUBSET\"][\"ALIGNMENTS\"][\"V14\"][\"CONCAT\"]\n",
    "\n",
    "## Directory for tree\n",
    "TREE_DIR = paths_dict[\"ANALYSIS_DATA\"][\"CONCAT_GENE_ANALYSIS\"][\"SUBSET\"][\"TREES\"][\"V14\"]"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Note that the IQTree commands include a \"-optlen BFGS\" flag. I asked Kelsey Williamson about it, and she relayed her communication with Minh:\n",
    "\n",
    "\"This is only for technical reason, if you want to know: the EM (expectation maximization) algorithm is used to estimate branch lengths for +H mixture model. Currently it does not work in the combination of two mixture models +C60 and +H model. The BFGS algorithm is an alternative, and it works for this combination.\n",
    "\n",
    "IQ-TREE could have switched internally, but we intentionally print this message to remind us that we can implement the EM algorithm for this case.”\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### 16.1. Four rate categories"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Estimate likelihoods with 4 linked heterotachy classes."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "%%bash -s \"$CONCAT_DIR\" \"$TREE_DIR\"\n",
    "\n",
    "sbatch ../../uppmax_scripts/script_bin/job_ghost.sh \"$1\"/concat_107t_93g_prequal_ginsi_bmge.fasta \"$2\"/ghost/ghost_4_new-sister-c_complex-mono \"$2\"/meow/meow.nex \"$2\"/ghost/concat_107t_93g_LG-MEOW6020-G_new-sister-c_complex-mono.treefile\n",
    "sleep 1\n",
    "\n",
    "sbatch ../../uppmax_scripts/script_bin/job_ghost.sh \"$1\"/concat_107t_93g_prequal_ginsi_bmge.fasta \"$2\"/ghost/ghost_4_new-sister-c_complex-non-mono \"$2\"/meow/meow.nex \"$2\"/ghost/concat_107t_93g_LG-MEOW6020-G_new-sister-c_complex-non-mono_2.treefile\n",
    "sleep 1\n",
    "\n",
    "sbatch ../../uppmax_scripts/script_bin/job_ghost.sh \"$1\"/concat_107t_93g_prequal_ginsi_bmge.fasta \"$2\"/ghost/ghost_4_new-sister-h-c_complex-mono \"$2\"/meow/meow.nex \"$2\"/ghost/concat_107t_93g_LG-MEOW6020-G_new-sister-h-c_complex-mono.treefile\n",
    "sleep 1\n",
    "\n",
    "sbatch ../../uppmax_scripts/script_bin/job_ghost.sh \"$1\"/concat_107t_93g_prequal_ginsi_bmge.fasta \"$2\"/ghost/ghost_4_new-sister-h-c_complex-non-mono \"$2\"/meow/meow.nex \"$2\"/ghost/concat_107t_93g_LG-MEOW6020-G_new-sister-h-c_complex-non-mono.treefile\n",
    "sleep 1\n",
    "\n",
    "sbatch ../../uppmax_scripts/script_bin/job_ghost.sh \"$1\"/concat_107t_93g_prequal_ginsi_bmge.fasta \"$2\"/ghost/ghost_4_new-sister-h_complex-mono \"$2\"/meow/meow.nex \"$2\"/ghost/concat_107t_93g_LG-MEOW6020-G_new-sister-h_complex-mono_2.treefile\n",
    "sleep 1\n",
    "\n",
    "sbatch ../../uppmax_scripts/script_bin/job_ghost.sh \"$1\"/concat_107t_93g_prequal_ginsi_bmge.fasta \"$2\"/ghost/ghost_4_new-sister-h_complex-non-mono \"$2\"/meow/meow.nex \"$2\"/ghost/concat_107t_93g_LG-MEOW6020-G_new-sister-h_complex-non-mono.treefile\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### 16.2. Six rate categories"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Estimate likelihoods with 6 linked heterotachy classes."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "%%bash -s \"$CONCAT_DIR\" \"$TREE_DIR\"\n",
    "\n",
    "sbatch ../../uppmax_scripts/script_bin/job_ghost.sh \"$1\"/concat_107t_93g_prequal_ginsi_bmge.fasta \"$2\"/ghost/ghost_6_new-sister-c_complex-mono \"$2\"/meow/meow.nex \"$2\"/ghost/concat_107t_93g_LG-MEOW6020-G_new-sister-c_complex-mono.treefile\n",
    "sleep 1\n",
    "\n",
    "sbatch ../../uppmax_scripts/script_bin/job_ghost.sh \"$1\"/concat_107t_93g_prequal_ginsi_bmge.fasta \"$2\"/ghost/ghost_6_new-sister-c_complex-non-mono \"$2\"/meow/meow.nex \"$2\"/ghost/concat_107t_93g_LG-MEOW6020-G_new-sister-c_complex-non-mono_2.treefile\n",
    "sleep 1\n",
    "\n",
    "sbatch ../../uppmax_scripts/script_bin/job_ghost.sh \"$1\"/concat_107t_93g_prequal_ginsi_bmge.fasta \"$2\"/ghost/ghost_6_new-sister-h-c_complex-mono \"$2\"/meow/meow.nex \"$2\"/ghost/concat_107t_93g_LG-MEOW6020-G_new-sister-h-c_complex-mono.treefile\n",
    "sleep 1\n",
    "\n",
    "sbatch ../../uppmax_scripts/script_bin/job_ghost.sh \"$1\"/concat_107t_93g_prequal_ginsi_bmge.fasta \"$2\"/ghost/ghost_6_new-sister-h-c_complex-non-mono \"$2\"/meow/meow.nex \"$2\"/ghost/concat_107t_93g_LG-MEOW6020-G_new-sister-h-c_complex-non-mono.treefile\n",
    "sleep 1\n",
    "\n",
    "sbatch ../../uppmax_scripts/script_bin/job_ghost.sh \"$1\"/concat_107t_93g_prequal_ginsi_bmge.fasta \"$2\"/ghost/ghost_6_new-sister-h_complex-mono \"$2\"/meow/meow.nex \"$2\"/ghost/concat_107t_93g_LG-MEOW6020-G_new-sister-h_complex-mono_2.treefile\n",
    "sleep 1\n",
    "\n",
    "sbatch ../../uppmax_scripts/script_bin/job_ghost.sh \"$1\"/concat_107t_93g_prequal_ginsi_bmge.fasta \"$2\"/ghost/ghost_6_new-sister-h_complex-non-mono \"$2\"/meow/meow.nex \"$2\"/ghost/concat_107t_93g_LG-MEOW6020-G_new-sister-h_complex-non-mono.treefile\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### 16.3. Eight rate categories"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "And now 8 classes."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "%%bash -s \"$CONCAT_DIR\" \"$TREE_DIR\"\n",
    "\n",
    "sbatch ../../uppmax_scripts/script_bin/job_ghost.sh \"$1\"/concat_107t_93g_prequal_ginsi_bmge.fasta \"$2\"/ghost/ghost_8_new-sister-c_complex-mono \"$2\"/meow/meow.nex \"$2\"/ghost/concat_107t_93g_LG-MEOW6020-G_new-sister-c_complex-mono.treefile\n",
    "sleep 1\n",
    "\n",
    "sbatch ../../uppmax_scripts/script_bin/job_ghost.sh \"$1\"/concat_107t_93g_prequal_ginsi_bmge.fasta \"$2\"/ghost/ghost_8_new-sister-c_complex-non-mono \"$2\"/meow/meow.nex \"$2\"/ghost/concat_107t_93g_LG-MEOW6020-G_new-sister-c_complex-non-mono_2.treefile\n",
    "sleep 1\n",
    "\n",
    "sbatch ../../uppmax_scripts/script_bin/job_ghost.sh \"$1\"/concat_107t_93g_prequal_ginsi_bmge.fasta \"$2\"/ghost/ghost_8_new-sister-h-c_complex-mono \"$2\"/meow/meow.nex \"$2\"/ghost/concat_107t_93g_LG-MEOW6020-G_new-sister-h-c_complex-mono.treefile\n",
    "sleep 1\n",
    "\n",
    "sbatch ../../uppmax_scripts/script_bin/job_ghost.sh \"$1\"/concat_107t_93g_prequal_ginsi_bmge.fasta \"$2\"/ghost/ghost_8_new-sister-h-c_complex-non-mono \"$2\"/meow/meow.nex \"$2\"/ghost/concat_107t_93g_LG-MEOW6020-G_new-sister-h-c_complex-non-mono.treefile\n",
    "sleep 1\n",
    "\n",
    "sbatch ../../uppmax_scripts/script_bin/job_ghost.sh \"$1\"/concat_107t_93g_prequal_ginsi_bmge.fasta \"$2\"/ghost/ghost_8_new-sister-h_complex-mono \"$2\"/meow/meow.nex \"$2\"/ghost/concat_107t_93g_LG-MEOW6020-G_new-sister-h_complex-mono_2.treefile\n",
    "sleep 1\n",
    "\n",
    "sbatch ../../uppmax_scripts/script_bin/job_ghost.sh \"$1\"/concat_107t_93g_prequal_ginsi_bmge.fasta \"$2\"/ghost/ghost_8_new-sister-h_complex-non-mono \"$2\"/meow/meow.nex \"$2\"/ghost/concat_107t_93g_LG-MEOW6020-G_new-sister-h_complex-non-mono.treefile\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### 16.4. Model test"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "But which GHOST model fits the best?"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "%%bash -s \"$CONCAT_DIR\" \"$TREE_DIR\"\n",
    "\n",
    "sbatch ../../uppmax_scripts/script_bin/job_ghost.sh \"$1\"/concat_107t_93g_prequal_ginsi_bmge.fasta \"$2\"/ghost/ghost_model_test \"$2\"/meow/meow.nex"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### 16.5. AU test"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "GHOST with 8 categories fits the best. Let's do an AU test of the trees. We concatenate the 6 different trees to test."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [],
   "source": [
    "%%bash -s \"$TREE_DIR\"\n",
    "\n",
    "cat \"$1\"/ghost/ghost_8_new-sister-c_complex-mono.treefile \\\n",
    "    \"$1\"/ghost/ghost_8_new-sister-c_complex-non-mono.treefile \\\n",
    "    \"$1\"/ghost/ghost_8_new-sister-h-c_complex-mono.treefile \\\n",
    "    \"$1\"/ghost/ghost_8_new-sister-h-c_complex-non-mono.treefile \\\n",
    "    \"$1\"/ghost/ghost_8_new-sister-h_complex-mono.treefile \\\n",
    "    \"$1\"/ghost/ghost_8_new-sister-h_complex-non-mono.treefile \\\n",
    "    > \"$1\"/ghost/ghost_8_au_test.trees"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "We will use the best scoring tree (ghost_8_new-sister-h-c_complex-mono.treefile) to estimate parameters for the topology tests."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "%%bash -s \"$CONCAT_DIR\" \"$TREE_DIR\"\n",
    "\n",
    "sbatch ../../uppmax_scripts/script_bin/job_topology_test.sh \"$1\"/concat_107t_93g_prequal_ginsi_bmge.fasta \"$2\"/ghost/ghost_8_AU_test \"$2\"/meow/meow.nex \"$2\"/ghost/ghost_8_new-sister-h-c_complex-mono.treefile \"$2\"/ghost/ghost_8_au_test.trees"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## 17. Fast taxa removal\n",
    "\n",
    "We now test the impact of fast evolving taxa on the resulting topology, by removing the fastest evolving taxa. We use a custom python script to identify the fastest evolving taxa (largest root to tip distances)."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "metadata": {},
   "outputs": [],
   "source": [
    "## Directory for tree\n",
    "TREE_DIR = paths_dict[\"ANALYSIS_DATA\"][\"CONCAT_GENE_ANALYSIS\"][\"SUBSET\"][\"TREES\"][\"V14\"]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 20,
   "metadata": {},
   "outputs": [],
   "source": [
    "%%bash -s \"$TREE_DIR\"\n",
    "\n",
    "python ../../src/root_to_tip_distances.py \"$1\"/meow/concat_107t_93g_prequal_ginsi_bmge_LG-MEOW6020-G.treefile Glaucocystophyceae \"$1\"/meow/bls_result.txt"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "We decided to remove the 10 taxa with the longest root to tip distances."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 21,
   "metadata": {},
   "outputs": [],
   "source": [
    "## Directory for fasta file\n",
    "CONCAT_DIR = paths_dict[\"ANALYSIS_DATA\"][\"CONCAT_GENE_ANALYSIS\"][\"SUBSET\"][\"ALIGNMENTS\"][\"V14\"][\"CONCAT\"]\n",
    "\n",
    "## Directory for tree\n",
    "TREE_DIR = paths_dict[\"ANALYSIS_DATA\"][\"CONCAT_GENE_ANALYSIS\"][\"SUBSET\"][\"TREES\"][\"V14\"]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "%%bash -s \"$CONCAT_DIR\" \"$TREE_DIR\"\n",
    "\n",
    "cat \"$2\"/meow/bls_result.txt | tr ' ' '\\t' | sort -nk2 | tail | cut -f1 | sed -E 's/taxo_/taxo=/' | sed -E 's/mag_/mag=/' | sed -E 's/accession_/accession=/' | sed -E 's/taxonomy_/taxonomy=/' > \"$1\"/lb_taxa.list\n",
    "\n",
    "seqkit grep -f \"$1\"/lb_taxa.list \"$1\"/concat_107t_93g_prequal_ginsi_bmge.fasta -v > \"$1\"/concat_97t_93g_prequal_ginsi_bmge_nLB.fasta"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "A quick test (not shown) showed that the nRCVFV value was now 0.00307385430785046."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Run tree with the newly generated alignment with the MEOW(60,20) model."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "%%bash -s \"$CONCAT_DIR\" \"$TREE_DIR\"\n",
    "\n",
    "sbatch ../../uppmax_scripts/script_bin/job_iqtree.sh \"$1\"/concat_97t_93g_prequal_ginsi_bmge_nLB.fasta \"$2\"/meow/concat_97t_93g_prequal_ginsi_bmge_LG-MEOW6020-G_nLB \"$2\"/meow/meow.nex"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## 18. Recoding amino acids (SR4 recoding)\n",
    "\n",
    "The results of the GFmix model indicate that our phylogenetic inference is impacted by compositional heterogeneity. We now try recoding our dataset to reduce the impact of compositional heterogeneity. We opted for SR4 recoding. This was done using the PhyloFisher script `aa_recoder.py` ([Tice et al 2021](https://doi.org/10.1371/journal.pbio.3001365)).  "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [],
   "source": [
    "## Directory for fasta file\n",
    "CONCAT_DIR = paths_dict[\"ANALYSIS_DATA\"][\"CONCAT_GENE_ANALYSIS\"][\"SUBSET\"][\"ALIGNMENTS\"][\"V14\"][\"CONCAT\"]\n",
    "\n",
    "## Directory for tree\n",
    "TREE_DIR = paths_dict[\"ANALYSIS_DATA\"][\"CONCAT_GENE_ANALYSIS\"][\"SUBSET\"][\"TREES\"][\"V14\"]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "aa_recoder.py -i concat_107t_93g_prequal_ginsi_bmge.fasta -o concat_107t_93g_prequal_ginsi_bmge.SR4 -re SR4"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "The nRCFV values is now at its lowest with a value of 0.000975221142460481."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "We can now set up two trees! A Bayesian inference under CAT-GTR, and a Maximum Likelihood tree using a recoded version of the C60 model obtained from: https://github.com/xgrau/recoded-mixture-models/blob/master/recoded_models/xmC60SR4.nex\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### 18.1 CAT-GTR tree"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Convert alignment to phylip format first."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "%%bash -s \"$CONCAT_DIR\"\n",
    "\n",
    "perl ../../src/fasta2phylip.pl -f \"$1\"/concat_107t_93g_prequal_ginsi_bmge.SR4.fas -o \"$1\"/concat_107t_93g_prequal_ginsi_bmge.SR4.phy\n",
    "\n",
    "cat \"$1\"/concat_107t_93g_prequal_ginsi_bmge.SR4.phy | tr '=' '_' > phylip\n",
    "mv phylip \"$1\"/concat_107t_93g_prequal_ginsi_bmge.SR4.phy"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Run three chains of PhyloBayes."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "%%bash -s \"$CONCAT_DIR\" \"$TREE_DIR\"\n",
    "\n",
    "sbatch ../../uppmax_scripts/script_bin/job_pb.sh \"$1\"/concat_107t_93g_prequal_ginsi_bmge.SR4.phy \"$2\"/recoded_sr4/concat_107t_93g_prequal_ginsi_bmge.SR4_catgtr_chain1\n",
    "sleep 1\n",
    "\n",
    "sbatch ../../uppmax_scripts/script_bin/job_pb.sh \"$1\"/concat_107t_93g_prequal_ginsi_bmge.SR4.phy \"$2\"/recoded_sr4/concat_107t_93g_prequal_ginsi_bmge.SR4_catgtr_chain2\n",
    "sleep 1\n",
    "\n",
    "sbatch ../../uppmax_scripts/script_bin/job_pb.sh \"$1\"/concat_107t_93g_prequal_ginsi_bmge.SR4.phy \"$2\"/recoded_sr4/concat_107t_93g_prequal_ginsi_bmge.SR4_catgtr_chain3\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "We checked convergence with bpcomp. "
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### 18.2. Constrained ML trees\n",
    "\n",
    "We now infer constrained trees with the recoded C60 model in a ML framework (GTR+C60+G)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [],
   "source": [
    "## Directory for fasta file\n",
    "CONCAT_DIR = paths_dict[\"ANALYSIS_DATA\"][\"CONCAT_GENE_ANALYSIS\"][\"SUBSET\"][\"ALIGNMENTS\"][\"V14\"][\"CONCAT\"]\n",
    "\n",
    "## Directory for tree\n",
    "TREE_DIR = paths_dict[\"ANALYSIS_DATA\"][\"CONCAT_GENE_ANALYSIS\"][\"SUBSET\"][\"TREES\"][\"V14\"]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "%%bash -s \"$CONCAT_DIR\" \"$TREE_DIR\"\n",
    "\n",
    "sbatch ../../uppmax_scripts/script_bin/job_iqtree.sh \"$1\"/concat_107t_93g_prequal_ginsi_bmge.SR4.fas \"$2\"/recoded_sr4/concat_107t_93g_sr4_LG-C60-G_new-sister-c_complex-mono \"$2\"/recoded_sr4/xmC60SR4.nex \"$2\"/constraints/new-sister-c_complex-mono.tre\n",
    "sleep 1\n",
    "\n",
    "sbatch ../../uppmax_scripts/script_bin/job_iqtree.sh \"$1\"/concat_107t_93g_prequal_ginsi_bmge.SR4.fas \"$2\"/recoded_sr4/concat_107t_93g_sr4_LG-C60-G_new-sister-c_complex-non-mono \"$2\"/recoded_sr4/xmC60SR4.nex \"$2\"/constraints/new-sister-c_complex-non-mono.tre\n",
    "sleep 1\n",
    "\n",
    "sbatch ../../uppmax_scripts/script_bin/job_iqtree.sh \"$1\"/concat_107t_93g_prequal_ginsi_bmge.SR4.fas \"$2\"/recoded_sr4/concat_107t_93g_sr4_LG-C60-G_new-sister-h_complex-mono \"$2\"/recoded_sr4/xmC60SR4.nex \"$2\"/constraints/new-sister-h_complex-mono.tre\n",
    "sleep 1\n",
    "\n",
    "sbatch ../../uppmax_scripts/script_bin/job_iqtree.sh \"$1\"/concat_107t_93g_prequal_ginsi_bmge.SR4.fas \"$2\"/recoded_sr4/concat_107t_93g_sr4_LG-C60-G_new-sister-h_complex-non-mono \"$2\"/recoded_sr4/xmC60SR4.nex \"$2\"/constraints/new-sister-h_complex-non-mono.tre\n",
    "sleep 1\n",
    "\n",
    "sbatch ../../uppmax_scripts/script_bin/job_iqtree.sh \"$1\"/concat_107t_93g_prequal_ginsi_bmge.SR4.fas \"$2\"/recoded_sr4/concat_107t_93g_sr4_LG-C60-G_new-sister-h-c_complex-mono \"$2\"/recoded_sr4/xmC60SR4.nex \"$2\"/constraints/new-sister-h-c_complex-mono.tre\n",
    "sleep 1\n",
    "\n",
    "sbatch ../../uppmax_scripts/script_bin/job_iqtree.sh \"$1\"/concat_107t_93g_prequal_ginsi_bmge.SR4.fas \"$2\"/recoded_sr4/concat_107t_93g_sr4_LG-C60-G_new-sister-h-c_complex-non-mono \"$2\"/recoded_sr4/xmC60SR4.nex \"$2\"/constraints/new-sister-h-c_complex-non-mono.tre"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## 19. Check the effect of alternative gene sets\n",
    "\n",
    "Previous studies frequently used the \"slow evolving\" and \"fast evolving\" gene sets as defined by [Janouskovec et al 2010](https://doi.org/10.1073/pnas.1003335107). These were the gene sets used for our preliminary analyses in fact. Here, we run phylogenies of this 107 taxon set based on the 32 and 66 gene set as requested by a reviewer. Note that while Janouskovec et al have 34 (not 32) genes in the slow set and 68 genes in the fast set. The discrepancy is because we excluded ascF (too few taxa), and psbH (slightly weird gene tree).\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### 19.1 32 genes (slow evolving genes)\n",
    "\n",
    "We don't need to align or trim the sequences as we are taking just a subset of the genes. "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Read_genes\n",
    "GENE_LIST = paths_dict['DATABASES']['GENE_LISTS']['GENES_32']\n",
    "\n",
    "# Directory for aligned and trimmed fasta files\n",
    "BMGE_DIR = paths_dict['ANALYSIS_DATA'][\"CONCAT_GENE_ANALYSIS\"][\"SUBSET\"][\"ALIGNMENTS\"][\"V14\"][\"BMGE\"]\n",
    "\n",
    "## Output directory for fasta files\n",
    "CONCAT_DIR = paths_dict[\"ANALYSIS_DATA\"][\"CONCAT_GENE_ANALYSIS\"][\"SUBSET\"][\"ALIGNMENTS\"][\"V14\"][\"CONCAT\"]"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Copy the relevant files to a directory."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 21,
   "metadata": {},
   "outputs": [],
   "source": [
    "%%bash -s \"$GENE_LIST\" \"$BMGE_DIR\"\n",
    "\n",
    "cat \"$1\" | while read line\n",
    "    do cp \"$2\"/\"$line\".fasta \"$2\"/32g/.\n",
    "done\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Concatenate the genes!"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 22,
   "metadata": {},
   "outputs": [],
   "source": [
    "%%bash -s \"$BMGE_DIR\" \"$CONCAT_DIR\"\n",
    "\n",
    "files=(\"$1\"/32g/*fasta)\n",
    "\n",
    "perl /home/mahja/beta-Cyclocitral/src/cat_fasta.pl -f \"${files[@]}\" > \"$2\"/concat_107t_32g_prequal_ginsi_bmge.fasta\n",
    "mv partitions.txt \"$2\"/partitions_prequal_ginsi_bmge_32g.txt"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "The alignment has 107 taxa, 32 genes, and 7,921 sites. \n",
    "\n",
    "I manually edited the header the names slightly. \n",
    "- Removed the * after New\n",
    "- Changed ZHAN22 to Haptophyte from New"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### 20.1.1. Run cpREV+C60+G tree\n",
    "\n",
    "We ran a tree with the cpREV+C60+G model. "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "metadata": {},
   "outputs": [],
   "source": [
    "## Directory for fasta file\n",
    "CONCAT_DIR = paths_dict[\"ANALYSIS_DATA\"][\"CONCAT_GENE_ANALYSIS\"][\"SUBSET\"][\"ALIGNMENTS\"][\"V14\"][\"CONCAT\"]\n",
    "\n",
    "## Output directory for tree\n",
    "TREE_DIR = paths_dict[\"ANALYSIS_DATA\"][\"CONCAT_GENE_ANALYSIS\"][\"SUBSET\"][\"TREES\"][\"V14\"]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "%%bash -s \"$CONCAT_DIR\" \"$TREE_DIR\"\n",
    "\n",
    "sbatch ../../uppmax_scripts/script_bin/job_iqtree.sh \"$1\"/concat_107t_32g_prequal_ginsi_bmge.fasta \"$2\"/concat_107t_32g_prequal_ginsi_bmge_cpREV-C60-G"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### 19.1.2. Run cat-gtr tree\n",
    "Convert alignment to phylip format. "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "%%bash -s \"$CONCAT_DIR\"\n",
    "\n",
    "perl ../../src/fasta2phylip.pl -f \"$1\"/concat_107t_32g_prequal_ginsi_bmge.fasta -o \"$1\"/concat_107t_32g_prequal_ginsi_bmge.phy\n",
    "\n",
    "cat \"$1\"/concat_107t_32g_prequal_ginsi_bmge.phy | tr '=' '_' > phylip\n",
    "mv phylip \"$1\"/concat_107t_32g_prequal_ginsi_bmge.phy"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Run three chains of PhyloBayes. "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "%%bash -s \"$CONCAT_DIR\" \"$TREE_DIR\"\n",
    "\n",
    "sbatch ../../uppmax_scripts/script_bin/job_pb.sh \"$1\"/concat_107t_32g_prequal_ginsi_bmge.phy \"$2\"/gene_subsets/concat_107t_32g_prequal_ginsi_bmge_catgtr_chain1\n",
    "sleep 1\n",
    "\n",
    "sbatch ../../uppmax_scripts/script_bin/job_pb.sh \"$1\"/concat_107t_32g_prequal_ginsi_bmge.phy \"$2\"/gene_subsets/concat_107t_32g_prequal_ginsi_bmge_catgtr_chain2\n",
    "sleep 1\n",
    "\n",
    "sbatch ../../uppmax_scripts/script_bin/job_pb.sh \"$1\"/concat_107t_32g_prequal_ginsi_bmge.phy \"$2\"/gene_subsets/concat_107t_32g_prequal_ginsi_bmge_catgtr_chain3\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "%%bash -s \"$TREE_DIR\"\n",
    "\n",
    "sbatch ../../uppmax_scripts/script_bin/job_pb.sh \"$1\"/gene_subsets/concat_107t_32g_prequal_ginsi_bmge_catgtr_chain1\n",
    "sleep 1\n",
    "\n",
    "sbatch ../../uppmax_scripts/script_bin/job_pb.sh \"$1\"/gene_subsets/concat_107t_32g_prequal_ginsi_bmge_catgtr_chain2\n",
    "sleep 1\n",
    "\n",
    "sbatch ../../uppmax_scripts/script_bin/job_pb.sh \"$1\"/gene_subsets/concat_107t_32g_prequal_ginsi_bmge_catgtr_chain3\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### 19.2 66 genes (slow+fast evolving genes)\n",
    "\n",
    "We don't need to align or trim the sequences as we are taking just a subset of the genes. "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 27,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Read_genes\n",
    "GENE_LIST = paths_dict['DATABASES']['GENE_LISTS']['GENES_66']\n",
    "\n",
    "# Directory for aligned and trimmed fasta files\n",
    "BMGE_DIR = paths_dict['ANALYSIS_DATA'][\"CONCAT_GENE_ANALYSIS\"][\"SUBSET\"][\"ALIGNMENTS\"][\"V14\"][\"BMGE\"]\n",
    "\n",
    "## Output directory for fasta files\n",
    "CONCAT_DIR = paths_dict[\"ANALYSIS_DATA\"][\"CONCAT_GENE_ANALYSIS\"][\"SUBSET\"][\"ALIGNMENTS\"][\"V14\"][\"CONCAT\"]"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Copy files to relevant directory."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 29,
   "metadata": {},
   "outputs": [],
   "source": [
    "%%bash -s \"$GENE_LIST\" \"$BMGE_DIR\"\n",
    "\n",
    "cat \"$1\" | while read line\n",
    "    do cp \"$2\"/\"$line\".fasta \"$2\"/66g/.\n",
    "done"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Concatenate the genes!"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 30,
   "metadata": {},
   "outputs": [],
   "source": [
    "%%bash -s \"$BMGE_DIR\" \"$CONCAT_DIR\"\n",
    "\n",
    "files=(\"$1\"/66g/*fasta)\n",
    "\n",
    "perl /home/mahja/beta-Cyclocitral/src/cat_fasta.pl -f \"${files[@]}\" > \"$2\"/concat_107t_66g_prequal_ginsi_bmge.fasta\n",
    "mv partitions.txt \"$2\"/partitions_prequal_ginsi_bmge_66g.txt"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "The alignment has 107 taxa, 32 genes, and 16,438 sites. \n",
    "\n",
    "I manually edited the header the names slightly. \n",
    "- Removed the * after New\n",
    "- Changed ZHAN22 to Haptophyte from New"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### 19.2.1. Run cpREV+C60+G tree\n",
    "\n",
    "We ran a tree with the cpREV+C60+G model. "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [],
   "source": [
    "## Directory for fasta file\n",
    "CONCAT_DIR = paths_dict[\"ANALYSIS_DATA\"][\"CONCAT_GENE_ANALYSIS\"][\"SUBSET\"][\"ALIGNMENTS\"][\"V14\"][\"CONCAT\"]\n",
    "\n",
    "## Output directory for tree\n",
    "TREE_DIR = paths_dict[\"ANALYSIS_DATA\"][\"CONCAT_GENE_ANALYSIS\"][\"SUBSET\"][\"TREES\"][\"V14\"]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "%%bash -s \"$CONCAT_DIR\" \"$TREE_DIR\"\n",
    "\n",
    "sbatch ../../uppmax_scripts/script_bin/job_iqtree.sh \"$1\"/concat_107t_66g_prequal_ginsi_bmge.fasta \"$2\"/gene_subsets/concat_107t_66g_prequal_ginsi_bmge_cpREV-C60-G"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### 19.2.2. Run cat-gtr tree\n",
    "Convert alignment to phylip format. "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "%%bash -s \"$CONCAT_DIR\"\n",
    "\n",
    "perl ../../src/fasta2phylip.pl -f \"$1\"/concat_107t_66g_prequal_ginsi_bmge.fasta -o \"$1\"/concat_107t_66g_prequal_ginsi_bmge.phy\n",
    "\n",
    "cat \"$1\"/concat_107t_66g_prequal_ginsi_bmge.phy | tr '=' '_' > phylip\n",
    "mv phylip \"$1\"/concat_107t_66g_prequal_ginsi_bmge.phy"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Run 3 chains of PhyloBayes."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "%%bash -s \"$CONCAT_DIR\" \"$TREE_DIR\"\n",
    "\n",
    "sbatch ../../uppmax_scripts/script_bin/job_pb.sh \"$1\"/concat_107t_66g_prequal_ginsi_bmge.phy \"$2\"/gene_subsets/concat_107t_66g_prequal_ginsi_bmge_catgtr_chain1\n",
    "sleep 1\n",
    "\n",
    "sbatch ../../uppmax_scripts/script_bin/job_pb.sh \"$1\"/concat_107t_66g_prequal_ginsi_bmge.phy \"$2\"/gene_subsets/concat_107t_66g_prequal_ginsi_bmge_catgtr_chain2\n",
    "sleep 1\n",
    "\n",
    "sbatch ../../uppmax_scripts/script_bin/job_pb.sh \"$1\"/concat_107t_66g_prequal_ginsi_bmge.phy \"$2\"/gene_subsets/concat_107t_66g_prequal_ginsi_bmge_catgtr_chain3\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "%%bash -s \"$TREE_DIR\"\n",
    "\n",
    "sbatch ../../uppmax_scripts/script_bin/job_pb.sh \"$1\"/gene_subsets/concat_107t_66g_prequal_ginsi_bmge_catgtr_chain1\n",
    "sleep 1\n",
    "\n",
    "sbatch ../../uppmax_scripts/script_bin/job_pb.sh \"$1\"/gene_subsets/concat_107t_66g_prequal_ginsi_bmge_catgtr_chain2\n",
    "sleep 1\n",
    "\n",
    "sbatch ../../uppmax_scripts/script_bin/job_pb.sh \"$1\"/gene_subsets/concat_107t_66g_prequal_ginsi_bmge_catgtr_chain3\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## References\n",
    "\n",
    "Katoh, K., & Standley, D. M. (2013). MAFFT multiple sequence alignment software version 7: improvements in performance and usability. Molecular biology and evolution, 30(4), 772-780. https://doi.org/10.1093/molbev/mst010\n",
    "\n",
    "Capella-Gutiérrez, S., Silla-Martínez, J. M., & Gabaldón, T. (2009). trimAl: a tool for automated alignment trimming in large-scale phylogenetic analyses. Bioinformatics, 25(15), 1972-1973. https://doi.org/10.1093/bioinformatics/btp348\n",
    "\n",
    "cat_fasta.pl script. Written by Martin Ryberg. Available at https://github.com/mr-y/my_bioinfPerls/blob/master/cat_fasta.pl. \n",
    "\n",
    "Susko, E. (2022). MAMMaL:(M)ultinomial (A)pproximate (M)ixture (Ma)ximum (L)ikelihood Accelerated Estimation of Frequency Classes in Site-heterogeneous Profile Mixture Models Version 1.1. 3 June 20, 2022. https://www.mathstat.dal.ca/~tsusko/doc/mammal.pdf \n",
    "\n",
    "Susko, E., Lincker, L., & Roger, A. J. (2018). Accelerated estimation of frequency classes in site-heterogeneous profile mixture models. Molecular Biology and Evolution, 35(5), 1266-1283. https://doi.org/10.1093/molbev/msy026\n",
    "\n",
    "Baños, H., Wong, T. K., Daneau, J., Susko, E., Minh, B. Q., Lanfear, R., Brown, M., Eme, L., & Roger, A. J. (2024). GTRpmix: A linked general-time reversible model for profile mixture models. bioRxiv, 2024-03.\n",
    "\n",
    "Zhou, H. Q., Ning, L. W., Zhang, H. X., & Guo, F. B. (2014). Analysis of the relationship between genomic GC content and patterns of base usage, codon usage and amino acid usage in prokaryotes: similar GC content adopts similar compositional frequencies regardless of the phylogenetic lineages. PloS one, 9(9), e107319. \n",
    "\n",
    "Fleming, J. F., & Struck, T. H. (2023). nRCFV: a new, dataset-size-independent metric to quantify compositional heterogeneity in nucleotide and amino acid datasets. BMC bioinformatics, 24(1), 145.\n",
    "\n",
    "Hernandez, A. M., & Ryan, J. F. (2021). Six-state amino acid recoding is not an effective strategy to offset compositional heterogeneity and saturation in phylogenetic analyses. Systematic Biology, 70(6), 1200-1212. \n",
    "\n",
    "MEOW (2024). https://github.com/jdaneau/pm\n",
    "\n",
    "Williamson, K., Eme, L., Baños, H., McCarthy, C., Susko, E., Kamikawa, R., ... & Roger, A. J. (2024). A robustly rooted tree of eukaryotes reveals their excavate ancestry. bioRxiv, 2024-09. https://doi.org/10.1101/2024.09.04.611237\n",
    "\n",
    "Tice, A. K., Žihala, D., Pánek, T., Jones, R. E., Salomaki, E. D., Nenarokov, S., ... & Brown, M. W. (2021). PhyloFisher: a phylogenomic package for resolving eukaryotic relationships. PLoS Biology, 19(8), e3001365.\n",
    "\n",
    "Crotty, S. M., Minh, B. Q., Bean, N. G., Holland, B. R., Tuke, J., Jermiin, L. S., & Haeseler, A. V. (2020). GHOST: recovering historical signal from heterotachously evolved sequence alignments. Systematic biology, 69(2), 249-264. https://doi.org/10.1093/sysbio/syz051\n",
    "\n",
    "Eglit, Y., Shiratori, T., Jerlström-Hultqvist, J., Williamson, K., Roger, A. J., Ishida, K. I., & Simpson, A. G. (2024). Meteora sporadica, a protist with incredible cell architecture, is related to Hemimastigophora. Current Biology, 34(2), 451-459. https://doi.org/10.1016/j.cub.2023.12.032"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "ptMAGs",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.10.5"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
