{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# rpl36 phylogeny\n",
    "\n",
    "In this notebook, we'll look at the rpl36 phylogeny and alignment. This is an interesting gene because [Rice and Palmer, 2006](https://doi.org/10.1186/1741-7007-4-31) found that haptophytes and cryptophytes share a laterally transferred rpl36 gene, called rpl36-c (c for cryptophyte), which is distinct from the rpl36-p type gene (p for plastid) found in other plastid types. This rpl36-c gene was horizontally transferred from bacteria (donor is difficult to identify), and has completely replaced the native rpl36 gene. \n",
    "\n",
    "If the new group (leptophytes) also has the rpl36-c gene:\n",
    "- it provides further evidence that the new lineage is linked to cryptophytes and haptophytes"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "3.10.5 | packaged by conda-forge | (main, Jun 14 2022, 07:04:59) [GCC 10.3.0]\n"
     ]
    }
   ],
   "source": [
    "# Check if python is 3.10.5\n",
    "import json\n",
    "import os\n",
    "import pandas as pd\n",
    "import sys\n",
    "import numpy as np\n",
    "import __init__\n",
    "\n",
    "\n",
    "print(sys.version)\n",
    "%load_ext autoreload\n",
    "%autoreload 2"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [],
   "source": [
    "# we store the important data paths in PATH_FILE\n",
    "PATH_FILE = \"../../PATHS.json\"\n",
    "\n",
    "paths_dict = json.load(open(PATH_FILE, \"r\"));"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Version 3\n",
    "\n",
    "Extract rpl36 gene from the references and MAGs in the subset phylogeny in Figure 2 (107 taxa, 93 genes).\n",
    "\n",
    "### 3.1 Extract rpl36 gene\n",
    "We prepare the file `rpl36_replaced.fasta`."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### 3.2 Align\n",
    "\n",
    "We align with mafft-linsi."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [],
   "source": [
    "## Define data folder\n",
    "ALL_FASTA = paths_dict[\"ANALYSIS_DATA\"][\"RPL36\"][\"DATASET\"][\"ROOT\"]\n",
    "\n",
    "## Define output folder\n",
    "MAFFT = paths_dict[\"ANALYSIS_DATA\"][\"RPL36\"][\"ALIGNMENTS\"][\"ALIGNED\"]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "%%bash -s \"$ALL_FASTA\" \"$MAFFT\"\n",
    "\n",
    "sbatch /crex/proj/naiss2023-6-81/Mahwash/beta-Cyclocitral/uppmax_scripts/script_bin/job_mafft-linsi.sh \"$1\"/v3/rpl36_replaced.fasta \"$2\"/v3/rpl36.mafft.fasta"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "We trim manually to remove the ends. The final alignment was 103 taxa (after removing emoty sequences) and 48 amino acids."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [],
   "source": [
    "## Define input folder\n",
    "FASTA = paths_dict[\"ANALYSIS_DATA\"][\"RPL36\"][\"ALIGNMENTS\"][\"TRIMMED\"]\n",
    "\n",
    "## Define output folder\n",
    "OUT = paths_dict[\"ANALYSIS_DATA\"][\"RPL36\"][\"TREES\"]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "%%bash -s \"$FASTA\" \"$OUT\"\n",
    "\n",
    "sbatch /crex/proj/naiss2023-6-81/Mahwash/beta-Cyclocitral/uppmax_scripts/script_bin/job_iqtree.sh \"$1\"/v3/rpl36.mafft.trimmed.fasta \"$2\"/v3/rpl36"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "I didn't really use the phylogeny as it the support values were (expectedly) very low. The rpl36 alignment was plotted next to the phylogeny in Figure 2 using the script `plot_tree_rpl36aln.R` (provided in the `src` folder)."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## References\n",
    "\n"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "ptMAGs",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.10.5"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
